{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "728abf83-8476-43c8-b376-c23b297309ed",
   "metadata": {},
   "source": [
    "## STRATEGY for getting AMR gene lists"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7ff0436-f243-423a-9012-28e1eabe67d4",
   "metadata": {},
   "source": [
    "Even if we narrow our search on only ESKAPE pathogens, the number of AMR genes is still pretty high ~ 3.5million entries for ~ 250,000 unique genomes, so here's my plan of action :\n",
    "Extract the AMR genes for each pathogen entry by filtering on the following:\n",
    "- subtype = 'AMR'\n",
    "- scope = 'core'\n",
    "- select first 15 unique genomes\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c827ac9a-fa6e-4f9f-98c1-9936005a8863",
   "metadata": {},
   "source": [
    "The above steps done for 6 pathogen species:\n",
    "- Enterococcus faecium\n",
    "- Staphylococcus aureus\n",
    "- Klebsiella pneumoniae\n",
    "- Acinetobacter baumannii\n",
    "- Pseudomonas aeruginosa\n",
    "- other Enterobacterales group"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e978d815-5e2a-469d-b34d-487b921fb48a",
   "metadata": {},
   "source": [
    "## NEW strategy!!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "848345db-ef9a-4679-8c67-6ba824151b0c",
   "metadata": {},
   "source": [
    "(by Madeline and Parul)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f202caf1-2ab1-4213-8f33-75e46d573a89",
   "metadata": {},
   "source": [
    "1. Downloaded list of all ESKAPE pathogens from ISOLATE dataset wtih:\n",
    "   - LEVEL = 'chromosome'\n",
    "   - LEVEL = 'complete'\n",
    "2. Download all the 'protein-fasta' files using the Genome Accession Ids through ncbi-genome-download tool\n",
    "3. Download list of all AMR genes belonging to ESKAPE pathogens from microbiggE dataset with:\n",
    "   - subtype = 'AMR'\n",
    "   - scope = 'core'\n",
    "4. Match the AMR genes (ID) with the genes (ID) in the protein-fasta to tag them as AMR v/s non-AMR.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "899b0230-a464-4318-9793-85057160a65c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad0babf0-89fe-4869-bffa-b415f42d7bae",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf980625-e568-4ced-b9d7-fe8d6d422403",
   "metadata": {},
   "outputs": [],
   "source": [
    "#read and combine data from tsv files \n",
    "#df1 = pd.read_csv('AMR_metadata/E.tsv', sep='\\t')\n",
    "#df2 = pd.read_csv('AMR_metadata/S.tsv', sep='\\t')\n",
    "#df3 = pd.read_csv('AMR_metadata/K.tsv', sep='\\t')\n",
    "#df4 = pd.read_csv('AMR_metadata/A.tsv', sep='\\t')\n",
    "#df5 = pd.read_csv('AMR_metadata/P.tsv', sep='\\t')\n",
    "#df6 = pd.read_csv('AMR_metadata/otherE.tsv', sep='\\t')\n",
    "\n",
    "#combined_AMR_metadata = pd.concat([df1,df2,df3,df4,df5,df6], ignore_index= 'TRUE')\n",
    "#combined_AMR_metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7065f247-7c2e-43f5-90e7-f889d466d510",
   "metadata": {},
   "source": [
    "### Now let's get the list of genome assemblies we have to download"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "a5ee188c-bed5-4871-9529-bed858117982",
   "metadata": {},
   "outputs": [],
   "source": [
    "# assemblies = combined_AMR_metadata['Assembly'].unique().tolist()\n",
    "# print (assemblies)\n",
    "\n",
    "# # Save the list to a file\n",
    "# with open('AMR_metadata/unique_assemblies.txt', 'w') as f:\n",
    "#     for element in assemblies:\n",
    "#         f.write(f\"{element}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "bd2d2676-1b39-40e5-9a33-15e1aa43896a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ! head -10 AMR_metadata/unique_assemblies.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d1a9268-c331-40aa-addc-05426b90dd1f",
   "metadata": {},
   "source": [
    "### Downloading the above list of genomes using ncbi-genome-download"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "f194ca19-e60d-4369-b9dc-ab6dd28dce09",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Channels:\n",
      " - defaults\n",
      " - bioconda\n",
      "Platform: linux-64\n",
      "Collecting package metadata (repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/maaly7/anaconda3/envs/general\n",
      "\n",
      "  added / updated specs:\n",
      "    - bioconda::ncbi-genome-download\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    brotli-python-1.0.9        |  py311h6a678d5_8         358 KB\n",
      "    certifi-2024.8.30          |  py311h06a4308_0         163 KB\n",
      "    charset-normalizer-3.3.2   |     pyhd3eb1b0_0          44 KB\n",
      "    idna-3.7                   |  py311h06a4308_0         133 KB\n",
      "    ncbi-genome-download-0.3.3 |     pyh7cba7a3_0          29 KB  bioconda\n",
      "    pysocks-1.7.1              |  py311h06a4308_0          35 KB\n",
      "    requests-2.32.3            |  py311h06a4308_0         126 KB\n",
      "    tqdm-4.66.5                |  py311h92b7b1e_0         163 KB\n",
      "    urllib3-2.2.2              |  py311h06a4308_0         227 KB\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:         1.2 MB\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  appdirs            pkgs/main/noarch::appdirs-1.4.4-pyhd3eb1b0_0 \n",
      "  brotli-python      pkgs/main/linux-64::brotli-python-1.0.9-py311h6a678d5_8 \n",
      "  certifi            pkgs/main/linux-64::certifi-2024.8.30-py311h06a4308_0 \n",
      "  charset-normalizer pkgs/main/noarch::charset-normalizer-3.3.2-pyhd3eb1b0_0 \n",
      "  idna               pkgs/main/linux-64::idna-3.7-py311h06a4308_0 \n",
      "  ncbi-genome-downl~ bioconda/noarch::ncbi-genome-download-0.3.3-pyh7cba7a3_0 \n",
      "  pysocks            pkgs/main/linux-64::pysocks-1.7.1-py311h06a4308_0 \n",
      "  requests           pkgs/main/linux-64::requests-2.32.3-py311h06a4308_0 \n",
      "  tqdm               pkgs/main/linux-64::tqdm-4.66.5-py311h92b7b1e_0 \n",
      "  urllib3            pkgs/main/linux-64::urllib3-2.2.2-py311h06a4308_0 \n",
      "\n",
      "\n",
      "\n",
      "Downloading and Extracting Packages:\n",
      "brotli-python-1.0.9  | 358 KB    |                                       |   0% \n",
      "urllib3-2.2.2        | 227 KB    |                                       |   0% \u001b[A\n",
      "\n",
      "certifi-2024.8.30    | 163 KB    |                                       |   0% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "tqdm-4.66.5          | 163 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "idna-3.7             | 133 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "requests-2.32.3      | 126 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "charset-normalizer-3 | 44 KB     |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "pysocks-1.7.1        | 35 KB     |                                       |   0% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "brotli-python-1.0.9  | 358 KB    | #############2                        |  36% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "charset-normalizer-3 | 44 KB     | #############3                        |  36% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "requests-2.32.3      | 126 KB    | ####6                                 |  13% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "pysocks-1.7.1        | 35 KB     | ################7                     |  45% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "certifi-2024.8.30    | 163 KB    | ##################################### | 100% \u001b[A\u001b[A\n",
      "\n",
      "certifi-2024.8.30    | 163 KB    | ##################################### | 100% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "idna-3.7             | 133 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "idna-3.7             | 133 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "brotli-python-1.0.9  | 358 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "urllib3-2.2.2        | 227 KB    | ##################################### | 100% \u001b[A\n",
      "urllib3-2.2.2        | 227 KB    | ##################################### | 100% \u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "charset-normalizer-3 | 44 KB     | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "charset-normalizer-3 | 44 KB     | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "tqdm-4.66.5          | 163 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "tqdm-4.66.5          | 163 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "requests-2.32.3      | 126 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "requests-2.32.3      | 126 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "pysocks-1.7.1        | 35 KB     | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "pysocks-1.7.1        | 35 KB     | ##################################### | 100% \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "                                                                                \u001b[A\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n",
      "\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "conda install bioconda::ncbi-genome-download\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "4c2c9b6d-0999-410c-828c-b7a4425fd0f9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: ncbi-genome-download [-h] [-s {refseq,genbank}] [-F FILE_FORMATS]\n",
      "                            [-l ASSEMBLY_LEVELS] [-g GENERA] [--genus GENERA]\n",
      "                            [--fuzzy-genus] [-S STRAINS] [-T SPECIES_TAXIDS]\n",
      "                            [-t TAXIDS] [-A ASSEMBLY_ACCESSIONS]\n",
      "                            [--fuzzy-accessions] [-R REFSEQ_CATEGORIES]\n",
      "                            [--refseq-category REFSEQ_CATEGORIES] [-o OUTPUT]\n",
      "                            [--flat-output] [-H] [-P] [-u URI] [-p N] [-r N]\n",
      "                            [-m METADATA_TABLE] [-n] [-N] [-v] [-d] [-V]\n",
      "                            [-M TYPE_MATERIALS]\n",
      "                            groups\n",
      "\n",
      "positional arguments:\n",
      "  groups                The NCBI taxonomic groups to download (default: all).\n",
      "                        A comma-separated list of taxonomic groups is also\n",
      "                        possible. For example: \"bacteria,viral\"Choose from:\n",
      "                        ['all', 'archaea', 'bacteria', 'fungi',\n",
      "                        'invertebrate', 'metagenomes', 'plant', 'protozoa',\n",
      "                        'vertebrate_mammalian', 'vertebrate_other', 'viral']\n",
      "\n",
      "options:\n",
      "  -h, --help            show this help message and exit\n",
      "  -s {refseq,genbank}, --section {refseq,genbank}\n",
      "                        NCBI section to download (default: refseq)\n",
      "  -F FILE_FORMATS, --formats FILE_FORMATS\n",
      "                        Which formats to download (default: genbank).A comma-\n",
      "                        separated list of formats is also possible. For\n",
      "                        example: \"fasta,assembly-report\". Choose from:\n",
      "                        ['genbank', 'fasta', 'rm', 'features', 'gff',\n",
      "                        'protein-fasta', 'genpept', 'wgs', 'cds-fasta', 'rna-\n",
      "                        fna', 'rna-fasta', 'assembly-report', 'assembly-\n",
      "                        stats', 'translated-cds', 'all']\n",
      "  -l ASSEMBLY_LEVELS, --assembly-levels ASSEMBLY_LEVELS\n",
      "                        Assembly levels of genomes to download (default: all).\n",
      "                        A comma-separated list of assembly levels is also\n",
      "                        possible. For example: \"complete,chromosome\". Choose\n",
      "                        from: ['all', 'complete', 'chromosome', 'scaffold',\n",
      "                        'contig']\n",
      "  -g GENERA, --genera GENERA\n",
      "                        Only download sequences of the provided genera. A\n",
      "                        comma-seperated list of genera is also possible. For\n",
      "                        example: \"Streptomyces coelicolor,Escherichia coli\".\n",
      "                        (default: [])\n",
      "  --genus GENERA        Deprecated alias of --genera\n",
      "  --fuzzy-genus         Use a fuzzy search on the organism name instead of an\n",
      "                        exact match.\n",
      "  -S STRAINS, --strains STRAINS\n",
      "                        Only download sequences of the given strain(s). A\n",
      "                        comma-separated list of strain names is possible, as\n",
      "                        well as a path to a filename containing one name per\n",
      "                        line.\n",
      "  -T SPECIES_TAXIDS, --species-taxids SPECIES_TAXIDS\n",
      "                        Only download sequences of the provided species NCBI\n",
      "                        taxonomy IDs. A comma-separated list of species taxids\n",
      "                        is also possible. For example: \"52342,12325\".\n",
      "                        (default: [])\n",
      "  -t TAXIDS, --taxids TAXIDS\n",
      "                        Only download sequences of the provided NCBI taxonomy\n",
      "                        IDs. A comma-separated list of taxids is also\n",
      "                        possible. For example: \"9606,9685\". (default: [])\n",
      "  -A ASSEMBLY_ACCESSIONS, --assembly-accessions ASSEMBLY_ACCESSIONS\n",
      "                        Only download sequences matching the provided NCBI\n",
      "                        assembly accession(s). A comma-separated list of\n",
      "                        accessions is possible, as well as a path to a\n",
      "                        filename containing one accession per line.\n",
      "  --fuzzy-accessions    Use a fuzzy search on the entry accession instead of\n",
      "                        an exact match.\n",
      "  -R REFSEQ_CATEGORIES, --refseq-categories REFSEQ_CATEGORIES\n",
      "                        Only download sequences of the provided refseq\n",
      "                        categories [refrerence, representative, na]. A comma-\n",
      "                        separated list of categories is also possible.\n",
      "                        (default: download all categories)\n",
      "  --refseq-category REFSEQ_CATEGORIES\n",
      "                        Deprecated alias for --refseq-categories\n",
      "  -o OUTPUT, --output-folder OUTPUT\n",
      "                        Create output hierarchy in specified folder (default:\n",
      "                        /home/maaly7/notebooks/amr-2024-team-\n",
      "                        nassar/amr_analysis)\n",
      "  --flat-output         Dump all files right into the output folder without\n",
      "                        creating any subfolders.\n",
      "  -H, --human-readable  Create links in human-readable hierarchy (might fail\n",
      "                        on Windows)\n",
      "  -P, --progress-bar    Create a progress bar for indicating the download\n",
      "                        progress\n",
      "  -u URI, --uri URI     NCBI base URI to use (default:\n",
      "                        https://ftp.ncbi.nih.gov/genomes)\n",
      "  -p N, --parallel N    Run N downloads in parallel (default: 1)\n",
      "  -r N, --retries N     Retry download N times when connection to NCBI fails\n",
      "                        (default: 0)\n",
      "  -m METADATA_TABLE, --metadata-table METADATA_TABLE\n",
      "                        Save tab-delimited file with genome metadata\n",
      "  -n, --dry-run         Only check which files to download, don't download\n",
      "                        genome files.\n",
      "  -N, --no-cache        Don't cache the assembly summary file in\n",
      "                        /home/maaly7/.cache/ncbi-genome-download.\n",
      "  -v, --verbose         increase output verbosity\n",
      "  -d, --debug           print debugging information\n",
      "  -V, --version         print version information\n",
      "  -M TYPE_MATERIALS, --type-materials TYPE_MATERIALS\n",
      "                        Specifies the relation to type material for the\n",
      "                        assembly (default: any). \"any\" will include assemblies\n",
      "                        with no relation to type material value defined, \"all\"\n",
      "                        will download only assemblies with a defined value. A\n",
      "                        comma-separated list of relatons. For example:\n",
      "                        \"reference,synonym\". Choose from: ['any', 'all',\n",
      "                        'type', 'reference', 'synonym', 'proxytype',\n",
      "                        'neotype'] .\n"
     ]
    }
   ],
   "source": [
    "! ncbi-genome-download --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "272b7bf2-22a8-406e-a59e-871de46d0d3b",
   "metadata": {},
   "source": [
    "### List of genomes to download (directly downloaded from Isolates browser) are the following:\n",
    "GCAs_to_download_Sa.txt \\\n",
    "GCAs_to_download_Pa.txt \\\n",
    "GCAs_to_download_Kp.txt \\\n",
    "GCAs_to_download_Ef.txt \\\n",
    "GCAs_to_download_E.txt \\\n",
    "GCAs_to_download_Ab.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "12b7ed3e-9de5-4102-8b83-27985825ceb2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Using cached summary.\n",
      "INFO: Checking record 'GCA_000009645.1'\n"
     ]
    }
   ],
   "source": [
    " ! ncbi-genome-download -s genbank -F 'cds-fasta' -A 'GCA_000009645.1' -o genome_assemblies/ --flat-output -v bacteria"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a751545-aad6-4fe8-9cd8-941ddb69f59a",
   "metadata": {},
   "outputs": [],
   "source": [
    "! ncbi-genome-download -s genbank -F 'translated-cds' -A 'GCA_000009645.1' -o genome_assemblies/ --flat-output -v bacteria"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "d1975233-b33c-45ee-a4a1-7e1dd97ec6d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "! gzip -d genome_assemblies/GCA_000009645.1_ASM964v1_cds_from_genomic.fna.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "46891477-cb7e-417c-ac61-d9a3191976e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">lcl|BA000018.3_cds_BAB41217.1_1 [gene=dnaA] [protein=chromosomal replication initiator protein] [protein_id=BAB41217.1] [location=517..1878] [gbkey=CDS]\n",
      "ATGTCGGAAAAAGAAATTTGGGAAAAAGTGCTTGAAATTGCTCAAGAAAAATTATCAGCTGTAAGTTACTCAACTTTCCT\n",
      "AAAAGATACTGAGCTTTACACGATTAAAGATGGTGAAGCTATCGTATTATCGAGTATTCCTTTTAATGCAAATTGGTTAA\n",
      "ATCAACAATATGCTGAAATTATCCAAGCAATCTTATTTGATGTTGTAGGCTATGAAGTTAAACCTCACTTTATTACTACT\n",
      "GAAGAATTAGCAAATTATAGTAATAATGAAACTGCTACTCCAAAAGAAACAACAAAACCTTCTACTGAAACAACTGAGGA\n",
      "TAATCATGTGCTTGGTAGAGAGCAATTCAATGCCCATAACACATTTGACACTTTTGTAATCGGACCTGGTAACCGCTTTC\n",
      "CACATGCAGCAAGTTTAGCTGTAGCCGAAGCACCAGCCAAAGCGTACAATCCATTATTTATCTATGGAGGGGTTGGTTTA\n",
      "GGAAAAACCCATTTAATGCATGCCATTGGTCATCATGTTTTAGATAATAATCCAGATGCCAAAGTGATTTACACATCAAG\n",
      "TGAAAAATTCACAAATGAATTTATTAAATCAATTCGTGATAACGAAGGTGAAGCTTTCAGAGAAAGATATCGTAATATCG\n",
      "ACGTCTTATTAATCGATGATATTCAGTTCATACAAAATAAAGTACAAACACAAGAAGAATTTTTCTATACTTTTAATGAA\n",
      "TTGCATCAGAATAACAAGCAAATAGTTATTTCGAGTGATCGACCACCAAAGGAAATTGCACAATTAGAAGACCGATTACG\n",
      "TTCACGCTTTGAATGGGGGCTAATTGTTGATATTACGCCACCAGATTATGAAACTCGAATGGCAATTTTGCAGAAGAAAA\n",
      "TTGAAGAAGAAAAATTAGATATTCCACCAGAAGCTTTAAATTATATAGCAAATCAAATTCAATCTAATATTCGTGAATTA\n",
      "GAAGGTGCATTAACACGTTTACTTGCATATTCACAATTATTAGGAAAACCAATTACAACTGAATTAACTGCTGAAGCTTT\n",
      "AAAAGATATCATTCAAGCACCAAAATCTAAAAAGATTACCATCCAAGATATTCAAAAAATTGTAGGCCAGTACTATAATG\n",
      "TTAGGATTGAAGATTTCAGTGCAAAAAAACGTACAAAGTCAATTGCATATCCGCGTCAAATAGCTATGTACTTGTCTAGA\n",
      "GAGCTTACAGATTTCTCATTACCTAAAATTGGTGAAGAATTTGGTGGGCGTGATCATACGACCGTCATTCATGCTCATGA\n",
      "AAAAATATCTAAAGATTTAAAAGAAGATCCTATTTTTAAACAAGAAGTAGAGAATCTTGAAAAAGAAATAAGAAATGTAT\n",
      "AA\n",
      ">lcl|BA000018.3_cds_BAB41218.1_2 [gene=dnaN] [protein=DNA polymerase III, beta chain] [protein_id=BAB41218.1] [location=2156..3289] [gbkey=CDS]\n"
     ]
    }
   ],
   "source": [
    "! head -20 genome_assemblies/GCA_000009645.1_ASM964v1_cds_from_genomic.fna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "5f22b620-2993-41ab-8829-b7038acbd1c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "! grep 'WP_001242578.1' genome_assemblies/GCA_000009645.1_ASM964v1_translated_cds.faa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "40025d38-ffec-4a55-a0ec-db301c4ca825",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Using cached summary.\n",
      "INFO: Checking record 'GCA_000750905.1'\n",
      "INFO: Checking record 'GCA_000763245.3'\n",
      "INFO: Checking record 'GCA_000816985.1'\n",
      "INFO: Checking record 'GCA_000829255.1'\n",
      "INFO: Checking record 'GCA_000829275.1'\n",
      "INFO: Checking record 'GCA_000829885.1'\n",
      "INFO: Checking record 'GCA_000981825.1'\n",
      "INFO: Checking record 'GCA_000988505.1'\n",
      "INFO: Checking record 'GCA_001077475.1'\n",
      "INFO: Checking record 'GCA_001293085.1'\n",
      "INFO: Checking record 'GCA_001447845.1'\n",
      "INFO: Checking record 'GCA_001457615.1'\n",
      "INFO: Checking record 'GCA_001482325.1'\n",
      "INFO: Checking record 'GCA_001515845.2'\n",
      "INFO: Checking record 'GCA_001515915.2'\n",
      "INFO: Checking record 'GCA_001516005.1'\n",
      "INFO: Checking record 'GCA_001516105.1'\n",
      "INFO: Checking record 'GCA_001516165.2'\n",
      "INFO: Checking record 'GCA_001516185.2'\n",
      "INFO: Checking record 'GCA_001516205.2'\n",
      "INFO: Checking record 'GCA_001516225.2'\n",
      "INFO: Checking record 'GCA_001516245.2'\n",
      "INFO: Checking record 'GCA_001516265.1'\n",
      "INFO: Checking record 'GCA_001516305.2'\n",
      "INFO: Checking record 'GCA_001516325.2'\n",
      "INFO: Checking record 'GCA_001516345.1'\n",
      "INFO: Checking record 'GCA_001516385.1'\n",
      "INFO: Checking record 'GCA_001518975.1'\n",
      "INFO: Checking record 'GCA_001542795.2'\n",
      "INFO: Checking record 'GCA_001542835.1'\n",
      "INFO: Checking record 'GCA_001547955.1'\n",
      "INFO: Checking record 'GCA_001548135.1'\n",
      "INFO: Checking record 'GCA_001548335.1'\n",
      "INFO: Checking record 'GCA_001594325.2'\n",
      "INFO: Checking record 'GCA_001606045.1'\n",
      "INFO: Checking record 'GCA_001618925.1'\n",
      "INFO: Checking record 'GCA_001632245.1'\n",
      "INFO: Checking record 'GCA_001679685.1'\n",
      "INFO: Checking record 'GCA_001687285.1'\n",
      "INFO: Checking record 'GCA_001721745.1'\n",
      "INFO: Checking record 'GCA_001721765.1'\n",
      "INFO: Checking record 'GCA_001721785.1'\n",
      "INFO: Checking record 'GCA_001721805.1'\n",
      "INFO: Checking record 'GCA_001721825.1'\n",
      "INFO: Checking record 'GCA_001721845.1'\n",
      "INFO: Checking record 'GCA_001722005.2'\n",
      "INFO: Checking record 'GCA_001722025.1'\n",
      "INFO: Checking record 'GCA_001722045.1'\n",
      "INFO: Checking record 'GCA_001729505.1'\n",
      "INFO: Checking record 'GCA_001750705.1'\n",
      "INFO: Checking record 'GCA_001792835.1'\n",
      "INFO: Checking record 'GCA_001792855.1'\n",
      "INFO: Checking record 'GCA_001792875.1'\n",
      "INFO: Checking record 'GCA_001870265.1'\n",
      "INFO: Checking record 'GCA_001874465.1'\n",
      "INFO: Checking record 'GCA_001879525.1'\n",
      "INFO: Checking record 'GCA_001900195.1'\n",
      "INFO: Checking record 'GCA_001900225.1'\n",
      "INFO: Checking record 'GCA_001900265.1'\n",
      "INFO: Checking record 'GCA_002075065.1'\n",
      "INFO: Checking record 'GCA_002085605.1'\n",
      "INFO: Checking record 'GCA_002085755.1'\n",
      "INFO: Checking record 'GCA_002104595.1'\n",
      "INFO: Checking record 'GCA_002104615.1'\n",
      "INFO: Checking record 'GCA_002116945.1'\n",
      "INFO: Checking record 'GCA_002116985.1'\n",
      "INFO: Checking record 'GCA_002117025.1'\n",
      "INFO: Checking record 'GCA_002117045.1'\n",
      "INFO: Checking record 'GCA_002192475.1'\n",
      "INFO: Checking record 'GCA_002192495.1'\n",
      "INFO: Checking record 'GCA_002205335.1'\n",
      "INFO: Checking record 'GCA_002205355.1'\n",
      "INFO: Checking record 'GCA_002205375.1'\n",
      "INFO: Checking record 'GCA_002208645.1'\n",
      "INFO: Checking record 'GCA_002215345.1'\n",
      "INFO: Checking record 'GCA_002223805.1'\n",
      "INFO: Checking record 'GCA_002237405.1'\n",
      "INFO: Checking record 'GCA_002237425.1'\n",
      "INFO: Checking record 'GCA_002287725.2'\n",
      "INFO: Checking record 'GCA_002411865.3'\n",
      "INFO: Checking record 'GCA_002442855.1'\n",
      "INFO: Checking record 'GCA_002753655.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_002812845.1'\n",
      "INFO: Checking record 'GCA_002812865.1'\n",
      "INFO: Checking record 'GCA_002812885.1'\n",
      "INFO: Checking record 'GCA_002812905.2'\n",
      "INFO: Checking record 'GCA_002812925.3'\n",
      "INFO: Checking record 'GCA_002946935.1'\n",
      "INFO: Checking record 'GCA_002968515.1'\n",
      "INFO: Checking record 'GCA_002968585.1'\n",
      "INFO: Checking record 'GCA_002968655.1'\n",
      "INFO: Checking record 'GCA_002968695.1'\n",
      "INFO: Checking record 'GCA_002968955.1'\n",
      "INFO: Checking record 'GCA_002997005.1'\n",
      "INFO: Checking record 'GCA_003028335.1'\n",
      "INFO: Checking record 'GCA_003052005.2'\n",
      "INFO: Checking record 'GCA_003060845.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_003073615.1'\n",
      "INFO: Checking record 'GCA_003073735.1'\n",
      "INFO: Checking record 'GCA_003073795.1'\n",
      "INFO: Checking record 'GCA_003073895.1'\n",
      "INFO: Checking record 'GCA_003074095.1'\n",
      "INFO: Checking record 'GCA_003074115.1'\n",
      "INFO: Checking record 'GCA_003193645.1'\n",
      "INFO: Checking record 'GCA_003194245.1'\n",
      "INFO: Checking record 'GCA_003204335.1'\n",
      "INFO: Checking record 'GCA_003206535.1'\n",
      "INFO: Checking record 'GCA_003288335.1'\n",
      "INFO: Checking record 'GCA_003288435.1'\n",
      "INFO: Checking record 'GCA_003319235.1'\n",
      "INFO: Checking record 'GCA_003332705.2'\n",
      "INFO: Checking record 'GCA_003347525.2'\n",
      "INFO: Checking record 'GCA_003369735.1'\n",
      "INFO: Checking record 'GCA_003369755.1'\n",
      "INFO: Checking record 'GCA_003369775.1'\n",
      "INFO: Checking record 'GCA_003408495.1'\n",
      "INFO: Checking record 'GCA_003411785.2'\n",
      "INFO: Checking record 'GCA_003429185.1'\n",
      "INFO: Checking record 'GCA_003429205.1'\n",
      "INFO: Checking record 'GCA_003433235.1'\n",
      "INFO: Checking record 'GCA_003441515.1'\n",
      "INFO: Checking record 'GCA_003441535.1'\n",
      "INFO: Checking record 'GCA_003441555.1'\n",
      "INFO: Checking record 'GCA_003441575.1'\n",
      "INFO: Checking record 'GCA_003571505.1'\n",
      "INFO: Checking record 'GCA_003571805.1'\n",
      "INFO: Checking record 'GCA_003626935.2'\n",
      "INFO: Checking record 'GCA_003641125.1'\n",
      "INFO: Checking record 'GCA_003665495.1'\n",
      "INFO: Checking record 'GCA_003716765.1'\n",
      "INFO: Checking record 'GCA_003795165.1'\n",
      "INFO: Checking record 'GCA_003798105.1'\n",
      "INFO: Checking record 'GCA_003798125.1'\n",
      "INFO: Checking record 'GCA_003798145.1'\n",
      "INFO: Checking record 'GCA_003812165.1'\n",
      "INFO: Checking record 'GCA_003812885.1'\n",
      "INFO: Checking record 'GCA_003812905.1'\n",
      "INFO: Checking record 'GCA_003813005.1'\n",
      "INFO: Checking record 'GCA_003813025.1'\n",
      "INFO: Checking record 'GCA_003950015.1'\n",
      "INFO: Checking record 'GCA_003952745.1'\n",
      "INFO: Checking record 'GCA_003952765.1'\n",
      "INFO: Checking record 'GCA_003957825.1'\n",
      "INFO: Checking record 'GCA_004000495.1'\n",
      "INFO: Checking record 'GCA_004010895.1'\n",
      "INFO: Checking record 'GCA_004014755.2'\n",
      "INFO: Checking record 'GCA_004102665.1'\n",
      "INFO: Checking record 'GCA_004193735.1'\n",
      "INFO: Checking record 'GCA_004291075.1'\n",
      "INFO: Checking record 'GCA_004346985.1'\n",
      "INFO: Checking record 'GCA_004347005.1'\n",
      "INFO: Checking record 'GCA_004347025.1'\n",
      "INFO: Checking record 'GCA_004347045.1'\n",
      "INFO: Checking record 'GCA_004355125.1'\n",
      "INFO: Checking record 'GCA_004355145.1'\n",
      "INFO: Checking record 'GCA_004769055.1'\n",
      "INFO: Checking record 'GCA_005154325.1'\n",
      "INFO: Checking record 'GCA_005305005.1'\n",
      "INFO: Checking record 'GCA_006364735.1'\n",
      "INFO: Checking record 'GCA_006364795.1'\n",
      "INFO: Checking record 'GCA_006704595.2'\n",
      "INFO: Checking record 'GCA_006971785.1'\n",
      "INFO: Checking record 'GCA_007559065.1'\n",
      "INFO: Checking record 'GCA_007559085.1'\n",
      "INFO: Checking record 'GCA_007559105.1'\n",
      "INFO: Checking record 'GCA_007559125.1'\n",
      "INFO: Checking record 'GCA_007566955.1'\n",
      "INFO: Checking record 'GCA_007567125.1'\n",
      "INFO: Checking record 'GCA_007567285.1'\n",
      "INFO: Checking record 'GCA_007833895.1'\n",
      "INFO: Checking record 'GCA_008033725.1'\n",
      "INFO: Checking record 'GCA_008033745.1'\n",
      "INFO: Checking record 'GCA_008033765.1'\n",
      "INFO: Checking record 'GCA_008195485.1'\n",
      "INFO: Checking record 'GCA_008245185.1'\n",
      "INFO: Checking record 'GCA_008330985.1'\n",
      "INFO: Checking record 'GCA_008705235.1'\n",
      "INFO: Checking record 'GCA_009497675.1'\n",
      "INFO: Checking record 'GCA_009648875.1'\n",
      "INFO: Checking record 'GCA_009662315.1'\n",
      "INFO: Checking record 'GCA_009664165.1'\n",
      "INFO: Checking record 'GCA_009676765.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_009676785.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_009676885.1'\n",
      "INFO: Checking record 'GCA_009720405.1'\n",
      "INFO: Checking record 'GCA_009720425.1'\n",
      "INFO: Checking record 'GCA_009800105.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_009800125.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_009800145.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_009800165.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_009867055.1'\n",
      "INFO: Checking record 'GCA_009905195.1'\n",
      "INFO: Checking record 'GCA_009911735.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_011045375.1'\n",
      "INFO: Checking record 'GCA_011106815.1'\n",
      "INFO: Checking record 'GCA_011466835.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_012276675.1'\n",
      "INFO: Checking record 'GCA_012935295.1'\n",
      "INFO: Checking record 'GCA_012971705.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_013114915.1'\n",
      "INFO: Checking record 'GCA_013114935.1'\n",
      "INFO: Checking record 'GCA_013114955.1'\n",
      "INFO: Checking record 'GCA_013122735.1'\n",
      "INFO: Checking record 'GCA_013201015.1'\n",
      "INFO: Checking record 'GCA_013201095.1'\n",
      "INFO: Checking record 'GCA_013201115.1'\n",
      "INFO: Checking record 'GCA_013255565.1'\n",
      "INFO: Checking record 'GCA_013260445.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_013305645.2'\n",
      "INFO: Checking record 'GCA_013305765.1'\n",
      "INFO: Checking record 'GCA_013305815.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_013305845.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_013341295.1'\n",
      "INFO: Checking record 'GCA_013343315.1'\n",
      "INFO: Checking record 'GCA_013343335.1'\n",
      "INFO: Checking record 'GCA_013343355.1'\n",
      "INFO: Checking record 'GCA_013343375.1'\n",
      "INFO: Checking record 'GCA_013343395.1'\n",
      "INFO: Checking record 'GCA_013343415.1'\n",
      "INFO: Checking record 'GCA_013343435.1'\n",
      "INFO: Checking record 'GCA_013343455.1'\n",
      "INFO: Checking record 'GCA_013343475.1'\n",
      "INFO: Checking record 'GCA_013343495.1'\n",
      "INFO: Checking record 'GCA_013343515.1'\n",
      "INFO: Checking record 'GCA_013343535.1'\n",
      "INFO: Checking record 'GCA_013343555.1'\n",
      "INFO: Checking record 'GCA_013343575.1'\n",
      "INFO: Checking record 'GCA_013347245.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_013375755.1'\n",
      "INFO: Checking record 'GCA_013375775.1'\n",
      "INFO: Checking record 'GCA_013375795.1'\n",
      "INFO: Checking record 'GCA_013375815.1'\n",
      "INFO: Checking record 'GCA_013375835.1'\n",
      "INFO: Checking record 'GCA_013375855.1'\n",
      "INFO: Checking record 'GCA_013375875.1'\n",
      "INFO: Checking record 'GCA_013375895.1'\n",
      "INFO: Checking record 'GCA_013375915.1'\n",
      "INFO: Checking record 'GCA_013376775.1'\n",
      "INFO: Checking record 'GCA_013393665.2'\n",
      "INFO: Checking record 'GCA_013393685.1'\n",
      "INFO: Checking record 'GCA_013393705.1'\n",
      "INFO: Checking record 'GCA_013394455.1'\n",
      "INFO: Checking record 'GCA_013394475.2'\n",
      "INFO: Checking record 'GCA_013395035.1'\n",
      "INFO: Checking record 'GCA_013414905.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_013414925.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_013414945.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_013414965.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_013414985.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_013415005.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_013415025.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_013415045.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_013415065.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_013415085.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_013467585.1'\n",
      "INFO: Checking record 'GCA_013467605.1'\n",
      "INFO: Checking record 'GCA_014109785.1'\n",
      "INFO: Checking record 'GCA_014155905.1'\n",
      "INFO: Checking record 'GCA_014217315.1'\n",
      "INFO: Checking record 'GCA_014467035.1'\n",
      "INFO: Checking record 'GCA_014490645.1'\n",
      "INFO: Checking record 'GCA_014775635.1'\n",
      "INFO: Checking record 'GCA_014792125.1'\n",
      "INFO: Checking record 'GCA_014854655.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_014930935.1'\n",
      "INFO: Checking record 'GCA_015034565.1'\n",
      "INFO: Checking record 'GCA_015697465.1'\n",
      "INFO: Checking record 'GCA_015697605.1'\n",
      "INFO: Checking record 'GCA_015697645.1'\n",
      "INFO: Checking record 'GCA_015697665.1'\n",
      "INFO: Checking record 'GCA_015832055.1'\n",
      "INFO: Checking record 'GCA_015832075.1'\n",
      "INFO: Checking record 'GCA_016064595.1'\n",
      "INFO: Checking record 'GCA_016105505.1'\n",
      "INFO: Checking record 'GCA_016105665.1'\n",
      "INFO: Checking record 'GCA_016105885.1'\n",
      "INFO: Checking record 'GCA_016107485.1'\n",
      "INFO: Checking record 'GCA_016126955.1'\n",
      "INFO: Checking record 'GCA_016134915.1'\n",
      "INFO: Checking record 'GCA_016584725.1'\n",
      "INFO: Checking record 'GCA_016743035.1'\n",
      "INFO: Checking record 'GCA_016743115.1'\n",
      "INFO: Checking record 'GCA_016745115.1'\n",
      "INFO: Checking record 'GCA_016745135.1'\n",
      "INFO: Checking record 'GCA_016745155.1'\n",
      "INFO: Checking record 'GCA_016745175.1'\n",
      "INFO: Checking record 'GCA_016745195.1'\n",
      "INFO: Checking record 'GCA_016811975.1'\n",
      "INFO: Checking record 'GCA_016864415.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_016919425.1'\n",
      "INFO: Checking record 'GCA_016922515.2'\n",
      "INFO: Checking record 'GCA_016923395.2'\n",
      "INFO: Checking record 'GCA_016924255.2'\n",
      "INFO: Checking record 'GCA_016924475.2'\n",
      "INFO: Checking record 'GCA_016924585.2'\n",
      "INFO: Checking record 'GCA_016925455.1'\n",
      "INFO: Checking record 'GCA_016925475.1'\n",
      "INFO: Checking record 'GCA_017378395.1'\n",
      "INFO: Checking record 'GCA_017378475.1'\n",
      "INFO: Checking record 'GCA_017900915.1'\n",
      "INFO: Checking record 'GCA_018138045.1'\n",
      "INFO: Checking record 'GCA_018138065.1'\n",
      "INFO: Checking record 'GCA_018141565.1'\n",
      "INFO: Checking record 'GCA_018141625.1'\n",
      "INFO: Checking record 'GCA_018141645.1'\n",
      "INFO: Checking record 'GCA_018279245.1'\n",
      "INFO: Checking record 'GCA_018279265.1'\n",
      "INFO: Checking record 'GCA_018279285.1'\n",
      "INFO: Checking record 'GCA_018279305.1'\n",
      "INFO: Checking record 'GCA_018409365.1'\n",
      "INFO: Checking record 'GCA_018448985.1'\n",
      "INFO: Checking record 'GCA_018598285.3'\n",
      "INFO: Checking record 'GCA_019084125.3'\n",
      "INFO: Checking record 'GCA_019364615.1'\n",
      "INFO: Checking record 'GCA_019378915.1'\n",
      "INFO: Checking record 'GCA_019379355.1'\n",
      "INFO: Checking record 'GCA_019434075.1'\n",
      "INFO: Checking record 'GCA_019434095.1'\n",
      "INFO: Checking record 'GCA_019434135.1'\n",
      "INFO: Checking record 'GCA_019434235.1'\n",
      "INFO: Checking record 'GCA_019443665.1'\n",
      "INFO: Checking record 'GCA_019466095.1'\n",
      "INFO: Checking record 'GCA_019466145.1'\n",
      "INFO: Checking record 'GCA_019710495.1'\n",
      "INFO: Checking record 'GCA_019711215.1'\n",
      "INFO: Checking record 'GCA_019720855.1'\n",
      "INFO: Checking record 'GCA_019720875.1'\n",
      "INFO: Checking record 'GCA_019738995.2'\n",
      "INFO: Checking record 'GCA_019857245.1'\n",
      "INFO: Checking record 'GCA_019857265.1'\n",
      "INFO: Checking record 'GCA_019857285.1'\n",
      "INFO: Checking record 'GCA_019857305.1'\n",
      "INFO: Checking record 'GCA_019857325.1'\n",
      "INFO: Checking record 'GCA_019857345.1'\n",
      "INFO: Checking record 'GCA_019857365.1'\n",
      "INFO: Checking record 'GCA_019857405.1'\n",
      "INFO: Checking record 'GCA_019857425.1'\n",
      "INFO: Checking record 'GCA_019857445.1'\n",
      "INFO: Checking record 'GCA_019857465.1'\n",
      "INFO: Checking record 'GCA_019857485.1'\n",
      "INFO: Checking record 'GCA_019857505.1'\n",
      "INFO: Checking record 'GCA_019857525.1'\n",
      "INFO: Checking record 'GCA_019857545.1'\n",
      "INFO: Checking record 'GCA_019857565.1'\n",
      "INFO: Checking record 'GCA_019857585.1'\n",
      "INFO: Checking record 'GCA_019857605.1'\n",
      "INFO: Checking record 'GCA_019857625.1'\n",
      "INFO: Checking record 'GCA_019857645.1'\n",
      "INFO: Checking record 'GCA_019915445.1'\n",
      "INFO: Checking record 'GCA_019915465.1'\n",
      "INFO: Checking record 'GCA_019915485.1'\n",
      "INFO: Checking record 'GCA_020771675.1'\n",
      "INFO: Checking record 'GCA_020771755.1'\n",
      "INFO: Checking record 'GCA_020990445.1'\n",
      "INFO: Checking record 'GCA_020990465.1'\n",
      "INFO: Checking record 'GCA_020990485.1'\n",
      "INFO: Checking record 'GCA_021166255.2'\n",
      "INFO: Checking record 'GCA_021166275.2'\n",
      "INFO: Checking record 'GCA_021166295.2'\n",
      "INFO: Checking record 'GCA_021166315.2'\n",
      "INFO: Checking record 'GCA_021166335.2'\n",
      "INFO: Checking record 'GCA_021166355.2'\n",
      "INFO: Checking record 'GCA_021166375.2'\n",
      "INFO: Checking record 'GCA_021184245.1'\n",
      "INFO: Checking record 'GCA_021184265.1'\n",
      "INFO: Checking record 'GCA_021249245.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_021266605.1'\n",
      "INFO: Checking record 'GCA_021378395.1'\n",
      "INFO: Checking record 'GCA_021398755.1'\n",
      "INFO: Checking record 'GCA_021460035.1'\n",
      "INFO: Checking record 'GCA_021497285.1'\n",
      "INFO: Checking record 'GCA_021497305.1'\n",
      "INFO: Checking record 'GCA_021497325.1'\n",
      "INFO: Checking record 'GCA_021497345.2'\n",
      "INFO: Checking record 'GCA_021497365.2'\n",
      "INFO: Checking record 'GCA_021497385.1'\n",
      "INFO: Checking record 'GCA_021497405.1'\n",
      "INFO: Checking record 'GCA_021497425.1'\n",
      "INFO: Checking record 'GCA_021497445.1'\n",
      "INFO: Checking record 'GCA_021513295.1'\n",
      "INFO: Checking record 'GCA_021513375.1'\n",
      "INFO: Checking record 'GCA_021513395.1'\n",
      "INFO: Checking record 'GCA_021513415.1'\n",
      "INFO: Checking record 'GCA_021513435.1'\n",
      "INFO: Checking record 'GCA_021513455.1'\n",
      "INFO: Checking record 'GCA_021513475.1'\n",
      "INFO: Checking record 'GCA_021513495.1'\n",
      "INFO: Checking record 'GCA_021513515.1'\n",
      "INFO: Checking record 'GCA_021513535.1'\n",
      "INFO: Checking record 'GCA_021513555.1'\n",
      "INFO: Checking record 'GCA_021513575.1'\n",
      "INFO: Checking record 'GCA_021513595.1'\n",
      "INFO: Checking record 'GCA_021513615.1'\n",
      "INFO: Checking record 'GCA_021513635.1'\n",
      "INFO: Checking record 'GCA_021513675.1'\n",
      "INFO: Checking record 'GCA_021601605.3'\n",
      "INFO: Checking record 'GCA_021601685.2'\n",
      "INFO: Checking record 'GCA_021601745.3'\n",
      "INFO: Checking record 'GCA_021644285.3'\n",
      "INFO: Checking record 'GCA_021644425.2'\n",
      "INFO: Checking record 'GCA_021644505.2'\n",
      "INFO: Checking record 'GCA_021645225.2'\n",
      "INFO: Checking record 'GCA_021645325.2'\n",
      "INFO: Checking record 'GCA_021645445.2'\n",
      "INFO: Checking record 'GCA_021660795.4'\n",
      "INFO: Checking record 'GCA_021660815.2'\n",
      "INFO: Checking record 'GCA_021660875.3'\n",
      "INFO: Checking record 'GCA_021661475.3'\n",
      "INFO: Checking record 'GCA_021661615.2'\n",
      "INFO: Checking record 'GCA_021661655.3'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_021664795.2'\n",
      "INFO: Checking record 'GCA_021666815.3'\n",
      "INFO: Checking record 'GCA_021666875.3'\n",
      "INFO: Checking record 'GCA_021669275.3'\n",
      "INFO: Checking record 'GCA_021669335.3'\n",
      "INFO: Checking record 'GCA_021669535.2'\n",
      "INFO: Checking record 'GCA_022213165.2'\n",
      "INFO: Checking record 'GCA_022220245.1'\n",
      "INFO: Checking record 'GCA_022424055.2'\n",
      "INFO: Checking record 'GCA_022453765.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_022453825.1'\n",
      "INFO: Checking record 'GCA_022475555.3'\n",
      "INFO: Checking record 'GCA_022533445.1'\n",
      "INFO: Checking record 'GCA_022559545.1'\n",
      "INFO: Checking record 'GCA_022559565.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_022569855.1'\n",
      "INFO: Checking record 'GCA_022569875.1'\n",
      "INFO: Checking record 'GCA_022569895.1'\n",
      "INFO: Checking record 'GCA_022569915.1'\n",
      "INFO: Checking record 'GCA_022569935.1'\n",
      "INFO: Checking record 'GCA_022569955.1'\n",
      "INFO: Checking record 'GCA_022569975.1'\n",
      "INFO: Checking record 'GCA_022569995.1'\n",
      "INFO: Checking record 'GCA_022570115.1'\n",
      "INFO: Checking record 'GCA_022570235.1'\n",
      "INFO: Checking record 'GCA_022570395.1'\n",
      "INFO: Checking record 'GCA_022570415.1'\n",
      "INFO: Checking record 'GCA_022570435.1'\n",
      "INFO: Checking record 'GCA_022570455.1'\n",
      "INFO: Checking record 'GCA_022570475.1'\n",
      "INFO: Checking record 'GCA_022593675.3'\n",
      "INFO: Checking record 'GCA_022637915.1'\n",
      "INFO: Checking record 'GCA_022638055.1'\n",
      "INFO: Checking record 'GCA_022638175.1'\n",
      "INFO: Checking record 'GCA_022649105.1'\n",
      "INFO: Checking record 'GCA_022649245.1'\n",
      "INFO: Checking record 'GCA_022670795.1'\n",
      "INFO: Checking record 'GCA_022699485.1'\n",
      "INFO: Checking record 'GCA_022699505.1'\n",
      "INFO: Checking record 'GCA_022699525.1'\n",
      "INFO: Checking record 'GCA_022700635.2'\n",
      "INFO: Checking record 'GCA_022810865.1'\n",
      "INFO: Checking record 'GCA_022810885.1'\n",
      "INFO: Checking record 'GCA_022810905.1'\n",
      "INFO: Checking record 'GCA_022810925.1'\n",
      "INFO: Checking record 'GCA_022810945.1'\n",
      "INFO: Checking record 'GCA_022810965.1'\n",
      "INFO: Checking record 'GCA_022810985.1'\n",
      "INFO: Checking record 'GCA_022811005.1'\n",
      "INFO: Checking record 'GCA_022811025.1'\n",
      "INFO: Checking record 'GCA_022811045.1'\n",
      "INFO: Checking record 'GCA_022811065.1'\n",
      "INFO: Checking record 'GCA_022811805.1'\n",
      "INFO: Checking record 'GCA_022870025.1'\n",
      "INFO: Checking record 'GCA_022870525.1'\n",
      "INFO: Checking record 'GCA_022870545.1'\n",
      "INFO: Checking record 'GCA_022870565.1'\n",
      "INFO: Checking record 'GCA_022915975.2'\n",
      "INFO: Checking record 'GCA_022916475.3'\n",
      "INFO: Checking record 'GCA_023066825.1'\n",
      "INFO: Checking record 'GCA_023066845.1'\n",
      "INFO: Checking record 'GCA_023066865.1'\n",
      "INFO: Checking record 'GCA_023093935.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_023093955.2'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_023093975.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_023101265.1'\n",
      "INFO: Checking record 'GCA_023101285.1'\n",
      "INFO: Checking record 'GCA_023101305.1'\n",
      "INFO: Checking record 'GCA_023272755.1'\n",
      "INFO: Checking record 'GCA_023380065.1'\n",
      "INFO: Checking record 'GCA_023520735.1'\n",
      "INFO: Checking record 'GCA_023571485.1'\n",
      "INFO: Checking record 'GCA_023572735.3'\n",
      "INFO: Checking record 'GCA_023612335.1'\n",
      "INFO: Checking record 'GCA_023615765.2'\n",
      "INFO: Checking record 'GCA_023822305.1'\n",
      "INFO: Checking record 'GCA_024089395.1'\n",
      "INFO: Checking record 'GCA_024126335.1'\n",
      "INFO: Checking record 'GCA_024126355.1'\n",
      "INFO: Checking record 'GCA_024266805.1'\n",
      "INFO: Checking record 'GCA_024266915.1'\n",
      "INFO: Checking record 'GCA_024266935.1'\n",
      "INFO: Checking record 'GCA_024300845.1'\n",
      "INFO: Checking record 'GCA_024397355.1'\n",
      "INFO: Checking record 'GCA_024507915.1'\n",
      "INFO: Checking record 'GCA_024507955.1'\n",
      "INFO: Checking record 'GCA_024584495.1'\n",
      "INFO: Checking record 'GCA_024652945.3'\n",
      "INFO: Checking record 'GCA_024734795.1'\n",
      "INFO: Checking record 'GCA_024734815.1'\n",
      "INFO: Checking record 'GCA_024760445.1'\n",
      "INFO: Checking record 'GCA_024803805.1'\n",
      "INFO: Checking record 'GCA_024922265.2'\n",
      "INFO: Checking record 'GCA_024972855.1'\n",
      "INFO: Checking record 'GCA_025021565.1'\n",
      "INFO: Checking record 'GCA_025021745.1'\n",
      "INFO: Checking record 'GCA_025200845.1'\n",
      "INFO: Checking record 'GCA_025244965.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_025263585.1'\n",
      "INFO: Checking record 'GCA_025263605.1'\n",
      "INFO: Checking record 'GCA_025263705.1'\n",
      "INFO: Checking record 'GCA_025398075.1'\n",
      "INFO: Checking record 'GCA_025398095.1'\n",
      "INFO: Checking record 'GCA_025398115.1'\n",
      "INFO: Checking record 'GCA_025398135.1'\n",
      "INFO: Checking record 'GCA_025398835.1'\n",
      "INFO: Checking record 'GCA_025452135.1'\n",
      "INFO: Checking record 'GCA_025580225.2'\n",
      "INFO: Checking record 'GCA_025723085.1'\n",
      "INFO: Checking record 'GCA_025734455.1'\n",
      "INFO: Checking record 'GCA_025736955.1'\n",
      "INFO: Checking record 'GCA_025736975.1'\n",
      "INFO: Checking record 'GCA_025790885.1'\n",
      "INFO: Checking record 'GCA_025908315.1'\n",
      "INFO: Checking record 'GCA_025985465.1'\n",
      "INFO: Checking record 'GCA_026625905.1'\n",
      "INFO: Checking record 'GCA_026636135.1'\n",
      "INFO: Checking record 'GCA_026636195.1'\n",
      "INFO: Checking record 'GCA_026727695.1'\n",
      "INFO: Checking record 'GCA_026727715.1'\n",
      "INFO: Checking record 'GCA_026727735.1'\n",
      "INFO: Checking record 'GCA_026727755.1'\n",
      "INFO: Checking record 'GCA_026870135.1'\n",
      "INFO: Checking record 'GCA_027171385.1'\n",
      "INFO: Checking record 'GCA_027359235.1'\n",
      "INFO: Checking record 'GCA_027570415.1'\n",
      "INFO: Checking record 'GCA_027570435.1'\n",
      "INFO: Checking record 'GCA_027570455.1'\n",
      "INFO: Checking record 'GCA_027570475.1'\n",
      "INFO: Checking record 'GCA_027570495.1'\n",
      "INFO: Checking record 'GCA_027570515.1'\n",
      "INFO: Checking record 'GCA_027570535.1'\n",
      "INFO: Checking record 'GCA_027570555.1'\n",
      "INFO: Checking record 'GCA_027570575.1'\n",
      "INFO: Checking record 'GCA_027570595.1'\n",
      "INFO: Checking record 'GCA_027570615.1'\n",
      "INFO: Checking record 'GCA_027570635.1'\n",
      "INFO: Checking record 'GCA_027570655.1'\n",
      "INFO: Checking record 'GCA_027570675.1'\n",
      "INFO: Checking record 'GCA_027570715.1'\n",
      "INFO: Checking record 'GCA_027570735.1'\n",
      "INFO: Checking record 'GCA_027570755.1'\n",
      "INFO: Checking record 'GCA_027570775.1'\n",
      "INFO: Checking record 'GCA_027570795.1'\n",
      "INFO: Checking record 'GCA_027570815.1'\n",
      "INFO: Checking record 'GCA_027570835.1'\n",
      "INFO: Checking record 'GCA_027570855.1'\n",
      "INFO: Checking record 'GCA_027570875.1'\n",
      "INFO: Checking record 'GCA_027570895.1'\n",
      "INFO: Checking record 'GCA_027570915.1'\n",
      "INFO: Checking record 'GCA_027570935.1'\n",
      "INFO: Checking record 'GCA_027570955.1'\n",
      "INFO: Checking record 'GCA_027570975.1'\n",
      "INFO: Checking record 'GCA_027570995.1'\n",
      "INFO: Checking record 'GCA_027571015.1'\n",
      "INFO: Checking record 'GCA_027571035.1'\n",
      "INFO: Checking record 'GCA_027571055.1'\n",
      "INFO: Checking record 'GCA_027571075.1'\n",
      "INFO: Checking record 'GCA_027571095.1'\n",
      "INFO: Checking record 'GCA_027571115.1'\n",
      "INFO: Checking record 'GCA_027571135.1'\n",
      "INFO: Checking record 'GCA_027571155.1'\n",
      "INFO: Checking record 'GCA_027571175.1'\n",
      "INFO: Checking record 'GCA_027571195.1'\n",
      "INFO: Checking record 'GCA_027571215.1'\n",
      "INFO: Checking record 'GCA_027571285.1'\n",
      "INFO: Checking record 'GCA_027571305.1'\n",
      "INFO: Checking record 'GCA_027571325.1'\n",
      "INFO: Checking record 'GCA_027571345.1'\n",
      "INFO: Checking record 'GCA_027571365.1'\n",
      "INFO: Checking record 'GCA_027594685.1'\n",
      "INFO: Checking record 'GCA_027594965.1'\n",
      "INFO: Checking record 'GCA_027594985.1'\n",
      "INFO: Checking record 'GCA_027595005.1'\n",
      "INFO: Checking record 'GCA_027595025.1'\n",
      "INFO: Checking record 'GCA_028053955.3'\n",
      "INFO: Checking record 'GCA_028366335.3'\n",
      "INFO: Checking record 'GCA_028366415.3'\n",
      "INFO: Checking record 'GCA_028366535.3'\n",
      "INFO: Checking record 'GCA_028403965.1'\n",
      "INFO: Checking record 'GCA_028403985.1'\n",
      "INFO: Checking record 'GCA_028404005.1'\n",
      "INFO: Checking record 'GCA_028404025.1'\n",
      "INFO: Checking record 'GCA_028404045.1'\n",
      "INFO: Checking record 'GCA_028404065.1'\n",
      "INFO: Checking record 'GCA_028404085.1'\n",
      "INFO: Checking record 'GCA_028404105.1'\n",
      "INFO: Checking record 'GCA_028404125.1'\n",
      "INFO: Checking record 'GCA_028551535.1'\n",
      "INFO: Checking record 'GCA_028595525.2'\n",
      "INFO: Checking record 'GCA_028595865.2'\n",
      "INFO: Checking record 'GCA_028607125.1'\n",
      "INFO: Checking record 'GCA_028607145.1'\n",
      "INFO: Checking record 'GCA_028607165.1'\n",
      "INFO: Checking record 'GCA_028607185.1'\n",
      "INFO: Checking record 'GCA_028607205.1'\n",
      "INFO: Checking record 'GCA_028607225.1'\n",
      "INFO: Checking record 'GCA_028607245.1'\n",
      "INFO: Checking record 'GCA_028607265.1'\n",
      "INFO: Checking record 'GCA_028607285.1'\n",
      "INFO: Checking record 'GCA_028607305.1'\n",
      "INFO: Checking record 'GCA_028607325.1'\n",
      "INFO: Checking record 'GCA_028607345.1'\n",
      "INFO: Checking record 'GCA_028607365.1'\n",
      "INFO: Checking record 'GCA_028607385.1'\n",
      "INFO: Checking record 'GCA_028608485.1'\n",
      "INFO: Checking record 'GCA_028608505.1'\n",
      "INFO: Checking record 'GCA_028608525.1'\n",
      "INFO: Checking record 'GCA_028608545.1'\n",
      "INFO: Checking record 'GCA_028608565.1'\n",
      "INFO: Checking record 'GCA_028608585.1'\n",
      "INFO: Checking record 'GCA_028608605.1'\n",
      "INFO: Checking record 'GCA_028608625.1'\n",
      "INFO: Checking record 'GCA_028608645.1'\n",
      "INFO: Checking record 'GCA_028608665.1'\n",
      "INFO: Checking record 'GCA_028608685.1'\n",
      "INFO: Checking record 'GCA_028608705.1'\n",
      "INFO: Checking record 'GCA_028608725.1'\n",
      "INFO: Checking record 'GCA_028608745.1'\n",
      "INFO: Checking record 'GCA_028608765.1'\n",
      "INFO: Checking record 'GCA_028608785.1'\n",
      "INFO: Checking record 'GCA_028608805.1'\n",
      "INFO: Checking record 'GCA_028608825.1'\n",
      "INFO: Checking record 'GCA_028608845.1'\n",
      "INFO: Checking record 'GCA_028608865.1'\n",
      "INFO: Checking record 'GCA_028608885.1'\n",
      "INFO: Checking record 'GCA_028608905.1'\n",
      "INFO: Checking record 'GCA_028608925.1'\n",
      "INFO: Checking record 'GCA_028608945.1'\n",
      "INFO: Checking record 'GCA_028608965.1'\n",
      "INFO: Checking record 'GCA_028608985.1'\n",
      "INFO: Checking record 'GCA_028609005.1'\n",
      "INFO: Checking record 'GCA_028609025.1'\n",
      "INFO: Checking record 'GCA_028609045.1'\n",
      "INFO: Checking record 'GCA_028609065.1'\n",
      "INFO: Checking record 'GCA_028609085.1'\n",
      "INFO: Checking record 'GCA_028609105.1'\n",
      "INFO: Checking record 'GCA_028609125.1'\n",
      "INFO: Checking record 'GCA_028609145.1'\n",
      "INFO: Checking record 'GCA_028609165.1'\n",
      "INFO: Checking record 'GCA_028609185.1'\n",
      "INFO: Checking record 'GCA_028609205.1'\n",
      "INFO: Checking record 'GCA_028609225.1'\n",
      "INFO: Checking record 'GCA_028609245.1'\n",
      "INFO: Checking record 'GCA_028609265.1'\n",
      "INFO: Checking record 'GCA_028609285.1'\n",
      "INFO: Checking record 'GCA_028609305.1'\n",
      "INFO: Checking record 'GCA_028609325.1'\n",
      "INFO: Checking record 'GCA_028609345.1'\n",
      "INFO: Checking record 'GCA_028609365.1'\n",
      "INFO: Checking record 'GCA_028609385.1'\n",
      "INFO: Checking record 'GCA_028609405.1'\n",
      "INFO: Checking record 'GCA_028609425.1'\n",
      "INFO: Checking record 'GCA_028609445.1'\n",
      "INFO: Checking record 'GCA_028609465.1'\n",
      "INFO: Checking record 'GCA_028609485.1'\n",
      "INFO: Checking record 'GCA_028609505.1'\n",
      "INFO: Checking record 'GCA_028609525.1'\n",
      "INFO: Checking record 'GCA_028609545.1'\n",
      "INFO: Checking record 'GCA_028609565.1'\n",
      "INFO: Checking record 'GCA_028609585.1'\n",
      "INFO: Checking record 'GCA_028609605.1'\n",
      "INFO: Checking record 'GCA_028609905.1'\n",
      "INFO: Checking record 'GCA_028609925.1'\n",
      "INFO: Checking record 'GCA_028609945.1'\n",
      "INFO: Checking record 'GCA_028609965.1'\n",
      "INFO: Checking record 'GCA_028609985.1'\n",
      "INFO: Checking record 'GCA_028610005.1'\n",
      "INFO: Checking record 'GCA_028610025.1'\n",
      "INFO: Checking record 'GCA_028610045.1'\n",
      "INFO: Checking record 'GCA_028610065.1'\n",
      "INFO: Checking record 'GCA_028610085.1'\n",
      "INFO: Checking record 'GCA_028610105.1'\n",
      "INFO: Checking record 'GCA_028610125.1'\n",
      "INFO: Checking record 'GCA_028610145.1'\n",
      "INFO: Checking record 'GCA_028610165.1'\n",
      "INFO: Checking record 'GCA_028610185.1'\n",
      "INFO: Checking record 'GCA_028610205.1'\n",
      "INFO: Checking record 'GCA_028622015.1'\n",
      "INFO: Checking record 'GCA_028622895.1'\n",
      "INFO: Checking record 'GCA_028743595.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_028751785.1'\n",
      "INFO: Checking record 'GCA_028961985.1'\n",
      "INFO: Checking record 'GCA_029201245.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_029856995.1'\n",
      "INFO: Checking record 'GCA_029857015.1'\n",
      "INFO: Checking record 'GCA_029916825.1'\n",
      "INFO: Checking record 'GCA_029958385.1'\n",
      "INFO: Checking record 'GCA_029961245.1'\n",
      "INFO: Checking record 'GCA_029961265.1'\n",
      "INFO: Checking record 'GCA_029961285.1'\n",
      "INFO: Checking record 'GCA_029961305.1'\n",
      "INFO: Checking record 'GCA_029961325.1'\n",
      "INFO: Checking record 'GCA_029961345.1'\n",
      "INFO: Checking record 'GCA_029961385.1'\n",
      "INFO: Checking record 'GCA_029961405.1'\n",
      "INFO: Checking record 'GCA_029961425.1'\n",
      "INFO: Checking record 'GCA_029961445.1'\n",
      "INFO: Checking record 'GCA_029961465.1'\n",
      "INFO: Checking record 'GCA_029961485.1'\n",
      "INFO: Checking record 'GCA_029961505.1'\n",
      "INFO: Checking record 'GCA_029961525.1'\n",
      "INFO: Checking record 'GCA_029961545.1'\n",
      "INFO: Checking record 'GCA_029961565.1'\n",
      "INFO: Checking record 'GCA_029961585.1'\n",
      "INFO: Checking record 'GCA_029961605.1'\n",
      "INFO: Checking record 'GCA_029961625.1'\n",
      "INFO: Checking record 'GCA_029961645.1'\n",
      "INFO: Checking record 'GCA_029961665.1'\n",
      "INFO: Checking record 'GCA_029962145.1'\n",
      "INFO: Checking record 'GCA_029962165.1'\n",
      "INFO: Checking record 'GCA_029962185.1'\n",
      "INFO: Checking record 'GCA_029962205.1'\n",
      "INFO: Checking record 'GCA_029962505.1'\n",
      "INFO: Checking record 'GCA_030028275.1'\n",
      "INFO: Checking record 'GCA_030034615.1'\n",
      "INFO: Checking record 'GCA_030034635.1'\n",
      "INFO: Checking record 'GCA_030034655.1'\n",
      "INFO: Checking record 'GCA_030034675.1'\n",
      "INFO: Checking record 'GCA_030121855.1'\n",
      "INFO: Checking record 'GCA_030121875.1'\n",
      "INFO: Checking record 'GCA_030121895.1'\n",
      "INFO: Checking record 'GCA_030121915.1'\n",
      "INFO: Checking record 'GCA_030121935.1'\n",
      "INFO: Checking record 'GCA_030121955.1'\n",
      "INFO: Checking record 'GCA_030121975.1'\n",
      "INFO: Checking record 'GCA_030121995.1'\n",
      "INFO: Checking record 'GCA_030122015.1'\n",
      "INFO: Checking record 'GCA_030122035.1'\n",
      "INFO: Checking record 'GCA_030122055.1'\n",
      "INFO: Checking record 'GCA_030122075.1'\n",
      "INFO: Checking record 'GCA_030122095.1'\n",
      "INFO: Checking record 'GCA_030122115.1'\n",
      "INFO: Checking record 'GCA_030122135.1'\n",
      "INFO: Checking record 'GCA_030122155.1'\n",
      "INFO: Checking record 'GCA_030179725.1'\n",
      "INFO: Checking record 'GCA_030179745.1'\n",
      "INFO: Checking record 'GCA_030179765.1'\n",
      "INFO: Checking record 'GCA_030183355.2'\n",
      "INFO: Checking record 'GCA_030183855.2'\n",
      "INFO: Checking record 'GCA_030185815.2'\n",
      "INFO: Checking record 'GCA_030185915.2'\n",
      "INFO: Checking record 'GCA_030186885.2'\n",
      "INFO: Checking record 'GCA_030252465.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_030253475.1'\n",
      "INFO: Checking record 'GCA_030284605.1'\n",
      "INFO: Checking record 'GCA_030316095.1'\n",
      "INFO: Checking record 'GCA_030366015.2'\n",
      "INFO: Checking record 'GCA_030368395.2'\n",
      "INFO: Checking record 'GCA_030368495.2'\n",
      "INFO: Checking record 'GCA_030369815.1'\n",
      "INFO: Checking record 'GCA_030369835.1'\n",
      "INFO: Checking record 'GCA_030369855.1'\n",
      "INFO: Checking record 'GCA_030376865.1'\n",
      "INFO: Checking record 'GCA_030377005.1'\n",
      "INFO: Checking record 'GCA_030388285.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_030413195.1'\n",
      "INFO: Checking record 'GCA_030436095.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_030444495.1'\n",
      "INFO: Checking record 'GCA_030444635.1'\n",
      "INFO: Checking record 'GCA_030444675.1'\n",
      "INFO: Checking record 'GCA_030489625.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_030489645.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_030490735.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_030505595.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_030563585.1'\n",
      "INFO: Checking record 'GCA_030563765.1'\n",
      "INFO: Checking record 'GCA_030564045.1'\n",
      "INFO: Checking record 'GCA_030758435.1'\n",
      "INFO: Checking record 'GCA_030848745.1'\n",
      "INFO: Checking record 'GCA_030866885.1'\n",
      "INFO: Checking record 'GCA_030866905.1'\n",
      "INFO: Checking record 'GCA_030866925.2'\n",
      "INFO: Checking record 'GCA_030908425.1'\n",
      "INFO: Checking record 'GCA_030908445.1'\n",
      "INFO: Checking record 'GCA_031177475.1'\n",
      "INFO: Checking record 'GCA_031211485.1'\n",
      "INFO: Checking record 'GCA_031212645.1'\n",
      "INFO: Checking record 'GCA_031218375.1'\n",
      "INFO: Checking record 'GCA_031219395.1'\n",
      "INFO: Checking record 'GCA_031220265.1'\n",
      "INFO: Checking record 'GCA_031221745.1'\n",
      "INFO: Checking record 'GCA_031222225.1'\n",
      "INFO: Checking record 'GCA_031223175.1'\n",
      "INFO: Checking record 'GCA_031224255.1'\n",
      "INFO: Checking record 'GCA_031224875.1'\n",
      "INFO: Checking record 'GCA_031326305.1'\n",
      "INFO: Checking record 'GCA_031326325.1'\n",
      "INFO: Checking record 'GCA_031851725.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_032231025.1'\n",
      "INFO: Checking record 'GCA_032231045.1'\n",
      "INFO: Checking record 'GCA_032231065.1'\n",
      "INFO: Checking record 'GCA_032231085.1'\n",
      "INFO: Checking record 'GCA_032231105.1'\n",
      "INFO: Checking record 'GCA_032252935.1'\n",
      "INFO: Checking record 'GCA_032252955.1'\n",
      "INFO: Checking record 'GCA_032252975.1'\n",
      "INFO: Checking record 'GCA_032256525.1'\n",
      "INFO: Checking record 'GCA_032258305.1'\n",
      "INFO: Checking record 'GCA_032258985.1'\n",
      "INFO: Checking record 'GCA_032292255.2'\n",
      "INFO: Checking record 'GCA_032463705.1'\n",
      "INFO: Checking record 'GCA_032598925.1'\n",
      "INFO: Checking record 'GCA_032681185.1'\n",
      "INFO: Checking record 'GCA_032696765.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_032699645.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_032761935.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_032811685.1'\n",
      "INFO: Checking record 'GCA_032843885.3'\n",
      "INFO: Checking record 'GCA_032847385.1'\n",
      "INFO: Checking record 'GCA_032934835.1'\n",
      "INFO: Checking record 'GCA_032941915.1'\n",
      "INFO: Checking record 'GCA_033216995.1'\n",
      "INFO: Checking record 'GCA_033217895.1'\n",
      "INFO: Checking record 'GCA_033219075.1'\n",
      "INFO: Checking record 'GCA_033219955.1'\n",
      "INFO: Checking record 'GCA_033221155.1'\n",
      "INFO: Checking record 'GCA_033222185.1'\n",
      "INFO: Checking record 'GCA_033223225.1'\n",
      "INFO: Checking record 'GCA_033224565.1'\n",
      "INFO: Checking record 'GCA_033225445.1'\n",
      "INFO: Checking record 'GCA_033226745.1'\n",
      "INFO: Checking record 'GCA_033227745.1'\n",
      "INFO: Checking record 'GCA_033228545.1'\n",
      "INFO: Checking record 'GCA_033229485.1'\n",
      "INFO: Checking record 'GCA_033230305.1'\n",
      "INFO: Checking record 'GCA_033232085.1'\n",
      "INFO: Checking record 'GCA_033335935.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_033374175.1'\n",
      "INFO: Checking record 'GCA_033375015.1'\n",
      "INFO: Checking record 'GCA_033375815.1'\n",
      "INFO: Checking record 'GCA_033392255.1'\n",
      "INFO: Checking record 'GCA_033393195.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_033394235.1'\n",
      "INFO: Checking record 'GCA_033395015.1'\n",
      "INFO: Checking record 'GCA_033541995.2'\n",
      "INFO: Checking record 'GCA_033547135.1'\n",
      "INFO: Checking record 'GCA_033615795.1'\n",
      "INFO: Checking record 'GCA_033807835.1'\n",
      "INFO: Checking record 'GCA_033971305.1'\n",
      "INFO: Checking record 'GCA_034127025.2'\n",
      "INFO: Checking record 'GCA_034190815.1'\n",
      "INFO: Checking record 'GCA_034258535.1'\n",
      "INFO: Checking record 'GCA_034438375.1'\n",
      "INFO: Checking record 'GCA_034480705.1'\n",
      "INFO: Checking record 'GCA_034480885.1'\n",
      "INFO: Checking record 'GCA_034625575.1'\n",
      "INFO: Checking record 'GCA_035197065.1'\n",
      "INFO: Checking record 'GCA_035197965.1'\n",
      "INFO: Checking record 'GCA_035205985.1'\n",
      "INFO: Checking record 'GCA_035319925.1'\n",
      "INFO: Checking record 'GCA_035320045.1'\n",
      "INFO: Checking record 'GCA_035320405.1'\n",
      "INFO: Checking record 'GCA_035320545.1'\n",
      "INFO: Checking record 'GCA_035320715.1'\n",
      "INFO: Checking record 'GCA_035557195.2'\n",
      "INFO: Checking record 'GCA_035621175.1'\n",
      "INFO: Checking record 'GCA_035621255.1'\n",
      "INFO: Checking record 'GCA_035621395.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_035666215.1'\n",
      "INFO: Checking record 'GCA_035984435.1'\n",
      "INFO: Checking record 'GCA_036232065.1'\n",
      "INFO: Checking record 'GCA_036232085.1'\n",
      "INFO: Checking record 'GCA_036232105.1'\n",
      "INFO: Checking record 'GCA_036232125.1'\n",
      "INFO: Checking record 'GCA_036232145.1'\n",
      "INFO: Checking record 'GCA_036232165.1'\n",
      "INFO: Checking record 'GCA_036232185.1'\n",
      "INFO: Checking record 'GCA_036232205.1'\n",
      "INFO: Checking record 'GCA_036232225.1'\n",
      "INFO: Checking record 'GCA_036232245.1'\n",
      "INFO: Checking record 'GCA_036232265.1'\n",
      "INFO: Checking record 'GCA_036232285.1'\n",
      "INFO: Checking record 'GCA_036232305.1'\n",
      "INFO: Checking record 'GCA_036232325.1'\n",
      "INFO: Checking record 'GCA_036232345.1'\n",
      "INFO: Checking record 'GCA_036232365.1'\n",
      "INFO: Checking record 'GCA_036232385.1'\n",
      "INFO: Checking record 'GCA_036232405.1'\n",
      "INFO: Checking record 'GCA_036232425.1'\n",
      "INFO: Checking record 'GCA_036232455.1'\n",
      "INFO: Checking record 'GCA_036232495.1'\n",
      "INFO: Checking record 'GCA_036232525.1'\n",
      "INFO: Checking record 'GCA_036232565.1'\n",
      "INFO: Checking record 'GCA_036232585.1'\n",
      "INFO: Checking record 'GCA_036232605.1'\n",
      "INFO: Checking record 'GCA_036232625.1'\n",
      "INFO: Checking record 'GCA_036232645.1'\n",
      "INFO: Checking record 'GCA_036232665.1'\n",
      "INFO: Checking record 'GCA_036232685.1'\n",
      "INFO: Checking record 'GCA_036232705.1'\n",
      "INFO: Checking record 'GCA_036232725.1'\n",
      "INFO: Checking record 'GCA_036232745.1'\n",
      "INFO: Checking record 'GCA_036232765.1'\n",
      "INFO: Checking record 'GCA_036232785.1'\n",
      "INFO: Checking record 'GCA_036232805.1'\n",
      "INFO: Checking record 'GCA_036232825.1'\n",
      "INFO: Checking record 'GCA_036232845.1'\n",
      "INFO: Checking record 'GCA_036232865.1'\n",
      "INFO: Checking record 'GCA_036232885.1'\n",
      "INFO: Checking record 'GCA_036232905.1'\n",
      "INFO: Checking record 'GCA_036232925.1'\n",
      "INFO: Checking record 'GCA_036232945.1'\n",
      "INFO: Checking record 'GCA_036233095.1'\n",
      "INFO: Checking record 'GCA_036233115.1'\n",
      "INFO: Checking record 'GCA_036233135.1'\n",
      "INFO: Checking record 'GCA_036233155.1'\n",
      "INFO: Checking record 'GCA_036233175.1'\n",
      "INFO: Checking record 'GCA_036233195.1'\n",
      "INFO: Checking record 'GCA_036233215.1'\n",
      "INFO: Checking record 'GCA_036233235.1'\n",
      "INFO: Checking record 'GCA_036233255.1'\n",
      "INFO: Checking record 'GCA_036233275.1'\n",
      "INFO: Checking record 'GCA_036233295.1'\n",
      "INFO: Checking record 'GCA_036233315.1'\n",
      "INFO: Checking record 'GCA_036233335.1'\n",
      "INFO: Checking record 'GCA_036233355.1'\n",
      "INFO: Checking record 'GCA_036233375.1'\n",
      "INFO: Checking record 'GCA_036233395.1'\n",
      "INFO: Checking record 'GCA_036233415.1'\n",
      "INFO: Checking record 'GCA_036233435.1'\n",
      "INFO: Checking record 'GCA_036233455.1'\n",
      "INFO: Checking record 'GCA_036233475.1'\n",
      "INFO: Checking record 'GCA_036233495.1'\n",
      "INFO: Checking record 'GCA_036233515.1'\n",
      "INFO: Checking record 'GCA_036233535.1'\n",
      "INFO: Checking record 'GCA_036233555.1'\n",
      "INFO: Checking record 'GCA_036233575.1'\n",
      "INFO: Checking record 'GCA_036233595.1'\n",
      "INFO: Checking record 'GCA_036233615.1'\n",
      "INFO: Checking record 'GCA_036233635.1'\n",
      "INFO: Checking record 'GCA_036233655.1'\n",
      "INFO: Checking record 'GCA_036233675.1'\n",
      "INFO: Checking record 'GCA_036233695.1'\n",
      "INFO: Checking record 'GCA_036233715.1'\n",
      "INFO: Checking record 'GCA_036233735.1'\n",
      "INFO: Checking record 'GCA_036233755.1'\n",
      "INFO: Checking record 'GCA_036234315.1'\n",
      "INFO: Checking record 'GCA_036236015.1'\n",
      "INFO: Checking record 'GCA_036236975.1'\n",
      "INFO: Checking record 'GCA_036236995.1'\n",
      "INFO: Checking record 'GCA_036237015.1'\n",
      "INFO: Checking record 'GCA_036237035.1'\n",
      "INFO: Checking record 'GCA_036237055.1'\n",
      "INFO: Checking record 'GCA_036237075.1'\n",
      "INFO: Checking record 'GCA_036237095.1'\n",
      "INFO: Checking record 'GCA_036237115.1'\n",
      "INFO: Checking record 'GCA_036326305.1'\n",
      "INFO: Checking record 'GCA_036417705.1'\n",
      "INFO: Checking record 'GCA_036417805.1'\n",
      "INFO: Checking record 'GCA_036417915.1'\n",
      "INFO: Checking record 'GCA_036417995.1'\n",
      "INFO: Checking record 'GCA_036418055.1'\n",
      "INFO: Checking record 'GCA_036418125.1'\n",
      "INFO: Checking record 'GCA_036418145.1'\n",
      "INFO: Checking record 'GCA_036418185.1'\n",
      "INFO: Checking record 'GCA_036418275.1'\n",
      "INFO: Checking record 'GCA_036418295.1'\n",
      "INFO: Checking record 'GCA_036542105.1'\n",
      "INFO: Checking record 'GCA_036688165.2'\n",
      "INFO: Checking record 'GCA_036776515.1'\n",
      "INFO: Checking record 'GCA_036835625.1'\n",
      "INFO: Checking record 'GCA_036872625.1'\n",
      "INFO: Checking record 'GCA_037039345.1'\n",
      "INFO: Checking record 'GCA_037039355.1'\n",
      "INFO: Checking record 'GCA_037039365.1'\n",
      "INFO: Checking record 'GCA_037039375.1'\n",
      "INFO: Checking record 'GCA_037039385.1'\n",
      "INFO: Checking record 'GCA_037215095.1'\n",
      "INFO: Checking record 'GCA_037223945.1'\n",
      "INFO: Checking record 'GCA_037267025.1'\n",
      "INFO: Checking record 'GCA_037285345.1'\n",
      "INFO: Checking record 'GCA_037287405.1'\n",
      "INFO: Checking record 'GCA_037287425.1'\n",
      "INFO: Checking record 'GCA_037287465.1'\n",
      "INFO: Checking record 'GCA_037287505.1'\n",
      "INFO: Checking record 'GCA_037287535.1'\n",
      "INFO: Checking record 'GCA_037287545.1'\n",
      "INFO: Checking record 'GCA_037287585.1'\n",
      "INFO: Checking record 'GCA_037482125.1'\n",
      "INFO: Checking record 'GCA_037890725.1'\n",
      "INFO: Checking record 'GCA_037914935.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_038069315.2'\n",
      "INFO: Checking record 'GCA_038069335.2'\n",
      "INFO: Checking record 'GCA_038396595.1'\n",
      "INFO: Checking record 'GCA_038396855.1'\n",
      "INFO: Checking record 'GCA_038396885.1'\n",
      "INFO: Checking record 'GCA_038397815.1'\n",
      "INFO: Checking record 'GCA_038418245.2'\n",
      "INFO: Checking record 'GCA_038420425.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_038431885.1'\n",
      "INFO: Checking record 'GCA_038431895.1'\n",
      "INFO: Checking record 'GCA_038431905.1'\n",
      "INFO: Checking record 'GCA_038431915.1'\n",
      "INFO: Checking record 'GCA_038431925.1'\n",
      "INFO: Checking record 'GCA_038431935.1'\n",
      "INFO: Checking record 'GCA_038431945.1'\n",
      "INFO: Checking record 'GCA_038432515.1'\n",
      "INFO: Checking record 'GCA_038433795.1'\n",
      "INFO: Checking record 'GCA_038435155.1'\n",
      "INFO: Checking record 'GCA_038753055.1'\n",
      "INFO: Checking record 'GCA_038755465.1'\n",
      "INFO: Checking record 'GCA_038757835.1'\n",
      "INFO: Checking record 'GCA_038759815.1'\n",
      "INFO: Checking record 'GCA_039512425.1'\n",
      "INFO: Checking record 'GCA_039555165.1'\n",
      "INFO: Checking record 'GCA_039622755.1'\n",
      "INFO: Checking record 'GCA_039622765.1'\n",
      "INFO: Checking record 'GCA_039622815.1'\n",
      "INFO: Checking record 'GCA_039622885.1'\n",
      "INFO: Checking record 'GCA_039702325.1'\n",
      "INFO: Checking record 'GCA_039830125.1'\n",
      "INFO: Checking record 'GCA_039904215.1'\n",
      "INFO: Checking record 'GCA_039908025.1'\n",
      "INFO: Checking record 'GCA_039908075.1'\n",
      "INFO: Checking record 'GCA_039919425.1'\n",
      "INFO: Checking record 'GCA_039921825.1'\n",
      "INFO: Checking record 'GCA_039924275.1'\n",
      "INFO: Checking record 'GCA_040008295.1'\n",
      "INFO: Checking record 'GCA_040084805.1'\n",
      "INFO: Checking record 'GCA_040268335.1'\n",
      "INFO: Checking record 'GCA_040429595.1'\n",
      "INFO: Checking record 'GCA_040429605.1'\n",
      "INFO: Checking record 'GCA_040560445.1'\n",
      "INFO: Checking record 'GCA_040562075.1'\n",
      "INFO: Checking record 'GCA_040571335.1'\n",
      "INFO: Checking record 'GCA_040580825.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_040581105.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_040581215.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_040735975.1'\n",
      "INFO: Checking record 'GCA_040931625.1'\n",
      "INFO: Checking record 'GCA_040931635.1'\n",
      "INFO: Checking record 'GCA_040965195.2'\n",
      "INFO: Checking record 'GCA_041014535.1'\n",
      "INFO: Checking record 'GCA_041080615.1'\n",
      "INFO: Checking record 'GCA_041154335.1'\n",
      "INFO: Checking record 'GCA_041344955.1'\n",
      "INFO: Checking record 'GCA_041344965.1'\n",
      "INFO: Checking record 'GCA_041344975.1'\n",
      "INFO: Checking record 'GCA_041344985.1'\n",
      "INFO: Checking record 'GCA_041378935.1'\n",
      "INFO: Checking record 'GCA_041501525.1'\n",
      "INFO: Checking record 'GCA_041501535.1'\n",
      "INFO: Checking record 'GCA_041501545.1'\n",
      "INFO: Checking record 'GCA_041514425.1'\n",
      "INFO: Checking record 'GCA_041519425.1'\n",
      "INFO: Checking record 'GCA_041519535.1'\n",
      "INFO: Checking record 'GCA_041519645.1'\n",
      "INFO: Checking record 'GCA_041931775.1'\n",
      "INFO: Checking record 'GCA_041931785.1'\n",
      "INFO: Checking record 'GCA_041931795.1'\n",
      "INFO: Checking record 'GCA_041931805.1'\n",
      "INFO: Checking record 'GCA_041931815.1'\n",
      "INFO: Checking record 'GCA_041933115.1'\n",
      "INFO: Checking record 'GCA_041951335.1'\n",
      "INFO: Checking record 'GCA_900070375.1'\n",
      "INFO: Checking record 'GCA_900095805.1'\n",
      "INFO: Checking record 'GCA_900149285.1'\n",
      "INFO: Checking record 'GCA_900497025.1'\n",
      "INFO: Checking record 'GCA_900618245.1'\n",
      "INFO: Checking record 'GCA_900618255.1'\n",
      "INFO: Checking record 'GCA_900618265.1'\n",
      "INFO: Checking record 'GCA_900618275.1'\n",
      "INFO: Checking record 'GCA_900618285.1'\n",
      "INFO: Checking record 'GCA_900618305.1'\n",
      "INFO: Checking record 'GCA_900618315.1'\n",
      "INFO: Checking record 'GCA_900618325.1'\n",
      "INFO: Checking record 'GCA_900618335.1'\n",
      "INFO: Checking record 'GCA_900636735.1'\n",
      "INFO: Checking record 'GCA_900636755.1'\n",
      "INFO: Checking record 'GCA_900636975.1'\n",
      "INFO: Checking record 'GCA_900637045.1'\n",
      "INFO: Checking record 'GCA_902703185.1'\n",
      "INFO: Checking record 'GCA_902703195.1'\n",
      "INFO: Checking record 'GCA_902703215.1'\n",
      "INFO: Checking record 'GCA_904866275.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_905071885.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_951691365.2'\n",
      "INFO: Checking record 'GCA_951802375.2'\n",
      "INFO: Checking record 'GCA_951805275.2'\n",
      "INFO: Checking record 'GCA_000014625.1'\n",
      "INFO: Checking record 'GCA_006974045.1'\n",
      "INFO: Checking record 'GCA_000006765.1'\n",
      "INFO: Checking record 'GCA_013001005.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_030444595.1'\n",
      "INFO: Checking record 'GCA_030444655.1'\n",
      "INFO: Checking record 'GCA_000017205.1'\n",
      "INFO: Checking record 'GCA_000168335.1'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_000026645.1'\n",
      "INFO: Checking record 'GCA_025490375.1'\n",
      "INFO: Checking record 'GCA_025490395.1'\n",
      "INFO: Checking record 'GCA_025490415.1'\n",
      "INFO: Checking record 'GCA_025490435.1'\n",
      "INFO: Checking record 'GCA_025490455.1'\n",
      "INFO: Checking record 'GCA_025490475.1'\n",
      "INFO: Checking record 'GCA_030264075.1'\n",
      "INFO: Checking record 'GCA_000148745.1'\n",
      "INFO: Checking record 'GCA_900185255.1'\n",
      "INFO: Checking record 'GCA_000226155.1'\n",
      "INFO: Checking record 'GCA_902172305.2'\n",
      "INFO: Checking record 'GCA_000223945.2'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_000223965.2'\n",
      "ERROR: No entry for file ending in '_protein.faa.gz'\n",
      "INFO: Checking record 'GCA_033108345.1'\n",
      "INFO: Checking record 'GCA_000284555.1'\n",
      "INFO: Checking record 'GCA_000271365.1'\n",
      "INFO: Checking record 'GCA_001045685.1'\n",
      "INFO: Checking record 'GCA_000714515.1'\n",
      "INFO: Checking record 'GCA_025079255.1'\n",
      "INFO: Checking record 'GCA_022810825.1'\n",
      "INFO: Checking record 'GCA_000271985.2'\n",
      "INFO: Checking record 'GCA_000412695.1'\n",
      "INFO: Checking record 'GCA_000496605.2'\n",
      "INFO: Checking record 'GCA_000496645.1'\n",
      "INFO: Checking record 'GCA_000359505.1'\n",
      "INFO: Checking record 'GCA_000414035.1'\n",
      "INFO: Checking record 'GCA_000468555.1'\n",
      "INFO: Checking record 'GCA_000468935.1'\n",
      "INFO: Checking record 'GCA_900069025.1'\n",
      "INFO: Checking record 'GCA_000484495.1'\n",
      "INFO: Checking record 'GCA_000484545.1'\n",
      "INFO: Checking record 'GCA_000583935.1'\n",
      "INFO: Checking record 'GCA_000508765.1'\n",
      "INFO: Checking record 'GCA_000583955.1'\n",
      "INFO: Checking record 'GCA_000583975.1'\n",
      "INFO: Checking record 'GCA_000583995.1'\n",
      "INFO: Checking record 'GCA_000583895.1'\n",
      "INFO: Checking record 'GCA_000583915.1'\n",
      "INFO: Checking record 'GCA_000496455.2'\n",
      "INFO: Checking record 'GCA_000504045.1'\n",
      "INFO: Checking record 'GCA_000510305.1'\n",
      "INFO: Checking record 'GCA_020978345.1'\n",
      "INFO: Checking record 'GCA_000524595.1'\n",
      "INFO: Checking record 'GCA_000626655.2'\n"
     ]
    }
   ],
   "source": [
    "#Don't run this again!!!!\n",
    "! ncbi-genome-download -s genbank -F 'protein-fasta' -A GCAs_to_download_Pa.txt -o protein_fastas/ --flat-output -v bacteria"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "589c0155-3686-452b-a179-6568e7804ddb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8267\n"
     ]
    }
   ],
   "source": [
    "! ls protein_fastas | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2a3ac02a-f07d-4665-a1b0-d4928abf4653",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCA_000006765.1_ASM676v1_protein.faa\n",
      "GCA_000009005.1_ASM900v1_protein.faa\n",
      "GCA_000009585.1_ASM958v1_protein.faa\n",
      "GCA_000009645.1_ASM964v1_protein.faa\n",
      "GCA_000009665.1_ASM966v1_protein.faa\n",
      "GCA_000009885.1_ASM988v1_protein.faa\n",
      "GCA_000010445.1_ASM1044v1_protein.faa\n",
      "GCA_000010465.1_ASM1046v1_protein.faa\n",
      "GCA_000011265.1_ASM1126v1_protein.faa\n",
      "GCA_000011505.1_ASM1150v1_protein.faa\n",
      "GCA_000011525.1_ASM1152v1_protein.faa\n",
      "GCA_000012045.1_ASM1204v1_protein.faa\n",
      "GCA_000013425.1_ASM1342v1_protein.faa\n",
      "GCA_000013465.1_ASM1346v1_protein.faa\n",
      "GCA_000014625.1_ASM1462v1_protein.faa\n",
      "GCA_000016305.1_ASM1630v1_protein.faa\n",
      "GCA_000016805.1_ASM1680v1_protein.faa\n",
      "GCA_000017085.1_ASM1708v1_protein.faa\n",
      "GCA_000017205.1_ASM1720v1_protein.faa\n",
      "GCA_000018445.1_ASM1844v1_protein.faa\n",
      "ls: write error: Broken pipe\n"
     ]
    }
   ],
   "source": [
    "! ls protein_fastas | head -20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "88eb2b81-1229-4fef-ba94-5be3d3ace0f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "! head -5 GCAs_to_download_Pa.txt > temp.txt\n",
    "! head -5 GCAs_to_download_Sa.txt >> temp.txt\n",
    "! head -5 GCAs_to_download_Kp.txt >> temp.txt\n",
    "! head -5 GCAs_to_download_Ef.txt >> temp.txt\n",
    "! head -5 GCAs_to_download_E.txt >> temp.txt\n",
    "! head -5 GCAs_to_download_Ab.txt >> temp.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "788febd9-66e7-4527-bbcb-9950e2accf17",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCA_000006765.1\n",
      "GCA_000014625.1\n",
      "GCA_000017205.1\n",
      "GCA_000026645.1\n",
      "GCA_000168335.1\n",
      "GCA_000009005.1\n",
      "GCA_000009645.1\n",
      "GCA_000009665.1\n",
      "GCA_000010445.1\n",
      "GCA_000010465.1\n",
      "GCA_000009885.1\n",
      "GCA_000016305.1\n",
      "GCA_000019565.1\n",
      "GCA_000025465.1\n",
      "GCA_000215745.1\n",
      "GCA_000174395.2\n",
      "GCA_000336405.1\n",
      "GCA_000411655.2\n",
      "GCA_000737555.1\n",
      "GCA_001298485.1\n",
      "GCA_000025565.1\n",
      "GCA_000235765.3\n",
      "GCA_000724505.1\n",
      "GCA_000770155.1\n",
      "GCA_001708345.1\n",
      "GCA_000018445.1\n",
      "GCA_000021245.2\n",
      "GCA_000069245.1\n",
      "GCA_000196795.1\n",
      "GCA_000191145.1\n"
     ]
    }
   ],
   "source": [
    "! head -40 temp.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "6ed691d2-c9b0-4487-8b68-64a8efebc6f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "! cat temp.txt | xargs -I {} find protein_fastas/ -name \"{}*.faa\" -print > temp2.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "56bf06b1-deb0-4878-b13a-8ec4436e87b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "! xargs -a temp2.txt -I {} mv {} sub-samples/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bdca63d-250b-467a-a971-482f4e3b193c",
   "metadata": {},
   "source": [
    "###  This means we have 8,267 unique genomes (with protein fastas available) from the ESKAPE pathogens  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "9abb8689-90df-4f6a-b3fd-60670f258e2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract all the gunzipped files  ------ Don't run this again\n",
    "! gzip -d protein_fastas/*.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6ed7802c-19d9-483b-ba65-cf8652f45701",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">AAG03391.1 chromosomal replication initiator protein DnaA [Pseudomonas aeruginosa PAO1]\n",
      "MSVELWQQCVDLLRDELPSQQFNTWIRPLQVEAEGDELRVYAPNRFVLDWVNEKYLGRLLELLGERGEGQLPALSLLIGS\n",
      "KRSRTPRAAIVPSQTHVAPPPPVAPPPAPVQPVSAAPVVVPREELPPVTTAPSVSSDPYEPEEPSIDPLAAAMPAGAAPA\n",
      "VRTERNVQVEGALKHTSYLNRTFTFENFVEGKSNQLARAAAWQVADNLKHGYNPLFLYGGVGLGKTHLMHAVGNHLLKKN\n",
      "PNAKVVYLHSERFVADMVKALQLNAINEFKRFYRSVDALLIDDIQFFARKERSQEEFFHTFNALLEGGQQVILTSDRYPK\n",
      "EIEGLEERLKSRFGWGLTVAVEPPELETRVAILMKKAEQAKIELPHDAAFFIAQRIRSNVRELEGALKRVIAHSHFMGRP\n",
      "ITIELIRESLKDLLALQDKLVSIDNIQRTVAEYYKIKISDLLSKRRSRSVARPRQVAMALSKELTNHSLPEIGVAFGGRD\n",
      "HTTVLHACRKIAQLRESDADIREDYKNLLRTLTT\n",
      ">AAG03392.1 DNA polymerase III, beta chain [Pseudomonas aeruginosa PAO1]\n",
      "MHFTIQREALLKPLQLVAGVVERRQTLPVLSNVLLVVEGQQLSLTGTDLEVELVGRVVLEDAAEPGEITVPARKLMDICK\n",
      "SLPNDVLIDIRVEEQKLLVKAGRSRFTLSTLPANDFPTVEEGPGSLNFSIAQSKLRRLIDRTSFAMAQQDVRYYLNGMLL\n",
      "EVNGGTLRSVATDGHRLAMCSLDAQIPSQDRHQVIVPRKGILELARLLTEQDGEVGIVLGQHHIRATTGEFTFTSKLVDG\n",
      "KFPDYERVLPRGGDKLVVGDRQQLREAFSRTAILSNEKYRGIRLQLSNGLLKIQANNPEQEEAEEEVQVEYNGGNLEIGF\n",
      "NVSYLLDVLGVIGTEQVRFILSDSNSSALVHEADNDDSAYVVMPMRL\n",
      ">AAG03393.1 RecF protein [Pseudomonas aeruginosa PAO1]\n",
      "MSLTRVSVTAVRNLHPVTLSPSPRINILYGDNGSGKTSVLEAIHLLGLARSFRSARLQPVIQYEEAACTVFGQVMLANGI\n",
      "ASNLGISRERQGEFTIRIDGQNARSAAQLAETLPLQLINPDSFRLLEGAPKIRRQFLDWGVFHVEPRFLPVWQRLQKALR\n",
      "QRNSWLRHGKLDPASQAAWDRELSLASDEIDAYRRSYIQALKPVFEETLAELVSLDDLTLSYYRGWDKDRDLLEVLASSL\n",
      "LRDQQMGHTQAGPQRADLRIRLAGHNAAEILSRGQQKLVVCALRIAQGHLINRAKRGQCVYLVDDLPSELDEQHRMALCR\n",
      "LLEDLGCQVFITCVDPQLLKDGWRTDTPVSMFHVEHGKVSQTTTIGSEA\n"
     ]
    }
   ],
   "source": [
    "#! head -20 genome_assemblies/GCA_018086425.1_PDT000701543.1_cds_from_genomic.fna\n",
    "! head -20 protein_fastas/GCA_000006765.1_ASM676v1_protein.faa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4004702a-ba57-430b-a837-b224c3061bf3",
   "metadata": {},
   "source": [
    "## Converting the fasta files into 1 big combined table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d579e8b3-6cf9-4e82-ace9-2eb667c6a423",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "096dc335-4212-4a8d-8344-ba46ebecc32f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_cds_dataframe(fasta_dir):\n",
    "    \"\"\"Creates a DataFrame from all cds FASTA files in a directory.\"\"\"\n",
    "    \n",
    "    data = []\n",
    "    \n",
    "    # Iterate over all files in the directory\n",
    "    for filename in os.listdir(fasta_dir):\n",
    "        # Check if the file ends with '.faa'\n",
    "        if filename.endswith(\".faa\"):\n",
    "            filepath = os.path.join(fasta_dir, filename)\n",
    "            \n",
    "            # Parse the FASTA file and extract headers and sequences\n",
    "            for record in SeqIO.parse(filepath, \"fasta\"):\n",
    "                data.append({\"header\": record.id, \"sequence\": str(record.seq)})\n",
    "    \n",
    "    return pd.DataFrame(data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "d44c58aa-c8e0-4da8-b73a-8ddb70a8333e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "            header                                           sequence\n",
      "0       CAI79689.1  MSEKEIWEKVLKIAQEKLSAVSYSTFLKDTELYTIKDGEAIVLSSI...\n",
      "1       CAI79690.1  MMEFTIKRDYFITQLNDTLKAISPRTTLPILTGIKIDAKEHEVILT...\n",
      "2       CAI79691.1  MIILVQEVVVEGDINLGQFLKTEGIIESGGQAKWFLQDVEVLINGV...\n",
      "3       CAI79692.1  MKLNTLQLENYRNYDEVTLKCHPDVNILIGENAQGKTNLLESIYTL...\n",
      "4       CAI79693.1  MVTALSDVNNTDNYGAGQIQVLEGLEAVRKRPGMYIGSTSERGLHH...\n",
      "...            ...                                                ...\n",
      "118374  UDP42675.1  MNLFKEEIGNWQDWSNVSQSLFAFSALTNFIFQKENLPLAKLENLP...\n",
      "118375  UDP42676.1  MDYGFISTIVRSELFMMQLDSVLVSGAQPNVLSKEIDSFNFMIPIL...\n",
      "118376  UDP42677.1  MPIAIGNKRLPVTLDEKRQKELQQLKQKYGKSESRIMCIALDLLIA...\n",
      "118377  UDP42678.1  DPVSFKPIYGRSGKLETAYWGSRASERQVRMYNKKLEQERKKQIVP...\n",
      "118378  UDP42679.1  MANENQLSEIQSKIEGRGAEKRTTINQLEKLNRISNRRLSVDDKVH...\n",
      "\n",
      "[118379 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "fasta_dir = \"sub-samples/\"  \n",
    "\n",
    "# Create the DataFrame from the FASTA files\n",
    "sub_sample_all_df = create_cds_dataframe(fasta_dir)\n",
    "\n",
    "print(sub_sample_all_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4b63eab4-7ebd-4545-b89b-ffb46ce451fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'QQF40241.1'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#all_df['header'][1]  #### the header is the protein ID !!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "398e8673-ddc7-49db-95dd-f1e3e92d90dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# all_df.to_csv(\"all_protein_df.tsv\", sep='\\t', index=False)  ##Save into a df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "82478504-a988-4e7f-bdbe-4c87a4a14472",
   "metadata": {},
   "outputs": [],
   "source": [
    "#####Subsample the big df\n",
    "\n",
    "# sampled_df = all_df.sample(n=10000, random_state=42)\n",
    "\n",
    "# sampled_df.to_csv(\"subsampled_protein_df.tsv\", sep='\\t', index=False)  ##Save into a df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d6590a6-8f9c-4d7a-925c-bf40f8f7ec36",
   "metadata": {},
   "source": [
    "### Now match the protien IDs with the genes from the AMR table to get AMR v/s non-AMR protein sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "028ddde3-6c2a-432f-b7b3-b5740805ba09",
   "metadata": {},
   "outputs": [],
   "source": [
    "# combined_AMR_metadata['Protein'].tolist()\n",
    "\n",
    "# protein_ids = combined_AMR_metadata['Protein'].tolist()\n",
    "# # Remove NaN values from the list\n",
    "# protein_ids = [protein_id for protein_id in protein_ids if pd.notna(protein_id)]\n",
    "# len(protein_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ebdde8cd-8107-48c0-8657-b87ef1b84b15",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Scientific name</th>\n",
       "      <th>Protein</th>\n",
       "      <th>Assembly</th>\n",
       "      <th>Scope</th>\n",
       "      <th>Subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Staphylococcus aureus subsp. aureus N315</td>\n",
       "      <td>WP_001242578.1</td>\n",
       "      <td>GCA_000009645.1</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Staphylococcus aureus subsp. aureus str. Newman</td>\n",
       "      <td>WP_001100300.1</td>\n",
       "      <td>GCA_000010465.1</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Staphylococcus aureus subsp. aureus MW2</td>\n",
       "      <td>WP_001801873.1</td>\n",
       "      <td>GCA_000011265.1</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Staphylococcus aureus subsp. aureus MSSA476</td>\n",
       "      <td>WP_001033157.1</td>\n",
       "      <td>GCA_000011525.1</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Staphylococcus aureus subsp. aureus COL</td>\n",
       "      <td>WP_063852655.1</td>\n",
       "      <td>GCA_000012045.1</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4050103</th>\n",
       "      <td>Acinetobacter baumannii</td>\n",
       "      <td>HAV4520103.1</td>\n",
       "      <td>GCA_016498365.1</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4050104</th>\n",
       "      <td>Acinetobacter baumannii</td>\n",
       "      <td>HAV4520104.1</td>\n",
       "      <td>GCA_016498365.1</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4050105</th>\n",
       "      <td>Acinetobacter baumannii</td>\n",
       "      <td>HAV4520105.1</td>\n",
       "      <td>GCA_016498365.1</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4050106</th>\n",
       "      <td>Acinetobacter baumannii</td>\n",
       "      <td>HAV4520106.1</td>\n",
       "      <td>GCA_016498365.1</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4050107</th>\n",
       "      <td>Acinetobacter baumannii</td>\n",
       "      <td>HAV4520107.1</td>\n",
       "      <td>GCA_016498365.1</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4050108 rows  5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        #Scientific name         Protein  \\\n",
       "0               Staphylococcus aureus subsp. aureus N315  WP_001242578.1   \n",
       "1        Staphylococcus aureus subsp. aureus str. Newman  WP_001100300.1   \n",
       "2                Staphylococcus aureus subsp. aureus MW2  WP_001801873.1   \n",
       "3            Staphylococcus aureus subsp. aureus MSSA476  WP_001033157.1   \n",
       "4                Staphylococcus aureus subsp. aureus COL  WP_063852655.1   \n",
       "...                                                  ...             ...   \n",
       "4050103                          Acinetobacter baumannii    HAV4520103.1   \n",
       "4050104                          Acinetobacter baumannii    HAV4520104.1   \n",
       "4050105                          Acinetobacter baumannii    HAV4520105.1   \n",
       "4050106                          Acinetobacter baumannii    HAV4520106.1   \n",
       "4050107                          Acinetobacter baumannii    HAV4520107.1   \n",
       "\n",
       "                Assembly Scope Subtype  \n",
       "0        GCA_000009645.1  core     AMR  \n",
       "1        GCA_000010465.1  core     AMR  \n",
       "2        GCA_000011265.1  core     AMR  \n",
       "3        GCA_000011525.1  core     AMR  \n",
       "4        GCA_000012045.1  core     AMR  \n",
       "...                  ...   ...     ...  \n",
       "4050103  GCA_016498365.1  core     AMR  \n",
       "4050104  GCA_016498365.1  core     AMR  \n",
       "4050105  GCA_016498365.1  core     AMR  \n",
       "4050106  GCA_016498365.1  core     AMR  \n",
       "4050107  GCA_016498365.1  core     AMR  \n",
       "\n",
       "[4050108 rows x 5 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "microbigge_df = pd.read_csv('AMR_metadata/All.tsv', sep='\\t')\n",
    "microbigge_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8bf2f52-cdb9-4a02-9b34-619def3acdd5",
   "metadata": {},
   "source": [
    "#### Some summary stats:\n",
    "- Total AMR genes found in the microbigge table for all ESKAPE pathogens = 4,050,108\n",
    "- Of this list, AMR genes with Protein IDs listed = 3,451,878"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "54efcac6-3190-4724-a8fa-8c94dfd4dac2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GCA_000006765.1',\n",
       " 'GCA_000014625.1',\n",
       " 'GCA_000017205.1',\n",
       " 'GCA_000026645.1',\n",
       " 'GCA_000168335.1',\n",
       " 'GCA_000009005.1',\n",
       " 'GCA_000009645.1',\n",
       " 'GCA_000009665.1',\n",
       " 'GCA_000010445.1',\n",
       " 'GCA_000010465.1',\n",
       " 'GCA_000009885.1',\n",
       " 'GCA_000016305.1',\n",
       " 'GCA_000019565.1',\n",
       " 'GCA_000025465.1',\n",
       " 'GCA_000215745.1',\n",
       " 'GCA_000174395.2',\n",
       " 'GCA_000336405.1',\n",
       " 'GCA_000411655.2',\n",
       " 'GCA_000737555.1',\n",
       " 'GCA_001298485.1',\n",
       " 'GCA_000025565.1',\n",
       " 'GCA_000235765.3',\n",
       " 'GCA_000724505.1',\n",
       " 'GCA_000770155.1',\n",
       " 'GCA_001708345.1',\n",
       " 'GCA_000018445.1',\n",
       " 'GCA_000021245.2',\n",
       " 'GCA_000069245.1',\n",
       " 'GCA_000196795.1',\n",
       " 'GCA_000191145.1']"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#extract the same Genomes that I selecetd for the sub-samples\n",
    "with open('temp.txt', 'r') as file:\n",
    "    genome_ids = [genome_ids.strip() for genome_ids in file]\n",
    "genome_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "d31db39c-71fe-4f39-a5c1-cde192c09d76",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Scientific name</th>\n",
       "      <th>Protein</th>\n",
       "      <th>Assembly</th>\n",
       "      <th>Scope</th>\n",
       "      <th>Subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Staphylococcus aureus subsp. aureus N315</td>\n",
       "      <td>WP_001242578.1</td>\n",
       "      <td>GCA_000009645.1</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Staphylococcus aureus subsp. aureus str. Newman</td>\n",
       "      <td>WP_001100300.1</td>\n",
       "      <td>GCA_000010465.1</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Acinetobacter baumannii ACICU</td>\n",
       "      <td>WP_002002480.1</td>\n",
       "      <td>GCA_000018445.1</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Acinetobacter baumannii AB0057</td>\n",
       "      <td>WP_001159761.1</td>\n",
       "      <td>GCA_000021245.2</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>Enterobacter cloacae subsp. dissolvens SDM</td>\n",
       "      <td>WP_043951555.1</td>\n",
       "      <td>GCA_000235765.3</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3927925</th>\n",
       "      <td>Enterobacter cloacae ECNIH2</td>\n",
       "      <td>WP_000186237.1</td>\n",
       "      <td>GCA_000724505.1</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3949091</th>\n",
       "      <td>Enterobacter cloacae ECNIH2</td>\n",
       "      <td>WP_000259031.1</td>\n",
       "      <td>GCA_000724505.1</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3980970</th>\n",
       "      <td>Enterobacter cloacae ECNIH2</td>\n",
       "      <td>WP_012695470.1</td>\n",
       "      <td>GCA_000724505.1</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3987338</th>\n",
       "      <td>Enterobacter cloacae ECNIH2</td>\n",
       "      <td>WP_004152396.1</td>\n",
       "      <td>GCA_000724505.1</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3993167</th>\n",
       "      <td>Enterobacter cloacae ECNIH2</td>\n",
       "      <td>WP_000027057.1</td>\n",
       "      <td>GCA_000724505.1</td>\n",
       "      <td>core</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>224 rows  5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        #Scientific name         Protein  \\\n",
       "0               Staphylococcus aureus subsp. aureus N315  WP_001242578.1   \n",
       "1        Staphylococcus aureus subsp. aureus str. Newman  WP_001100300.1   \n",
       "7                          Acinetobacter baumannii ACICU  WP_002002480.1   \n",
       "8                         Acinetobacter baumannii AB0057  WP_001159761.1   \n",
       "55            Enterobacter cloacae subsp. dissolvens SDM  WP_043951555.1   \n",
       "...                                                  ...             ...   \n",
       "3927925                      Enterobacter cloacae ECNIH2  WP_000186237.1   \n",
       "3949091                      Enterobacter cloacae ECNIH2  WP_000259031.1   \n",
       "3980970                      Enterobacter cloacae ECNIH2  WP_012695470.1   \n",
       "3987338                      Enterobacter cloacae ECNIH2  WP_004152396.1   \n",
       "3993167                      Enterobacter cloacae ECNIH2  WP_000027057.1   \n",
       "\n",
       "                Assembly Scope Subtype  \n",
       "0        GCA_000009645.1  core     AMR  \n",
       "1        GCA_000010465.1  core     AMR  \n",
       "7        GCA_000018445.1  core     AMR  \n",
       "8        GCA_000021245.2  core     AMR  \n",
       "55       GCA_000235765.3  core     AMR  \n",
       "...                  ...   ...     ...  \n",
       "3927925  GCA_000724505.1  core     AMR  \n",
       "3949091  GCA_000724505.1  core     AMR  \n",
       "3980970  GCA_000724505.1  core     AMR  \n",
       "3987338  GCA_000724505.1  core     AMR  \n",
       "3993167  GCA_000724505.1  core     AMR  \n",
       "\n",
       "[224 rows x 5 columns]"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "##select AMR genes in only the genomes listed in 'genome_ids'\n",
    "genome_ids = set(genome_ids)\n",
    "microbigge_subsample_df = microbigge_df[microbigge_df['Assembly'].isin(genome_ids)]\n",
    "\n",
    "microbigge_subsample_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "dc09f2ba-708e-44a7-af6f-9d8eeb71e6b8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "209"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_ids = microbigge_subsample_df['Protein'].tolist()\n",
    "# Remove NaN values from the list\n",
    "protein_ids = [protein_id for protein_id in protein_ids if pd.notna(protein_id)]\n",
    "len(protein_ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0aca148-3ecd-4fe1-85b7-d982e20c345f",
   "metadata": {},
   "source": [
    "### THE PROTEIN ID in MICROBIGGE TABLE DOES NOT MATCH PROTEIN ACCESSION IN PROTEIN FASTA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f99fc667-4536-468c-ae95-54665550aa03",
   "metadata": {},
   "source": [
    "### Hence downloading the protein sequences using protein IDS and match the sequences instead "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "009a490c-1dce-4cc8-8588-7f2c98255d87",
   "metadata": {},
   "outputs": [],
   "source": [
    "#! conda install bioconda::prodigal -y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "80cfec8d-bbe8-474c-a4a7-7c426e8b566c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#! pip install -r ../requirements.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "e42161b8-0239-40a3-9856-f7ff17827836",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import requests as r\n",
    "# from Bio import SeqIO\n",
    "# from io import StringIO\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06f7bc76-bc61-49ff-9210-24dd238bf636",
   "metadata": {},
   "outputs": [],
   "source": [
    "# baseUrl=\"http://www.uniprot.org/uniprot/\"\n",
    "# currentUrl=baseUrl+cID+\".fasta\"\n",
    "# response = r.post(currentUrl)\n",
    "# cData=''.join(response.text)\n",
    "\n",
    "# Seq=StringIO(cData)\n",
    "# pSeq=list(SeqIO.parse(Seq,'fasta'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "4de35db3-da33-4d4c-85ab-afc49b16c36f",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Channels:\n",
      " - defaults\n",
      " - bioconda\n",
      "Platform: linux-64\n",
      "Collecting package metadata (repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/maaly7/anaconda3/envs/general\n",
      "\n",
      "  added / updated specs:\n",
      "    - bioconda::entrez-direct\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    entrez-direct-22.4         |       he881be0_0        14.0 MB  bioconda\n",
      "    libidn2-2.3.4              |       h5eee18b_0         146 KB\n",
      "    libunistring-0.9.10        |       h27cfd23_0         536 KB\n",
      "    wget-1.24.5                |       h251f7ec_0         825 KB\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:        15.5 MB\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  entrez-direct      bioconda/linux-64::entrez-direct-22.4-he881be0_0 \n",
      "  libidn2            pkgs/main/linux-64::libidn2-2.3.4-h5eee18b_0 \n",
      "  libunistring       pkgs/main/linux-64::libunistring-0.9.10-h27cfd23_0 \n",
      "  wget               pkgs/main/linux-64::wget-1.24.5-h251f7ec_0 \n",
      "\n",
      "\n",
      "\n",
      "Downloading and Extracting Packages:\n",
      "entrez-direct-22.4   | 14.0 MB   |                                       |   0% \n",
      "wget-1.24.5          | 825 KB    |                                       |   0% \u001b[A\n",
      "\n",
      "libunistring-0.9.10  | 536 KB    |                                       |   0% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "libidn2-2.3.4        | 146 KB    |                                       |   0% \u001b[A\u001b[A\u001b[A\n",
      "wget-1.24.5          | 825 KB    | #####7                                |  16% \u001b[A\n",
      "\n",
      "\n",
      "libidn2-2.3.4        | 146 KB    | ####                                  |  11% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "entrez-direct-22.4   | 14.0 MB   | ####4                                 |  12% \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "libidn2-2.3.4        | 146 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "\n",
      "libidn2-2.3.4        | 146 KB    | ##################################### | 100% \u001b[A\u001b[A\u001b[A\n",
      "\n",
      "libunistring-0.9.10  | 536 KB    | ##################################### | 100% \u001b[A\u001b[A\n",
      "\n",
      "entrez-direct-22.4   | 14.0 MB   | ##############                        |  38% \u001b[A\u001b[A\n",
      "wget-1.24.5          | 825 KB    | ##################################### | 100% \u001b[A\n",
      "                                                                                \u001b[A\n",
      "                                                                                \u001b[A\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\n",
      "\n",
      "\n",
      "                                                                                \u001b[A\u001b[A\u001b[A\n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n"
     ]
    }
   ],
   "source": [
    "#### trying efetch \n",
    "! conda install bioconda::entrez-direct --yes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "e969df33-0838-4ede-8b3e-567cafda6cbd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "efetch 22.4\n",
      "\n",
      "\u001b[m\u001b[31m\u001b[1m\u001b[7m NOTE: \u001b[m\u001b[31m\u001b[1m EDirect version 22.6 is now available\u001b[m\n",
      "\n",
      "Format Selection\n",
      "\n",
      "  -format        Format of record or report\n",
      "  -mode          text, xml, asn.1, json\n",
      "  -style         master, conwithfeat\n",
      "\n",
      "Direct Record Selection\n",
      "\n",
      "  -db            Database name\n",
      "  -id            Unique identifier or accession number\n",
      "  -input         Read identifier(s) from file instead of stdin\n",
      "\n",
      "Sequence Range\n",
      "\n",
      "  -seq_start     First sequence position to retrieve\n",
      "  -seq_stop      Last sequence position to retrieve\n",
      "  -strand        1 = forward DNA strand, 2 = reverse complement\n",
      "                   (otherwise strand minus is set if start > stop)\n",
      "  -forward       Force strand 1\n",
      "  -revcomp       Force strand 2\n",
      "\n",
      "Gene Range\n",
      "\n",
      "  -chr_start     Sequence range from 0-based coordinates\n",
      "  -chr_stop        in gene docsum GenomicInfoType object\n",
      "\n",
      "Sequence Flags\n",
      "\n",
      "  -complexity    0 = default, 1 = bioseq, 3 = nuc-prot set\n",
      "  -extend        Extend sequence retrieval in both directions\n",
      "  -extrafeat     Bit flag specifying extra features\n",
      "  -showgaps      Propagate component gaps\n",
      "\n",
      "Subset Retrieval\n",
      "\n",
      "  -start         First record to fetch\n",
      "  -stop          Last record to fetch\n",
      "\n",
      "Miscellaneous\n",
      "\n",
      "  -raw           Skip database-specific XML modifications\n",
      "  -express       Direct sequence retrieval in groups of 5 \n",
      "  -immediate     Express mode on a single record at a time \n",
      "\n",
      "Format Examples\n",
      "\n",
      "  -db            -format            -mode    Report Type\n",
      "  ___            _______            _____    ___________\n",
      "\n",
      "  (all)\n",
      "                 docsum                      DocumentSummarySet XML\n",
      "                 docsum             json     DocumentSummarySet JSON\n",
      "                 full                        Same as native except for mesh\n",
      "                 uid                         Unique Identifier List\n",
      "                 url                         Entrez URL\n",
      "                 xml                         Same as -format full -mode xml\n",
      "\n",
      "  bioproject\n",
      "                 native                      BioProject Report\n",
      "                 native             xml      RecordSet XML\n",
      "\n",
      "  biosample\n",
      "                 native                      BioSample Report\n",
      "                 native             xml      BioSampleSet XML\n",
      "\n",
      "  clinvar\n",
      "                 variation                   Older Format\n",
      "                 variationid                 Transition Format\n",
      "                 vcv                         VCV Report\n",
      "                 clinvarset                  RCV Report\n",
      "\n",
      "  gds\n",
      "                 native             xml      RecordSet XML\n",
      "                 summary                     Summary\n",
      "\n",
      "  gene\n",
      "                 full_report                 Detailed Report\n",
      "                 gene_table                  Gene Table\n",
      "                 native                      Gene Report\n",
      "                 native             asn.1    Entrezgene ASN.1\n",
      "                 native             xml      Entrezgene-Set XML\n",
      "                 tabular                     Tabular Report\n",
      "\n",
      "  homologene\n",
      "                 alignmentscores             Alignment Scores\n",
      "                 fasta                       FASTA\n",
      "                 homologene                  Homologene Report\n",
      "                 native                      Homologene List\n",
      "                 native             asn.1    HG-Entry ASN.1\n",
      "                 native             xml      Entrez-Homologene-Set XML\n",
      "\n",
      "  mesh\n",
      "                 full                        Full Record\n",
      "                 native                      MeSH Report\n",
      "                 native             xml      RecordSet XML\n",
      "\n",
      "  nlmcatalog\n",
      "                 native                      Full Record\n",
      "                 native             xml      NLMCatalogRecordSet XML\n",
      "\n",
      "  pmc\n",
      "                 bioc                        PubTator Central BioC XML\n",
      "                 medline                     MEDLINE\n",
      "                 native             xml      pmc-articleset XML\n",
      "\n",
      "  pubmed\n",
      "                 abstract                    Abstract\n",
      "                 apa                         PMID plus APA citation\n",
      "                 bioc                        PubTator Central BioC XML\n",
      "                 medline                     MEDLINE\n",
      "                 native             asn.1    Pubmed-entry ASN.1\n",
      "                 native             xml      PubmedArticleSet XML\n",
      "\n",
      "  (sequences)\n",
      "                 acc                         Accession Number\n",
      "                 est                         EST Report\n",
      "                 fasta                       FASTA\n",
      "                 fasta              xml      TinySeq XML\n",
      "                 fasta_cds_aa                FASTA of CDS Products\n",
      "                 fasta_cds_na                FASTA of Coding Regions\n",
      "                 ft                          Feature Table\n",
      "                 gb                          GenBank Flatfile\n",
      "                 gb                 xml      GBSet XML\n",
      "                 gbc                xml      INSDSet XML\n",
      "                 gene_fasta                  FASTA of Gene\n",
      "                 gp                          GenPept Flatfile\n",
      "                 gp                 xml      GBSet XML\n",
      "                 gpc                xml      INSDSet XML\n",
      "                 gss                         GSS Report\n",
      "                 ipg                         Identical Protein Report\n",
      "                 ipg                xml      IPGReportSet XML\n",
      "                 native             text     Seq-entry ASN.1\n",
      "                 native             xml      Bioseq-set XML\n",
      "                 seqid                       Seq-id ASN.1\n",
      "\n",
      "  snp\n",
      "                 json                        Reference SNP Report\n",
      "\n",
      "  sra\n",
      "                 native             xml      EXPERIMENT_PACKAGE_SET XML\n",
      "                 runinfo            xml      SraRunInfo XML\n",
      "\n",
      "  structure\n",
      "                 mmdb                        Ncbi-mime-asn1 strucseq ASN.1\n",
      "                 native                      MMDB Report\n",
      "                 native             xml      RecordSet XML\n",
      "\n",
      "  taxonomy\n",
      "                 native                      Taxonomy List\n",
      "                 native             xml      TaxaSet XML\n",
      "\n",
      "Examples\n",
      "\n",
      "  efetch -db pubmed -id 6271474,5685784,4882854,6243420 -format xml |\n",
      "  xtract -pattern PubmedArticle -element MedlineCitation/PMID \"#Author\" \\\n",
      "    -block Author -position first -sep \" \" -element Initials,LastName \\\n",
      "    -block Article -element ArticleTitle\n",
      "\n",
      "  efetch -db nuccore -id CM000177.6 -format gb -style conwithfeat -showgaps\n",
      "\n",
      "  efetch -db nuccore -id 1121073309 -format gbc -style master\n",
      "\n",
      "  efetch -db nuccore -id JABRPF010000000 -format gb\n",
      "\n",
      "  efetch -db nuccore -id JABRPF010000001 -format gb\n",
      "\n",
      "  efetch -db protein -id 3OQZ_a -format fasta\n",
      "\n",
      "  esearch -db protein -query \"conotoxin AND mat_peptide [FKEY]\" |\n",
      "  efetch -format fasta -start 1 -stop 5\n",
      "\n",
      "  esearch -db protein -query \"conotoxin AND mat_peptide [FKEY]\" |\n",
      "  efetch -format gpc |\n",
      "  xtract -insd complete mat_peptide \"%peptide\" product mol_wt peptide |\n",
      "  grep -i conotoxin | sort -t $'\\t' -u -k 2,2n | head -n 8\n",
      "\n",
      "  esearch -db gene -query \"DDT [GENE] AND mouse [ORGN]\" |\n",
      "  efetch -format docsum |\n",
      "  xtract -pattern GenomicInfoType -element ChrAccVer ChrStart ChrStop |\n",
      "  xargs -n 3 sh -c 'efetch -db nuccore -format gb \\\n",
      "    -id \"$0\" -chr_start \"$1\" -chr_stop \"$2\"'\n",
      "\n",
      "  efetch -db sra -format docsum -id SRA030738,SRA030736 |\n",
      "  xtract -pattern DocumentSummary -element Run@acc\n",
      "\n",
      "\u001b[m\u001b[31m\u001b[1m\u001b[7m NOTE: \u001b[m\u001b[31m\u001b[1m EDirect version 22.6 is now available\u001b[m\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! efetch --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "2b1898a6-791a-497d-9768-38e873ca6478",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">WP_001242578.1 MULTISPECIES: bleomycin binding protein [Bacteria]\n",
      "MRMLQSIPALPVGDIKKSIGFYCDKLGFTLVHHEDGFAVLMCNEVRIHLWEASDEGWRSRSNDSPVCTGA\n",
      "ESFIAGTASCRIEVEGIDELYQHIKPLGILHPNTSLKDQWWDERDFAVIDPDNNLISFFQQIKS\n"
     ]
    }
   ],
   "source": [
    "! efetch -db protein -format fasta -id WP_001242578.1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d29e266b-42af-4edf-8d84-98706326e0d2",
   "metadata": {},
   "source": [
    "#### Running efetch on bulk to get all protein sequences in a dataframe format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "fd03107e-4fad-4f56-821d-0d139f80c34a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         protein_ID                                           sequence\n",
      "0    WP_001242578.1  MRMLQSIPALPVGDIKKSIGFYCDKLGFTLVHHEDGFAVLMCNEVR...\n",
      "1    WP_001100300.1  MNVEYSKIKKAVPILLFLFVFSLVIDNSFKLISVAIADDLNISVTT...\n",
      "2    WP_002002480.1  MKLLKILSLVCLSISIGACAEHSMSRAKTSTIPQVNNSIIDQNVQA...\n",
      "3    WP_001159761.1  MQFKKISCLLLSPLFIFSTSIYADNTPKDQEIKKLVDQNFKPLLEK...\n",
      "4    WP_043951555.1  MMKKSLSCALLLSVACSAFAAPMSEKQLANVVERNVTPLMKAQAIP...\n",
      "..              ...                                                ...\n",
      "204  WP_000186237.1  MTNYFDSPFKGKLLSEQVKNPNIKVGRYSYYSGYYHGHSFDDCARY...\n",
      "205  WP_000259031.1  MVTVFGILNLTEDSFFDESRRLDPAGAVTAAIEMLRVGSDVVDVGP...\n",
      "206  WP_012695470.1  MSHIQRETSCSRPRLNSNLDVDLYGYRWARDNVGQSGATIYRLYGK...\n",
      "207  WP_004152396.1  MSLYRRLVLLSCLSWPLAGFSATALTNLVAEPFAKLEQDFGGSIGV...\n",
      "208  WP_000027057.1  MSIQHFRVALIPFFAAFCLPVFAHPETLVKVKDAEDQLGARVGYIE...\n",
      "\n",
      "[209 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "# Initialize an empty list to store the data\n",
    "protein_data = []\n",
    "\n",
    "# Loop through each protein ID\n",
    "for protein_id in protein_ids:\n",
    "    # Run the efetch command using subprocess\n",
    "    result = subprocess.run(\n",
    "        ['efetch', '-db', 'protein', '-format', 'fasta', '-id', protein_id],\n",
    "        capture_output=True, text=True\n",
    "    )\n",
    "    \n",
    "    # Get the output from efetch (FASTA format)\n",
    "    fasta_output = result.stdout\n",
    "\n",
    "    # Extract the sequence (everything after the first line)\n",
    "    fasta_lines = fasta_output.split('\\n')\n",
    "    sequence = ''.join(fasta_lines[1:])  # Join lines after the header to get the full sequence\n",
    "    \n",
    "    # Append protein ID and sequence to the data list\n",
    "    protein_data.append({'protein_ID': protein_id, 'sequence': sequence})\n",
    "\n",
    "# Convert the list to a DataFrame\n",
    "df = pd.DataFrame(protein_data)\n",
    "\n",
    "print(df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "62658acc-b94a-4e31-9557-62b5e5d9d2d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_ID</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WP_001242578.1</td>\n",
       "      <td>MRMLQSIPALPVGDIKKSIGFYCDKLGFTLVHHEDGFAVLMCNEVR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WP_001100300.1</td>\n",
       "      <td>MNVEYSKIKKAVPILLFLFVFSLVIDNSFKLISVAIADDLNISVTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WP_002002480.1</td>\n",
       "      <td>MKLLKILSLVCLSISIGACAEHSMSRAKTSTIPQVNNSIIDQNVQA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WP_001159761.1</td>\n",
       "      <td>MQFKKISCLLLSPLFIFSTSIYADNTPKDQEIKKLVDQNFKPLLEK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WP_043951555.1</td>\n",
       "      <td>MMKKSLSCALLLSVACSAFAAPMSEKQLANVVERNVTPLMKAQAIP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>WP_000186237.1</td>\n",
       "      <td>MTNYFDSPFKGKLLSEQVKNPNIKVGRYSYYSGYYHGHSFDDCARY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>WP_000259031.1</td>\n",
       "      <td>MVTVFGILNLTEDSFFDESRRLDPAGAVTAAIEMLRVGSDVVDVGP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>WP_012695470.1</td>\n",
       "      <td>MSHIQRETSCSRPRLNSNLDVDLYGYRWARDNVGQSGATIYRLYGK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>WP_004152396.1</td>\n",
       "      <td>MSLYRRLVLLSCLSWPLAGFSATALTNLVAEPFAKLEQDFGGSIGV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>WP_000027057.1</td>\n",
       "      <td>MSIQHFRVALIPFFAAFCLPVFAHPETLVKVKDAEDQLGARVGYIE...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>209 rows  2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         protein_ID                                           sequence\n",
       "0    WP_001242578.1  MRMLQSIPALPVGDIKKSIGFYCDKLGFTLVHHEDGFAVLMCNEVR...\n",
       "1    WP_001100300.1  MNVEYSKIKKAVPILLFLFVFSLVIDNSFKLISVAIADDLNISVTT...\n",
       "2    WP_002002480.1  MKLLKILSLVCLSISIGACAEHSMSRAKTSTIPQVNNSIIDQNVQA...\n",
       "3    WP_001159761.1  MQFKKISCLLLSPLFIFSTSIYADNTPKDQEIKKLVDQNFKPLLEK...\n",
       "4    WP_043951555.1  MMKKSLSCALLLSVACSAFAAPMSEKQLANVVERNVTPLMKAQAIP...\n",
       "..              ...                                                ...\n",
       "204  WP_000186237.1  MTNYFDSPFKGKLLSEQVKNPNIKVGRYSYYSGYYHGHSFDDCARY...\n",
       "205  WP_000259031.1  MVTVFGILNLTEDSFFDESRRLDPAGAVTAAIEMLRVGSDVVDVGP...\n",
       "206  WP_012695470.1  MSHIQRETSCSRPRLNSNLDVDLYGYRWARDNVGQSGATIYRLYGK...\n",
       "207  WP_004152396.1  MSLYRRLVLLSCLSWPLAGFSATALTNLVAEPFAKLEQDFGGSIGV...\n",
       "208  WP_000027057.1  MSIQHFRVALIPFFAAFCLPVFAHPETLVKVKDAEDQLGARVGYIE...\n",
       "\n",
       "[209 rows x 2 columns]"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "amr_protein_df = df\n",
    "amr_protein_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90fc09f8-4617-4c23-906c-8cccfcf22dc6",
   "metadata": {},
   "source": [
    "### Get the AMR v/s non-AMR sequences by matching the Protein sequence in both"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6b285d3a-74a8-4ac6-a853-573152d00f9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# amr_df = pd.DataFrame()\n",
    "# #protein_ids = combined_AMR_metadata['Protein'].tolist()\n",
    "\n",
    "# # Iterate over the list of protein IDs and collect matching rows\n",
    "# for protein_id in protein_ids:\n",
    "#     # Get rows that match the protein ID in the 'header' column\n",
    "#     matches = all_df[all_df['header'] == protein_id]\n",
    "    \n",
    "#     # Append matching rows to the matching_df\n",
    "#     amr_df = pd.concat([amr_df, matches])\n",
    "\n",
    "# # After processing all protein IDs, the non-matching rows are the remaining ones\n",
    "# non_amr_df = all_df[~all_df.index.isin(amr_df.index)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "48977a12-7258-4275-8bd6-7d6d83638378",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'MAKLLIMSVVSFCFIFLLLVFFRYILKRYFNYSLNYKVWYLTMLAGLIPFIPIKFSFIKFNNVNNQAPTVESKSHDLNHNINTTKPIQEFTTDIHKFNWDSIDNICTVIWIVLVIILSFKFLKALLYLKYLKKQSLYLNENEKNKIDTILFNHQYKKNIVIRKAEAIQSPITFWYGKYIILIPSSYFKSVIDKRLKYIILHEYAHAKNRDTLHLIIFNIFSIIMSYNPLVHIVKRKIIHDNEVEADRFVLNNINKNEFKTYAESIMDSVLKTPFSNKNILSHSFNGKKSLLKSRLINIKEADLKKQSKLILIFICIFTFFIMIIQSQFLMGQSLTDYNYKKPLQSDYQILDESKNFGSNSGSFVMYSMKKDKYYIYNEKESRKRYSPDSTYKIYLALFGLDRHIISDKNSRMSWNHKHYLFESWNKEQDLNTAMQNSVNWYFERISNQIPKNYTAAQLKQLNYGNENLGSYKSYWMEDSLKISNLEQVIVFKNMMEQNNHFSKKAKNQLSSSLLIKKNEKYELYGKTGTGIVNGKYNNGWFVGYVITNHDKYYFATHLSDGNPSGKNAELISEKILKGMGVLNDQ',\n",
       " 'MAKMRISPELKKLIEKYRCVKDTEGMSPAKVYKLVGENENLYLKMTDSRYKGTTYDVEREKDMMLWLEGKLPVPKVLHFERHDGWSNLLMSEADGVLCSEEYEDEQSPEKIIELYAECIRLFHSIDISDCPYTNSLDSRLAELDYLLNNDLADVDCENWEEDTPFKDPRELYDFLKTEKPEEELVFSHGDLGDSNIFVKDGKVSGFIDLGRSGRADKWYDIAFCVRSIREDIGEEQYVELFFDLLGIKPDWEKIKYYILLDELF',\n",
       " 'MDFSRFFIDRPIFAAVLSILIFITGLIAIPLLPVSEYPDVVPPSVQVRAEYPGANPKVIAETVATPLEEAINGVENMMYMKSVAGSDGVLVTTVTFRPGTDPDQAQVQVQNRVAQAEARLPEDVRRLGITTQKQSPTLTLVVHLFSPGGKYDSLYMRNYATLKVKDELARLPGVGQIQIFGSGEYAMRVWLDPNKVAARGLTASDVVTAMQEQNVQVSAGQLGAEPLPKESDFLISINAQGRLHTEDEFGNIVLKTTQDGTVVRLRDVARVEMGSGSYALRSQLNNKDAVGIGIFQSPGANAIDLSNAVRAKMDELSTRFPDDMKWAAPYDPTVFVRDSIRAVVQTLLEAVVLVVLVVILFLQTWRASIIPLIAVPVSVVGTFSILYLLGFSLNTLSLFGLVLAIGIVVDDAIVVVENVERNIEEGLAPLAAAHQAMREVSGPIIAIALVLCAVFVPMAFLSGVTGQFYKQFAVTIAISTVISAINSLTLSPALAALLLKPHGAPKDFPTRLIDRLFGWIFRPFNRFFHRSSTGYQGLVGKTLGRRGAVFVVYLLLLCTAGVMFKAVPGGFIPTQDKLYLIGGVKMPEGSSLARTDAVIRKMSEIGMNTEGVDYAVAFPGLNALQFTNTPNTGTVFFGLKPFDQRKHTAAEINAEINAKIAQIQQGFGFSILPPPILGLGQGSGYSLYIQDRAGLGYGALQNAVNTMSGAIMQTPGMHFPISTYQANVPQLDVQVDRDKAKAQGVSLTDLFGTLQTYLGSSYVNDFNQFGRTWRVMAQADGQFRDSVEDIANLRTRNSQGEMVPIGSMVKITTTYGPDPVIRYNGYPAADLIGDADPRVLSSAQAMTQLDGMSKQILPNGMNIEWTDLSFQQATQGNTALIVFPVAVLLAFLVLAALYESWTLPLAVILIVPMTMLSALFGVWLTGGDNNVFVQVGLVVLMGLACKNAILIVEFARELEIQGKGIMEAALEACRLRLRPIVMTSIAFIAGTIPLILGHGAGAEVRGVTGITVFAGMLGVTLFGLFLTPVFYVTLRKFVTRGKAEREVLPA',\n",
       " 'MDFSRFFIDRPIFAAVLSILIFITGLIAIPLLPVSEYPDVVPPSVQVRAEYPGANPKVIAETVATPLEEAINGVENMMYMKSVAGSDGVLVTTVTFRPGTDPDQAQVQVQNRVAQAEARLPEDVRRLGITTQKQSPTLTLVVHLFSPKGKYDSLYMRNYATLKVKDELARLPGVGQIQIFGSGEYAMRVWLDPNKVAARGLTASDVVTAMQEQNVQVSAGQLGAEPLPQESDFLISINAQGRLHTEEEFGNIILKTAQDGSLVRLRDVARIEMGSGSYALRSQLNNKDAVGIGIFQSPGANAIDLSNAVRAKMAELATRFPEDMQWAAPYDPTVFVRDSIRAVVQTLLEAVVLVVLVVILFLQTWRASIIPLIAVPVSVVGTFSILYLLGFSLNTLSLFGLVLAIGIVVDDAIVVVENVERNIEEGLAPLAAAHQAMREVSGPIIAIALVLCAVFVPMAFLSGVTGQFYKQFAVTIAISTVISAINSLTLSPALAALLLKPHGAKKDLPTRLIDRLFGWIFRPFNRFFLRSSNGYQGLVGKTLGRRGAVFAVYLLLLCAAGVMFKVVPGGFIPTQDKLYLIGGVKMPEGSSLARTDAVIRKMSEIGMNTEGVDYAVAFPGLNALQFTNTPNTGTVFFGLKPFDQRKHTAAEINAEINAKIAQIQQGFGFSILPPPILGLGQGSGYSLYIQDRGGLGYGALQSAVNAMSGAIMQTPGMHFPISTYQANVPQLDVQVDRDKAKAQGVSLTELFGTLQTYLGSSYVNDFNQFGRTWRVMAQADGPYRESVEDIANLRTRNNQGEMVPIGSMVNISTTYGPDPVIRYNGYPAADLIGDADPRVLSSSQAMTHLEELSKQILPNGMNIEWTDLSFQQATQGNTALIVFPVAVLLAFLVLAALYESWTLPLAVILIVPMTMLSALFGVWLTGGDNNVFVQVGLVVLMGLACKNAILIVEFARELEIQGKGIMEAALEACRLRLRPIVMTSIAFIAGTIPLILGHGAGAEVRGVTGITVFSGMLGVTLFGLFLTPVFYVTLRKLVTRRKPVQEDLPA',\n",
       " 'MDFSRFFIDRPIFAAVLSILIFITGLIAIPLLPVSEYPDVVPPSVQVRAEYPGANPKVIAETVATPLEEAINGVENMMYMKSVAGSDGVLVTTVTFRPGTDPDQAQVQVQNRVAQAEARLPEDVRRLGITTQKQSPTLTLVVHLFSPNGKYDSLYMRNYATLKVKDELARLPGVGQIQIFGSGEYAMRVWLDPNKVAARGLTASDVVTAMQEQNVQVSAGQLGAEPLPQESDFLISINAQGRLHTEEEFGNIILKTAQDGSLVRLRDVARIEMGSGSYALRSQLNNKDAVGIGIFQSPGANAIDLSNAVRAKMAELATRFPEDMQWAAPYDPTVFVRDSIRAVVQTLLEAVVLVVLVVILFLQTWRASIIPLIAVPVSVVGTFSILYLLGFSLNTLSLFGLVLAIGIVVDDAIVVVENVERNIEEGLAPLAAAHQAMREVSGPIIAIALVLCAVFVPMAFLSGVTGQFYKQFAVTIAISTVISAINSLTLSPALAALLLKPHGAKKDLPTRLIDRLFGWIFRPFNRFFLRSSNGYQGLVSKTLGRRGAVFAVYLLLLCAAGVMFKVVPGGFIPTQDKLYLIGGVKMPEGSSLARTDAVIRKMSEIGMNTEGVDYAVAFPGLNALQFTNTPNTGTVFFGLKPFDQRKHTAAEINAEINAKIAQIQQGFGFSILPPPILGLGQGSGYSLYIQDRGGLGYGALQSAVNAMSGAIMQTPGMHFPISTYQANVPQLDVQVDRDKAKAQGVSLTELFGTLQTYLGSSYVNDFNQFGRTWRVMAQADGPYRESVEDIANLRTRNNQGEMVPIGSMVNISTTYGPDPVIRYNGYPAADLIGDADPRVLSSSQAMTHLEELSKQILPNGMNIEWTDLSFQQATQGNTALIVFPVAVLLAFLVLAALYESWTLPLAVILIVPMTMLSALFGVWLTGGDNNVFVQVGLVVLMGLACKNAILIVEFARELEIQGKGIMEAALEACRLRLRPIVMTSIAFIAGTIPLILGHGAGAEVRGVTGITVFSGMLGVTLFGLFLTPVFYVTLRKLVTRRKPVQEDLPA',\n",
       " 'MDFSRFFIDRPIFAAVLSILIFITGLIAIPLLPVSEYPDVVPPSVQVRAEYPGANPKVIAETVATPLEEAINGVENMMYMKSVAGSDGVLVTTVTFRPGTDPDQAQVQVQNRVSQAEARLPEDVRRLGITTQKQSPTLTLVVHLFSPGGKYDSLYMRNYATLKVKDELARLPGVGQIQIFGSGEYAMRVWLDPNKVAARGLTASDVVTAMQEQNVQVSAGQLGAEPLPKESDFLISINAQGRLHTEEEFGNIILKTAQDGSLVRLRDVARIEMGSGSYALRSQLNNKDAVGIGIFQSPGANAIDLSNAVRAKMAELSTRFPEDMKWAAPYDPTVFVRDSIRAVVQTLLEAVVLVVLVVILFLQTWRASIIPLIAVPVSVVGTFSILYLLGFSLNTLSLFGLVLAIGIVVDDAIVVVENVERNIEEGLAPLAAAHQAMREVSGPIIAIALVLCAVFVPMAFLSGVTGQFYKQFAVTIAISTVISAINSLTLSPALAALLLKPHGAPKDLPTRLIDRLFGWIFRPFNRFFHRSSNGYQGLVGKTLGRRGAVFVVYVLLLCAAGVMFKAVPGGFIPTQDKLYLIGGVKMPEGSSLARTDAVIRKMSEIGMNTEGVDYAVAFPGLNALQFTNTPNTGTVFFGLKPFDQRKHTAAEINAEINAKIAQIQQGFGFSILPPPILGLGQGSGYSLYIQDRAGLGYGALQNAVNTMSGAIMQTPGMHFPISTYQANVPQLDVQVDRDKAKAQGVSLTDLFGTLQTYLGSSYVNDFNQFGRTWRVMAQADGPYRDSVEDIANLRTRNNQGEMVPIGSMVNISTTYGPDPVIRYNGYPAADLIGDADPRVLSSTEAMTQLEGMSKQILPNGMNIEWTDLSFQQATQGNTALIVFPVAVLLAFLVLAALYESWTLPLAVILIVPMTMLSALFGVWLTGGDNNVFVQVGLVVLMGLACKNAILIVEFARELEIQGKGIMEAALEACRLRLRPIVMTSIAFIAGTIPLILGHGAGAEVRGVTGITVFSGMLGVTLFGLFLTPVFYVTLRKFVTRGKVEREVLPA',\n",
       " 'MDFSRFFIDRPIFAAVLSILIFITGLIAIPLLPVSEYPDVVPPSVQVRAEYPGANPKVIAETVATPLEEAINGVENMMYMKSVAGSDGVLVTTVTFRPGTDPDQAQVQVQNRVSQAEARLPEDVRRLGITTQKQSPTLTLVVHLFSPGGKYDSLYMRNYATLKVKDELARLPGVGQIQIFGSGEYAMRVWLDPNKVAARGLTASDVVTAMQEQNVQVSAGQLGAEPLPKESDFLISINAQGRLHTEEEFGNIILKTAQDGSLVRLRDVARIEMGSGSYALRSQLNNKDAVGIGIFQSPGANAIDLSNAVRAKMAELSTRFPEDMKWAAPYDPTVFVRDSIRAVVQTLLEAVVLVVLVVILFLQTWRASIIPLIAVPVSVVGTFSVLYLLGFSLNTLSLFGLVLAIGIVVDDAIVVVENVERNIEEGLAPLAAAHQAMREVSGPIIAIALVLCAVFVPMAFLSGVTGQFYKQFAVTIAISTVISAINSLTLSPALAALLLKPHGAPKDLPTRLIDRLFGWIFRPFNRFFHRSSNGYQGLVGKTLGRRGAVFVVYVLLLCAAGVMFKAVPGGFIPTQDKLYLIGGVKMPEGSSLARTDAVIRKMSEIGMNTEGVDYAVAFPGLNALQFTNTPNTGTVFFGLKPFDQRKHTAAEINAEINAKIAQIQQGFGFSILPPPILGLGQGSGYSLYIQDRAGLGYGALQNAVNTMSGAIMQTPGMHFPISTYQANVPQLDVQIDRDKAKAQGVSLTDLFGTLQTYLGSSYVNDFNQFGRTWRVMAQADGPYRDSVEDIANLRTRNNQGEMVPIGSMVNISTTYGPDPVIRYNGYPAADLIGDADPRVLSSTQAMTQLEGMSKQILPNGMNIEWTDLSFQQATQGNTALIVFPVAVLLAFLVLAALYESWTLPLAVILIVPMTMLSALFGVWLTGGDNNVFVQVGLVVLMGLACKNAILIVEFARELEIQGKGIMEAALEACRLRLRPIVMTSIAFIAGTIPLILGHGAGAEVRGVTGITVFSGMLGVTLFGLFLTPVFYVTLRKFVTRGKVEREVLPA',\n",
       " 'MDFSRFFIDRPIFAAVLSILIFITGLIAIPLLPVSEYPDVVPPSVQVRAEYPGANPKVIAETVATPLEEAINGVENMMYMKSVAGSDGVLVTTVTFRPGTDPDQAQVQVQNRVSQAEARLPEDVRRLGITTQKQSPTLTLVVHLFSPGGKYDSLYMRNYATLKVKDELARLPGVGQIQIFGSGEYAMRVWLDPNKVAARGLTASDVVTAMQEQNVQVSAGQLGAEPLPKESDFLISINAQGRLHTEEEFGNIILKTAQDGSLVRLRDVARIEMGSGSYALRSQLNNKDAVGIGIFQSPGANAIDLSNAVRAKMAELSTRFPEDMKWAAPYDPTVFVRDSIRAVVQTLLEAVVLVVMVVILFLQTWRASIIPLIAVPVSVVGTFSVLYLLGFSLNTLSLFGLVLAIGIVVDDAIVVVENVERNIEEGLAPLAAAHQAMREVSGPIIAIALVLCAVFVPMAFLSGVTGQFYKQFAVTIAISTVISAINSLTLSPALAALLLKPHGAPKDLPTRLIDRLFGWIFRPFNRFFHRSSNGYQGLVGKTLGRRGAVFVVYVLLLCAAGVMFKAVPGGFIPTQDKLYLIGGVKMPEGSSLARTDAVIRKMSEIGMNTEGVDYAVAFPGLNALQFTNTPNTGTVFFGLKPFDQRKHTAAEINAEINAKIAQIQQGFGFSILPPPILGLGQGSGYSLYIQDRAGLGYGALQNAVNTMSGAIMQTPGMHFPISTYQANVPQLDVQVDRDKAKAQGVSLTDLFGTLQTYLGSSYVNDFNQFGRTWRVMAQADGPYRDSVEDIANLRTRNNQGEMVPIGSMVNISTTYGPDPVIRYNGYPAADLIGDADPRVLSSTQAMTQLEGMSKQILPNGMNIEWTDLSFQQATQGNTALIVFPVAVLLAFLVLAALYESWTLPLAVILIVPMTMLSALFGVWLTGGDNNVFVQVGLVVLMGLACKNAILIVEFARELEIQGKGIMEAALEACRLRLRPIVMTSIAFIAGTIPLILGHGAGAEVRGVTGITVFSGMLGVTLFGLFLTPVFYVTLRKFVTRGKVEREVLPA',\n",
       " 'MDHATPNLPSRDFDATVAFYERLGFRLRFRDTGWMILERGGLVLEFLRHPELDPLTSWFSCCLRLDDLAGFYAHCEAAGITEATKGYPRLHAPQAQEWGGTMAALIDLDGTLLRLIANR',\n",
       " 'MDIIDKVFQQEDFSRQDLSDSRFRRCRFYQCDFSHCQLQDASFEDCSFIESGAVEGCHFSYADLRDASFKACRLSLANFSGANCFGIEFRECDLKGANFSRARFYNQVSHKMYFCSAYISGCNLAYTNLSGQCLEKCELFENNWSNANLSGASLMGSDLSRGTFSRDCWQQVNLRGCDLTFADLDGLDPRRVNLEGVKICAWQQEQLLEPLGVIVLPD',\n",
       " 'MDNKTYEISSAEWEFMNIIWMKKYASANNIIEEIQMQKDWSPKTIRTLITRLYKKGFIDRKKDNKIFQYYSLVEESDIKYKTSKNFINKVYKGGFNSLVLNFVEKEDLSQDEIEELRNILNKK',\n",
       " 'MDNKTYEISSAEWEVMNIIWMKKYASANNIIEEIQMQKDWSPKTIRTLITRLYKKGFIDRKKDNKIFQYYSLVEESDIKYKTSKNFINKVYKGGFNSLVLNFVEKEDLSQDEIEELRNILNKK',\n",
       " 'MDTTQVTLIHKILAAADERNLPLWIGGGWAIDARLGRVTRKHDDIDLTFPGERRGELEAIVEMLGGRVMEELDYGFLAEIGDELLDCEPAWWADEAYEIAEAPQGSCPEAAEGVIAGRPVRCNSWEAIIWDYFYYADEVPPVDWPTKHIESYRLACTSLGAEKVEVLRAAFRSRYAA',\n",
       " 'MEKKITGYTTVDISQWHRKEHFEAFQSVAQCTYNQTVQLDITAFLKTVKKNKHKFYPAFIHILARLMNAHPEFRMAMKDGELVIWDSIHPCYTVFHEQTETFSSLWSEYHDDFRQFLHIYSQDVACYGENLAYFPKGFIENMFFVSANPWVSFTSFDLNVANMDNFFAPVFTMGKYYTQGDKVLMPLAIQVHHAVCDGFHVGRMLNELQQYCNEWQGGA',\n",
       " 'MEKKITGYTTVDISQWHRKEHFEAFQSVAQCTYNQTVQLDITAFLKTVKKNKHKFYPAFIHILARLMNAHPEFRMAMKDGELVIWDSVHPCYTVFHEQTETFSSLWSEYHDDFRQFLHIYSQDVACYGENLAYFPKGFIENMFFVSANPWVSFTSFDLNVANMDNFFAPVFTMGKYYTQGDKVLMPLAIQVHHAVCDGFHVGRMLNELQQYCDEWQGGA',\n",
       " 'MENLAVNITNLQVSFGNQLELSIDSLRVYQQDRIGIIGENGVGKSTLLKLIAGELFPDHGKIQTEITFNYLPQLTYLAEAKDLNLELASHFQLRLEETSERKWSGGEERKIELIRLLSSYEQGMLLDEPTTHLDRKSIDRLIEELRYYYGTLVFVSHDRYFLDELASKIWEVKDGEIREFSGNYSAYLTQKELEKKTQLREAESIMKEKKRLEKSIQEKKKQAEKLEKVSSKKKKQQIRPDRLSSSKQKDSVQKAIQKNAKTLERRLQKIGETTKPQQMKQIRFPVPKSLELHSRYPIMGQNVQLERSGRTLLVNGDFQFSLGKKIAIVGENGSGKTTLLEHIRKQGEGILLSPKVSFQVYQQKGYQMTSEESIIRFVMRQTEFSESLVRSLLNHLGFAQETLTKPLCTLSGGEATRLTIALLFTKPSNVLLLDEPTNFIDMATIEALEKLMQIYPGTILFTSHDSYFVERTADEVYEIKGQKIKKVLTRNF',\n",
       " 'MENLAVNITNLQVSFGNQLELSIDSLRVYQQDRIGIIGENGVGKSTLLKLIAGELFPDHGKIQTEITFNYLPQLTYLAEAKDLNLELASHFQLRLEETSERKWSGGEERKIELIRLLSSYEQGMLLDEPTTHLDRKSIDRLIEELRYYYGTLVFVSHDRYFLDELASKIWEVKDGEIREFSGNYSAYLTQKELEKKTQLREAESIMKEKKRLEKSIQEKKKQAEKLEKVSSKKKKQQIRPDRLSSSKQKDSVQKAIQKNAKTLERRLQKIGETTKPQQMKQIRFPVPKSLELQSRYPIMGQNVQLERSGRTLLVNGDFQFSLGKKIAIVGENGSGKTTLLEHIRKQGEGILLSPKVSFQVYQQKGYQMTSEESIIRFVMRQTEFSESLVRSLLNHLGFAQETLTKPLCTLSGGEATRLTIALLFTKPSNVLLLDEPTNFIDMATIEALEKLMQIYPGTILFTSHDSYFVERTADEVYEIKGQKIKKVLTRNF',\n",
       " 'MFMPPVFPAHWHVSQPVLIADTFSSLVWKVFLPDGTPAIVKGLKPIEDIADELRGADYLVWRSGRGAVRLLGRENNLMLLEYAGERMLSHIVAEHGDYQATEIAAELMAKLYAASEEPLPSALLPIRERFAALFQRARDDQNAGRQTDYVHAAIIADQMMSNASELRGLHGDLHHENIMFSSRGWLVIDPVGLVGEVGFGAANMFYDPADRDDLCLDPRRIAQMADAFSRALDVDPRRLLDQAYAYGCLSAAWNADGEEEQRDLAIAAAIKHVRQTSY',\n",
       " 'MFMPPVFPAHWHVSQPVLIADTFSSLVWKVSLPDGTPAIVKGLKPIEDIADELRGADYLVWRNGRGAVRLLGRENNLMLLEYAGERMLSHIVAEHGDYQATEIAAELMAKLYAASEEPLPSALLPIRDRFAALFQRARDDQNAGCQTDYVHAAIIADQMMSNASELRGLHGDLHHENIMFSSRGWLVIDPVGLVGEVGFGAANMFYDPADRDDLCLDPRRIAQMADAFSRALDVDPRRLLDQAYAYGCLSAAWNADGEEEQRDLAIAAAIKQVRQTSY',\n",
       " 'MGIIRTCRLGPDQVKSMRAALDLFGREFGDVATYSQHQPDSDYLGNLLRSKTFIALAAFDQEAVVGALAAYVLPKFEQARSEIYIYDLAVSGEHRRQGIATALINLLKHEANALGAYVIYVQADYGDDPAVALYTKLGIREEVMHFDIDPSTAT',\n",
       " 'MGNYFESPFRGKLLSEQVSNPNIRVGRYSYYSGYYHGHSFDDCARYLMPDRDDVDKLVIGSFCSIGSGAAFIMAGNQGHRAEWASTFPFHFMHEEPAFAGAVNGYQPAGDTLIGHDVWIGTEAMFMPGVRVGHGAIIGSRALVTGDVEPYAIVGGNPARTIRKRFSDGDIQNLLEMAWWDWPLADIEAAMPLLCTGDIPALYRHWKQRQATA',\n",
       " 'MGNYFESPFRGKLLSEQVSNPNIRVGRYSYYSGYYHGHSFDDCARYLMPDRDDVDKLVIGSGAAFIMAGNQGHRAEWASTFPFHFMHEEPAFAGAVNGYQPAGDTLIGHDVWIGTEAMFMPGVRVGHGAIIGSRALVTGDVEPYAIVGGNPARTIRKRFSDGDIQNLLEMAWWDWPLADIEAAMPMLCTGDIPALYRHWKQRQATA',\n",
       " 'MHDAAASLPPQAPAAWADYLSGYRWQAQREGCSAAAVHRLDAARRPTLFVKHEALSEQAELPDEIARLRWLHGAGIDCPQVLNETQGGGRQWLLMSAVAGDTLSALAQRGELEPERLVRLVAAALRRLHDLDPAACPFDHRLERRLESVRRRVEAGLVEETDFDDEHRGRSAVELYRLLLDRRPAVEDLVVTHGDACLPNLLAEGRRFSGFIDCGRLGVADRHQDLALAARDIETDLGAPWAEAFLEAYGGDIDGERLAYFRLLDEFF',\n",
       " 'MHDAATSMPPQAPSTWADYLAGYRWRGQGEGCSAATVHRLEVARRPTLFVKQEVLSAHAELPAEIARLRWLHGAGIDCPQVLNETQSDGRQWLLMSAVPGDTLSALAQRGELEPERLVRLVAAALRRLHDLDPAACPFDHRLERRLDTVRQRVEAGLVDEADFDDDHRGRSATELYRLLLDRRPAVEDLVVAHGDACLPNLLAEGRRFSGFIDCGRLGVADRHQDLALAARDIEAELGAAWAEAFLVEYGGDIDGERLAYFRLLDEFF',\n",
       " 'MHDAATSMPQAPSTWADYLAGYRWRGQGEGCSAATVHRLEAERRPTLFVKQEVLSAHAELPAEIARLRWLHGAGIDCPQVLNETQSDGRQWLLMSAVPGDTLSALAQRGELEPERLVRLVAAALRRLHDLDPAACPFDHRLERRLDTVRQRVEAGLVDEADFDDDHRGRSATELYRLLLDRRPAVEDLVVAHGDACLPNLLAEGRRFSGFIDCGRLGVADRHQDLALAARDIEAELGAAWAEAFLVEYGGDIDGERLAYFRLLDEFF',\n",
       " 'MIEQDGLHAGSPAAWVERLFGYDWAQQTIGCSDAAVFRLSAQGRPVLFVKTDLSGALNELQDEAARLSWLATTGVPCAAVLDVVTEAGRDWLLLGEVPGQDLLSSHLAPAEKVSIMADAMRRLHTLDPATCPFDHQAKHRIERARTRMEAGLVDQDDLDEEHQGLAPAELFARLKARMPDGEDLVVTHGDACLPNIMVENGRFSGFIDCGRLGVADRYQDIALATRDIAEELGGEWADRFLVLYGIAAPDSQRIAFYRLLDEFF',\n",
       " 'MIIRFLALLFSAVVLVSLGHAQEKTHESSNWGKYFSDFNAKGTIVVVDERTNGNSTSVYNESRAQQRYSPASTFKIPHTLFALDAGAVRDEFHVFRWDGAKRSFAGHNQDQNLRSAMRNSTVWVYQLFAKEIGENKARSYLEKLNYGNADPSTKSGDYWIDGNLAISANEQISILKKLYRNELPFRVEHQRLVKDLMIVEAKRDWILRAKTGWDGQMGWWVGWVEWPTGPVFFALNIDTPNRMEDLHKREAIARAILQSVNALPPN',\n",
       " 'MIISEFDRNNLVLKDQLSDLLRLTWPEEYGDSSAEEVEEMMNPERIAVAAVDQDELVGFIGAIPQYGITGWELHPLVVESSRRKNQIGTRLVNYLEKEVASRGGITIYLGTDDLDHGTTLSQTDLYEHTFDKVASIQNLREHPYEFYEKLGYKIVGVLPNANGWDKPDIWMAKTIIPRPDSQ',\n",
       " 'MIISEFDRNNPVLKDQLSDLLRLTWPEEYGDSSAEEVEEMMNPERIAVAAVDQDELVGFIGAIPQYGITGWELHPLVVESSRRKNQIGTRLVNYLEKEVASRGGITIYLGTDDLDHGTTLSQTDLYEHTFDKVASIQNLREHPYEFYEKLGYKIVGVLPNANGWDKPDIWMAKTIIPRPDSQ',\n",
       " 'MIISEFDRNNPVLKDQLSDLLRLTWPEEYGDSSAEEVEEMMNPERIAVAAVDQDELVGFIGAIPQYGITGWELHPLVVESSRRKNQIGTRLVNYLEKEVASRGGITIYLGTDDLDHGTTLSQTDLYEHTFDKVASIQNLREHPYEFYEKLGYKIVGVLPNANGWDKPDIWMAKTIIPRPNSQ',\n",
       " 'MIRMILAINNQCFIGKNNTLMYRLKDDMLNFKKMTQNNIVVMGRKTFESLNNRGLPNRLNVVVTSKAETFEDIQTITTHDMKRSETFTKEGHVVYITPDSFINQFLPFHRDSEDEIWVIGGAQVYEAATPFASEIICTFVDDDEVGDVALKPKLFGGFTHLATLKSVDVDEDNDKPYEITQLVRHEDLEHKLRELQAQQHEMEKEQTQNNLSTPLENGGLRQGEAFVIAATTSAALSQIDTESREDSSDSDSSSSDSSSSSSD',\n",
       " 'MITEMKAGHLKDIDKPSEPFEVIGKIIPRYENENWTFTELLYEAPYLKSYQDEEDEEDEEADCLEYIDNTDKIIYLYYQDDKCVGKVKLRKNWNRYAYIEDIAVCKDFRGQGIGSALINISIEWAKHKNLHGLMLETQDNNLIACKFYHNCGFKIGSVDTMLYANFENNFEKAVFWYLRF',\n",
       " 'MKIINIGVLAHVDAGKTTLTESLLYNSGAITELGSVDKGTTRTDNTLLERQRGITIQTGITSFQWENTKVNIIDTPGHMDFLAEVYRSLSVLDGAILLISAKDGVQAQTRILFHALRKMGIPTIFFINKIDQNGIDLSTVYQDIKEKLSAEIVIKQKVELYPNMCVTNFTESEQWDTVIEGNDDLLEKYMSGKSLEALELEQEESIRFQNCSLFPLYHGSAKSNIGIDNLIEVITNKFYSSTHRGPSELCGNVFKIEYTKKRQRLAYIRLYSGVLHLRDSVRVSEKEKIKVTEMYTSINGELCKIDRAYSGEIVILQNEFLKLNSVLGDTKLLPQRKKIENPHPLLQTTVEPSKPEQREMLLDALLEISDSDPLLRYYVDSTTHEIILSFLGKVQMEVISALLQEKYHVEIELKEPTVIYMERPLKNAEYTIHIEVPPNPFWASIGLSVSPLPLGSGMQYESSVSLGYLNQSFQNAVMEGIRYGCEQGLYGWNVTDCKICFKYGLYYSPVSTPADFRMLTPIVLEQAFRKAGTELLEPYLSFKVYAPQEYLSRAYNDAPKYCANIVNTQLKNNEVIIIGEIPARCIQDYRNDLTFFTNGLSVCLAELKGYQVTTGEPVCQTRRLNSRIDKVRYMFNKIT',\n",
       " 'MKIINIGVLAHVDAGKTTLTESLLYNSGAITELGSVDKGTTRTDNTLLERQRGITIQTGITSFQWENTKVNIIDTPGHMDFLAEVYRSLSVLDGAILLISAKDGVQAQTRILFHALRKMGIPTIFFINKIDQNGIDLSTVYQDIKEKLSAEIVIKQKVELYPNVCVTNFTESEQWDTVIEGNDDLLEKYMSGKSLEALELEQEESIRFQNCSLFPLYHGSAKSNIGIDNLIEVITNKFYSSTHRGPSELCGNVFKIEYTKKRQRLAYIRLYSGVLHLRDSVRVSEKEKIKVTEMYTSINGELCKIDRAYSGEIVILQNEFLKLNSVLGDTKLLPQRKKIENPHPLLQTTVEPSKPEQREMLLDALLEISDSDPLLRYYVDSTTHEIILSFLGKVQMEVISALLQEKYHVEIELKEPTVIYMERPLKNAEYTIHIEVPPNPFWASIGLSVSPLPLGSGMQYESSVSLGYLNQSFQNAVMEGIRYGCEQGLYGWNVTDCKICFKYGLYYSPVSTPADFRMLAPIVLEQVLKKAGTELLEPYLSFKIYAPQEYLSRAYNDAPKYCANIVDTQLKNNEVILSGEIPARCIQEYRSDLTFFTNGRSVCLTELKGYHVTTGEPVCQPRRPNSRIDKVRYMFNKIT',\n",
       " 'MKIINIGVLAHVDAGKTTLTESLLYNSGAITELGSVDKGTTRTDNTLLERQRGITIQTGITSFQWENTKVNIIDTPGHMDFLAEVYRSLSVLDGAILLISAKDGVQAQTRILFHALRKMGIPTIFFINKIDQNGIDLSTVYQDIKEKLSMEIIIKQKVELHPNMCVMSCTEPEQWDVVIEGNDDLLEKYMSGKSLEALELEQEEIRRFQNCSLYPVYHGSAKSNIGIEQLIEVITNKFYSSTYRKKSELCGNVFKIEYSEERQRLAYVRLYGGILHLRDSVRISEKEKIKITEMYTSINGELCKIDKAYSGEIVILQNEFLKLNSVLGDTKLLPQRERIENPLPLLQTTVEPSKPQQREMLLDALLEISDSDPLLQYYVDSTTHEIILSFLGKVQMEVISALLQEKYHVEIELKEPTVIYMERPLKNAEYTIHIEVPPNPFWASIGLSVSPLPLGSGMQYESSVSLGYLNQSFQNAVMEGIRYGCEQGLYGWNVTDCKICFKYGLYYSPVSTPADFRMLAPIVLEQVLKKAGTELLEPYLSFKIYAPQEYLSRAYNDAPKYCANIVDTQLKNNEVILSGEIPARCIQEYRSDLTFFTNGRSVCLTELKGYHVTTGEPVCQPRRPNSRIDKVRYMFNKIT',\n",
       " 'MKIVKRILLVLLSLFFTIVYSNAQTDNLTLKIENVLKAKNARIGVAIFNSNEKDTLKINNDFHFPMQSVMKFPIALAVLSEIDKGNLSFEQKIEITPQDLLPKTWSPIKEEFPNGTTLTIEQILNYTVSESDNIGCDILLKLIGGTDSVQKFLNANHFTDISIKANEEQMHKDWNTQYQNWATPTAMNKLLIDTYNNKNQLLSKKSYDFIWKIMRETTTGSNRLKGQLPKNTIVAHKTGTSGINNGIAAATNDVGVITLPNGQLIFISVFVAESKETSEINEKIISDIAKITWNYYLNK',\n",
       " 'MKKLIFLIVIALVLSACNSNSSHAKELNDLEKKYNAHIGVYALDTKSGKEVKFNSDKRFAYASTSKAINSAILLEQVPYNKLNKKVHINKDDIVAYSPILEKYVGKDITLKELIEASMTYSDNTANNKIIKEIGGIKKVKQRLKELGDKVTNPVRYEIELNYYSPKSKKDTSTPAAFGKTLNKLIANGKLSKKNKNFLLDLMFNNKNGDTLIKDGVPKDYKVADKSGQAITYASRNDVAFVYPKGQSEPIVLVIFTNKDNKSDKPNDKLISETAKSVMKEF',\n",
       " 'MKLLKILSLVCLSISIGACAEHSMSRAKTSTIPQVNNSIIDQNVQALFNEISADAVFVTYDGQNIKKYGTHLDRAKTAYIPASTFKIANALIGLENHKATSTEIFKWDGKPRFFKAWDKDFTLGEAMQASTVPVYQELARRIGPSLMQSELQRIGYGNMQIGTEVDQFWLKGPLTITPIQEVKFVYDLAQGQLPFKPEVQQQVKEMLYVERRGENRLYAKSGWGMAVDPQVGWYVGFVEKADGQVVAFALNMQMKAGDDIALRKQLSLDVLDKLGVFHYL',\n",
       " 'MKLSLMVAISKNGVIGNGPDIPWSAKGEQLLFKAITYNQWLLVGRKTFESMGALPNRKYAVVTRSSFTSDNENVLIFPSIKDALTNLKKITDHVIVSGGGEIYKSLIDQVDTLHISTIDIEPEGDVYFPEIPSNFRPVFTQDFASNINYSYQIWQKG',\n",
       " 'MKPNRPLIVILSTVALDAVGIGLIMPVLPGLLRDLVHSNDVTAHYGILLALYALMQFACAPVLGALSDRFGRRPVLLVSLAGAAVDYAIMATAPFLWVLYIGRIVAGITGATGAVAGAYIADITDGDERARHFGFMSACFGFGMVAGPVLGGLMGGFSPHAPFFAAAALNGLNFLTGCFLLPESHKGERRPLRREALNPLASFRWARGMTVVAALMAVFFIMQLVGQVPAALWVIFGEDRFHWDATTIGISLAAFGILHSLAQAMITGPVAARLGERRALMLGMIADGTGYILLAFATRGWMAFPIMVLLASGGIGMPALQAMLSRQVDEERQGQLQGSLAALTSLTSIVGPLLFTAIYAASITTWNGWAWIAGAALYLLCLPALRRGLWSGAGQRADR',\n",
       " 'MKTFAAYVIIACLSSTALAGSITENTSWNKEFSAEAVNGVFVLCKSSSKSCATNDLARASKEYLPASTFKIPNAIIGLETGVIKNEHQVFKWDGKPRAMKQWERDLTLRGAIQVSAVPVFQQIAREVGEVRMQKYLKKFSYGNQNISGGIDKFWLEGQLRISAVNQVEFLESLYLNKLSASKENQLIVKEALVTEAAPEYLVHSKTGFSGVGTESNPGVAWWVGWVEKETEVYFFAFNMDIDNESKLPLRKSIPTKIMESEGIIGG',\n",
       " 'MKVSLMAAKAKNGVIGCGPDIPWSAKGEQLLFKALTYNQWLLVGRKTFESMGALPNRKYAVVTRSGWTSNDDNVVVFQSIEEAMDRLAEFTGHVIVSGGGEIYRETLPMASTLHLSTIDIEPEGDVFFPSIPNTFEVVFEQHFTSNINYCYQIWKKG',\n",
       " 'MLKQKELIANVKNLTESDERITACMMYGSFTKGEGDQYSDIEFYIFLKDSITSNFDSSNWLFDVAPYLMLYKNEYGTEVVIFDNLIRGEFHFLSEKDMNIIPSFKDSGYIPDTKAMLIYDETGQLENYLSEISGARPNRLTEENANFLLCNFSNLWLMGINVLKRGEYARSLELLSQLQKNTLQLIRMAEKNADNWLNMSKNLEKEISLENYKKFAKTTARLDKVELFEAYKNSLLLVMDLQSHLIEQYNLKVTHDILERLLNYISE',\n",
       " 'MLKSINHICFSVRNLNDSIHFYRDILLGKLLLTGKKTAYFELAGLWIALNEEKDIPRNEIHFSYTHIAFTIDDSEFKYWHQRLKDNNVNILEGRVRDIRDRQSIYFTDPDGHKLELHTGTLENRLNYYKEAKPHMTFYK',\n",
       " 'MLQSLNHLTLAVSDLQKSVTFWHELLGLALHARWNTGAYLTCGDLWVCLSYDEARQYVPPQESDYTHYAFTVAEEDFEPFSHRLEQAGVIVWKQNKSEGASFYFLDPDGHKLELHVGSLAARLAACREKPYAGMVFTSDGA',\n",
       " 'MLQSLNHLTLAVSDLQKSVTFWHELLGLALHARWNTGAYLTCGDLWVCLSYDEARQYVPPQESDYTHYAFTVAEEDFEPFSQRLEQAGVTVWKQNKSEGASFYFLDPDGHKLELHVGSLAARLAACREKPYAGMVFTSDEA',\n",
       " 'MLQSLNHLTLAVSDLQKSVTFWHELLGLALHARWNTGAYLTCGDLWVCLSYDEARQYVPPQESDYTHYAFTVSEEDFEPFSQRLEQAGVTVWKQNKSEGASFYFLDPDGHKLELHVGSLAARLAACREKPYAGMVFTSDEA',\n",
       " 'MLQSLNHLTLAVSDLQKSVTFWHELLGLALHARWNTGAYLTCGDLWVCLSYDEARRNVPPQESDYTHYAFTVAEEDFEPFSHRLEQAGVTVWKQNKSEGESFYFLDPDGHKLELHVGSLAARLAACREKPYAGMVFTSDGA',\n",
       " 'MLRSRVTVFGILNLTEDSFFDESRRLDPAGAVTAAIEMLRVGSDVVDVGPAASHPDARPVSPADEIRRIAPLLDALSDQMHRVSIDSFQPETQRYALKRGVGYLNDIQGFPDPALYPDIAEADCRLVVMHSAQRDGIATRTGHLRPEDALDEIVRFFEARVSALRRSGVAADRLILDPGMGFFLSPAPETSLHVLSNLQKLKSALGLPLLVSVSRKSFLGATVGLPVKDLGPASLAAELHAIGNGADYVRTHAPGDLRSAITFSETLAKFRSRDARDRGLDHA',\n",
       " 'MLSGLNHLTLAVSQLAPSVAFYQQLLGMTLHARWDSGAYLSCGDLWLCLSLDPQRRVTPPEESDYTHYAFSISEADFASFAARLEAAGVAIWKLNRSEGASHYFLDPDGHKLELHVGSLAQRLAACREQPYKGMVFFDQ',\n",
       " 'MLSSFLMLSIISSLLTICVIFLVRMLYIKYTQNIMSHKIWLLVLVSTLIPLIPFYKISNFTFSKDMMNRNVSDTTSSVSHMLDGQQSSVTKDLAINVNQFETSNITYMILLIWVFGSLLCLFYMIKAFRQIDVIKSSSLESSYLNERLKVCQSKMQFYKKHITISYSSNIDNPMVFGLVKSQIVLPTVVVETMNDKEIEYIILHELSHVKSHDLIFNQLYVVFKMIFWFNPALYISKTMMDNDCEKVCDRNVLKILNRHEHIRYGESILKCSILKSQHINNVAAQYLLGFNSNIKERVKYIALYDSMPKPNRNKRIVAYIVCSISLLIQAPLLSAHVQQDKYETNVSYKKLNQLAPYFKGFDGSFVLYNEREQAYSIYNEPESKQRYSPNSTYKIYLALMAFDQNLLSLNHTEQQWDKHQYPFKEWNQDQNLNSSMKYSVNWYYENLNKHLRQDEVKSYLDLIEYGNEEISGNENYWNESSLKISAIEQVNLLKNMKQHNMHFDNKAIEKVENSMTLKQKDTYKYVGKTGTGIVNHKEANGWFVGYVETKDNTYYFATHLKGEDNANGEKAQQISERILKEMELI',\n",
       " 'MLTGLNHLTLAVADLPASIAFYRDLLGFRLEARWDQGAYLELGSLWLCLSREPQYGGPAADYTHYAFGIAAADFARFAAQLRAHGVREWKQNRSEGDSFYFLDPDGHRLEAHVGDLRSRLAACRQAPYAGMRFAD',\n",
       " 'MLVEQGRLAIRALQKSDAPVMLHWLQDERVLEFYEGRDKRFDLQTVIEVFIEDQGETTPCLVLLDDKPLGYVQFYPLDSEDKQALELPVEDVIYGLDQFIGEPDFWGLGLGTELVSLVRDYLITDKAAQRLVLDPQSRNLRAIACYEKCGFEKLCLLPAHEMHEGQLQDCWLMQYYPARSSLIASSRPKI',\n",
       " 'MMKKIKIVPLILIVVVVGFGIYFYASKDKEINNTIDAIEDKNFKQVYKDSSYISKSDNGEVEMTERPIKIYNSLGVKDINIQDRKIKKVSKNKKRVDAQYKIKTNYGNIDRNVQFNFVKEDGMWKLDWDHSVIIPGMQKDQSIHIENLKSERGKILDRNNVELANTGTAYEIGIVPKNVSKKDYKAIAKELSISEDYIKQQMDQNWVQDDTFVPLKTVKKMDEYLSDFAKKFHLTTNETESRNYPLGKATSHLLGYVGPINSEELKQKEYKGYKDDAVIGKKGLEKLYDKKLQHEDGYRVTIVDDNSNTIAHTLIEKKKKDGKDIQLTIDAKVQKSIYNNMKNDYGSGTAIHPQTGELLALVSTPSYDVYPFMYGMSNEEYNKLTEDKKEPLLNKFQITTSPGSTQKILTAMIGLNNKTLDDKTSYKIDGKGWQKDKSWGGYNVTRYEVVNGNIDLKQAIESSDNIFFARVALELGSKKFEKGMKKLGVGEDIPSDYPFYNAQISNKNLDNEILLADSGYGQGEILINPVQILSIYSALENNGNINAPHLLKDTKNKVWKKNIISKENINLLTDGMQQVVNKTHKEDIYRSYANLIGKSGTAELKMKQGETGRQIGWFISYDKDNPNMMMAINVKDVQDKGMASYNAKISGKVYDELYENGNKKYDIDE',\n",
       " 'MMKKSLCCALLLGISCSALATPVSEKQLAEVVANTVTPLMKAQSVPGMAVAVIYQGKPHYYTFGKADIAANKPVTPQTLFELGSISKTFTGVLGGDAIARGEISLDDPVTRYWPQLTGKQWQGIRMLDLATYTAGGLPLQVPDEVTDNASLLRFYQNWQPQWKPGTTRLYANASIGLFGALAVKPSGMPYEQAMTTRVLKPLKLDHTWINVPKAEEAHYAWGYRDGKAVRVSPGMLDAQAYGVKTNVQDMANWVMANMAPENVADASLKQGIALAQSRYWRIGSMYQGLGWEMLNWPVEANTVVEGSDSKVALAPLPVAEVNPPAPPVKASWVHKTGSTGGFGSYVAFIPEKQIGIVMLANTSYPNPARVEAAYHILEALQ',\n",
       " 'MMKKSLSCALLLSVACSAFAAPMSEKQLADVVERTVTPLMKAQAIPGMAVAVIYQGQPHYFTFGKADVAANKPVTPQTLFELGSVSKTFTGVLGGDAIARKEISLADPVTKYWPELTGKQWQGIRLLDLATYTAGGLPLQVPDDVTDNASLLRFYQSWQPKWAPGTTRLYANTSIGLFGSLAVKPSGMRFEQAMAERVFKPLKLNHTWINVPHAEEPHYAWGYREGKAVHVSPGMLDAEAYGVKSNVKDMASWVMANMAPETLPQSTLQQGIALAQSRYWRVGAMYQGLGWEMLNWPVDAKTVVDGSDNKVALAPLPVAEVNPPAPPVKASWVHKTGSTGGFGSYVAFIPEKQIGIVMLANKSYPNPVRVETAYRILDALQ',\n",
       " 'MMKKSLSCALLLSVACSAFAAPMSEKQLADVVERTVTPLMKAQAIPGMAVAVIYQGQPHYFTFGKADVAANKPVTPQTLFELGSVSKTFTGVLGGDAIARKEISLADPVTKYWPELTGKQWQGIRLLDLATYTAGGLPLQVPDDVTDNASLLRFYQSWQPKWAPGTTRLYANTSIGLFGSLAVKPSGMRFEQAMAERVFKPLKLNHTWINVPHAEEPHYAWGYREGKAVHVSPGMLDAEAYGVKSNVKDMASWVMANMAPETLPQSTLQQGIALAQSRYWRVSAMYQGLGWEMLNWPVDAKTVVDGSDNKVALAPLPVAEVNPPAPPVKASWVHKTGSTGGFGSYVAFIPEKQIGIVMLANKSYPNPVRVETAYRILETLQ',\n",
       " 'MMKKSLSCALLLSVACSAFAAPMSEKQLANVVERNVTPLMKAQAIPGMAVAVIYQGQPHYFTFGKADVAANKPVTPQTLFELGSVSKTFTGVLGGDAIARKEISLADPVTKYWPELTGKQWQGIRLLDLATYTAGGLPLQVPDAVTDNASLLRFYQSWQPKWAPGTTRLYANTSIGLFGSLAVKPSGMRFEEAMTERVFKPLKLNHTWINIPHAEEPHYAWGYREGKAVHVSPGLLDAEAYGVKSNVKDMASWVMANMAPETIQQPSLKQGIALAQSRYWRVGAMYQGLGWEMLNWPVDAKTVVNGSDNKVALAPLPAAEVNPPAPPVKASWVHKTGSTGGFGSYVAFIPEKQIGIVMLANKSYPNPLRVETAYRILDTLQ',\n",
       " 'MNIKALLLITSAIFISACSPYIVTANPNHSASKSDDKAEKIKNLFNEAHTTGVLVIHQGQTQQSYGNDLARASTEYVPASTFKMLNALIGLEHHKATTTEVFKWDGEKRLFPEWEKNMTLGDAMKASAIPVYQDLARRIGLELMSKEVKRVGYGNADIGTQVDNFWLVGPLKITPQQEAQFAYKLANKTLPFSQKVQDEVQSMLFIEEKNGNKIYAKSGWGWDVNPQVGWLTGWVVQPQGNIVAFSLNLEMKKGIPSSVRKEITYKSLEQLGIL',\n",
       " 'MNIKALLLITSAIFISACSPYIVTANPNHSASKSDVKAEKIKNLFNEAHTTGVLVIQQGQTQQSYGNDLARASTEYVPASTFKMLNALIGLEHHKATTTEVFKWDGKKRLFPEWEKDMTLGDAMKASAIPVYQDLARRIGLELMSKEVKRVGYGNADIGTQVDNFWLVGPLKITPQQEAQFAYKLANKTLPFSQKVQDEVQSMLFIEEKNGNKIYAKSGWGWDVNPQVGWLTGWVVQPQGNIVAFSLNLEMKKGIPSSVRKEITYKSLEQLGIL',\n",
       " 'MNISLIFANELITRAFGNQGKLPWQFIKEDMQFFQKTTENSVVVMGLNTWRSLPKMKKLGRDFIVISSTITEHEVLNNNIQIFKSFESFLEAFRDTTKPINVIGGVGLLSEAIEHASTVYMSSIHMVKPVHADVYVPVELMNKLYSDFKYPENILWVGDPIDSVYSLSIDKFVRPASLVGVPNDINT',\n",
       " 'MNIVENEICIRTLIDDDFPLMLKWLTDERVLEFYGGRDKKYTLESLKKHYTEPWEDEVFRVIIEYNNVPIGYGQIYKMYDELYTDYHYPKTDEIVYGMDQFIGEPNYWSKGIGTRYIKLIFEFLKKERNANAVILDPHKNNPRAIRAYQKSGFRIIEDLPEHELHEGKKEDCYLMEYRYDDNATNVKAMKYLIEHYFDNFKVDSIEIIGSGYDSVAYLVNNEYIFKTKFSTNKKKGYAKEKAIYNFLNTNLETNVKIPNIEYSYISDELSILGYKEIKGTFLTPEIYSTMSEEEQNLLKRDIASFLRQMHGLDYTDISECTIDNKQNVLEEYILLRETIYNDLTDIEKDYIESFMERLNATTVFEGKKCLCHNDFSCNHLLLDGNNRLTGIIDFGDSGIIDEYCDFIYLLEDSEEEIGTNFGEDILRMYGNIDIEKAKEYQDIVEEYYPIETIVYGIKNIKQEFIENGRKEIYKRTYKD',\n",
       " 'MNKNIKYSQNFLTSEKVLNQIIKQLNLKETDTVYEIGTGKGHLTTKLAKISKQVTSIELDSHLFNLSSEKLKLNIRVTLIHQDILQFQFPNKQRYKIVGSIPYHLSTQIIKKVVFESHASDIYLIVEEGFYKRTLDIHRTLGLLLHTQVSIQQLLKLPAECFHPKPKVNSVLIKLTRHTTDVPDKYWKLYTYFVSKWVNREYRQLFTKNQFHQAMKHAKVNNLSTVTYEQVLSIFNSYLLFNGRK',\n",
       " 'MNKNIKYSQNFLTSEKVLNQIIKQLNLKETDTVYEIGTGKGHLTTKLAKISKQVTSIELDSHLFNLSSEKLKLNTRVTLIHQDILQFQFPNKQRYKIVGNIPYHLSTQIIKKVVFESRASDIYLIVEEGFYKRTLDIHRTLGLLLHTQVSIQQLLKLPAECFHPKPKVNSVLIKLTRHTTDVPDKYWKLYTYFVSKWVNREYRQLFTKNQFHQAMKHAKVNNLSTVTYEQVLSIFNSYLLFNGRK',\n",
       " 'MNKPAVIALVITLLDAMGIGLIMPVLPSLLREYLPEADVANHYGILLALYAVMQVCFAPLLGRWSDKLGRRPVLLLSLAGAAFDYTLLALSNVLWMLYLGRIISGITGATGAVAASVVADSTAVSERTAWFGRLGAAFGAGLIAGPAIGGLAGDISPHLPFVIAAILNACTFLMVFFIFKPAVQTEEKPAEQKQESAGISFITLLKPLALLLFVFFTAQLIGQIPATVWVLFTESRFAWDSAAVGFSLAGLGAMHALFQAVVAGALAKRLSEKTIIFAGFIADATAFLLMSAITSGWMVYPVLILLAGGGIALPALQGIISAGASAANQGKLQGVLVSLTNLTGVAGPLLFAFIFSQTQQSADGTVWLIGTALYGLLLAICLLIRKPAPVAATC',\n",
       " 'MNKSLIIFGIVNITSDSFSDGGRYLAPDAAIAQARKLMAEGADVIDLGPASSNPDAAPVSSDTEIARIAPVLDALKADGIPVSLDSYQPATQAYALSRGVAYLNDIRGFPDAAFYPQLAKSSAKLVVMHSVQDGQADRREAPAGDIMDHIAAFFDARIAALTGAGIKRNRLVLDPGMGFFLGAAPETSLSVLARFDELRLRFDLPVLLSVSRKSFLRALTGRGPGDVGAATLAAELAAAAGGADFIRTHEPRPLRDGLAVLAALKETARIR',\n",
       " 'MNKYFTCYVVASLFLSGCTVQHNLINETPSQIVQGHNQVIHQYFDEKNTSGVLVIQTDKKINLYGNALSRANTEYVPASTFKMLNALIGLENQKTDINEIFKWKGEKRSFTAWEKDMTLGEAMKLSAVPVYQELARRIGLDLMQKEVKRIGFGNAEIGQQVDNFWLVGPLKVTPIQEVEFVSQLAHTQLPFSEKVQANVKNMLLLEESNGYKIFGKTGWAMDIKPQVGWLTGWVEQPDGKIVAFALNMEMRSEMPASIRNELLMKSLKQLNII',\n",
       " 'MNQKNPKDTQNFITSKKHVKEILNHTNISKQDNVIEIGSGKGHFTKELVKMSRSVTAIEIDGGLCQVTKEAVNPSENIKVIQTDILKFSFPKHINYKIYGNIPYNISTDIVKRITFESQAKYSYLIVEKGFAKRLQNLQRALGLLLMVEMDIKMLKKVPPLYFHPKPSVDSVLIVLERHQPLISKKDYKKYRSFVYKWVNREYRVLFTKNQFRQALKHANVTNINKLSKEQFLSIFNSYKLFH',\n",
       " 'MNRTNIFFGESHSDWLPVRGGESGDFVFRRGDGHAFAKIAPASRRGELAGERDRLIWLKGRGVACPEVINWQEEQEGACLVITAIPGVPAADLSGADLLKAWPSMGQQLGAVHSLSVDQCPFERRLSRMFGRAVDVVSRNAVNPDFLPDEDKSTPQLDLLARVERELPVRLDQERTDMVVCHGDPCMPNFMVDPKTLQCTGLIDLGRLGTADRYADLALMIANAEENWAAPDEAERAFAVLFNVLGIEAPDRERLAFYLRLDPLTWG',\n",
       " 'MNRTYIFLGESHSDWLPVRGGESGDFVFRRGDGHAFAKIAPASRRGELAGERDRLIWLKGRGVACPEVINWQEEQEGACLVITAIPGVPAADLSGADLLKAWPSMGQQLGAVHSLSVDQCPFERRLSRMIGRAVDVVSRNAVNPDFLPDEDKSTPQLDLLARVERELPVRLDQERTDMVVCHGDPCMPNFMVDPRTLQCTGLIDLGRLGTADRYADLALMIANAEENWAAPDEAERAFAVLFNVLGIEAPDRERLAFYLRLDPLTWG',\n",
       " 'MNTSYSQSNLRHNQILIWLCILSFFSVLNEMVLNVSLPDIANDFNKPPASTNWVNTAFMLTFSIGTAVYGKLSDQLGIKRLLLFGIIINCFGSLIGFVGHSFFSLLIMARFIQGAGAAAFPALVMVVVARYIPKENRGKAFGLIGSIVAMGEGVGPAIGGMIAHYIHWSYLLLIPMITIITVPFLMKLLKKEVRIKGHFDIKGIILMSVGIVFFMLFTTSYSISFLIVSVLSFLIFVKHIRKVTDPFVDPGLGKNIPFMIGVLCGGIIFGTVAGFVSMVPYMMKDVHQLSTAEIGSVIIFPGTMSVIIFGYIGGILVDRRGPLYVLNIGVTFLSVSFLTASFLLETTSWFMTIIIVFVLGGLSFTKTVISTIVSSSLKQQEAGAGMSLLNFTSFLSEGTGIAIVGGLLSIPLLDQRLLPMEVDQSTYLYSNLLLLFSGIIVISWLVTLNVYKHSQRDF',\n",
       " 'MNTSYSQSNLRHNQILIWLCILSFFSVLNEMVLNVSLPDIANDFNKPPASTNWVNTAFMLTFSIGTAVYGKLSDQLGIKRLLLFGIIINCFGSVIGFVGHSFFSLLIMARFIQGAGAAAFPALVMVVVARYIPKENRGKAFGLIGSIVAMGEGVGPAIGGMIAHYIHWSYLLLIPMITIITVPFLMKLLKKEVRIKGHFDIKGIILMSVGIVFFMLFTTSYSISFLIVSVLSFLIFVKHIRKVTDPFVDPGLGKNIPFMIGVLCGGIIFGTVAGFVSMVPYMMKDVHQLSTAEIGSVIIFPGTMSVIIFGYIGGILVDRRGPLYVLNIGVTFLSVSFLTASFLLETTSWFMTIIIVFVLGGLSFTKTVISTIVSSSLKQQEAGAGMSLLNFTSFLSEGTGIAIVGGLLSIPLLDQRLLPMEVDQSTYLYSNLLLLFSGIIVISWLVTLNVYKHSQRDF',\n",
       " 'MNVEYSKIKKAVPILLFLFVFSLVIDNSFKLISVAIADDLNISVTTVSWQATLAGLVIGIGAVVYASLSDAISIRTLFIYGVILIIIGSIIGYIFQHQFALLLVGRIIQTAGLAAAETLYVIYVAKYLSKEDQKTYLGLSTSSYSLSLVIGTLSGGFISTYLHWTNMFLIALIVVFTLPFLFKLLPKENNTNKAHLDFVGLILVATIATTVMLFITNFNWLYMIGALIAIIVFALYIKNAQRPLVNKSFFQNKRYASFLFIVFVMYAIQLGYIFTFPFIMEQIYHLQLDTTSLLLVPGYIVAVIVGALSGKIGEYLNSKQAIITAIILIALSLILPAFAVGNHISIFVISMIFFAGSFALMYAPLLNEAIKTIDLNMTGVAIGFYNLIINVAVSVGIAIAAALIDFKALNFPGNDALSSHFGIILIILGLMSIVGLVLFVILNRWTQSEK',\n",
       " 'MNVEYSKIKKAVPILLFLFVFSLVIDNSFKLISVAIADDLNISVTTVSWQATLAGLVIGIGAVVYASLSDAISIRTLFIYGVILIIIGSIIGYIFQHQFPLLLVGRIIQTAGLAAAETLYVIYVAKYLSKEDQKTYLGLSTSSYSLSLVIGTLSGGFISTYLHWTNMFLIALIVVFTLPFLFKLLPKENNTNKAHLDFVGLILVATIATTVMLFITNFNWLYMIGALIAIIVFALYIKNAQRPLVNKSFFQNKRYASFLFIVFVMYAIQLGYIFTFPFIMEQIYHLQLDTTSLLLVPGYIVAVIVGALSGKIGEYLNSKQAIITAIILIALSLILPAFAVGNHISIFVISMIFFAGSFALMYAPLLNEAIKTIDLNMTGVAIGFYNLIINVAVSVGIAIAAALIDFKALNFPGNDALSSHFGIILIILGLMSIVGLVLFVILNRWTQSEK',\n",
       " 'MNVEYSKIKKAVPILLFLFVFSLVIDNSFKLISVAIADDLNISVTTVSWQATLAGLVIGIGAVVYASLSDAISIRTLFIYGVILIIIGSIIGYIFQHQFPLLLVGRIIQTAGLAAAETLYVIYVAKYLSKEDQKTYLGLSTSSYSLSLVIGTLSGGFISTYLHWTNMFLIALIVVFTLPFLFKLLPKENNTNKAHLDFVGLILVATIATTVMLFITNFNWLYMIGALIAIIVFALYIKNAQRPLVNKSFFQNKRYASFLFIVFVMYAIQLGYIFTFPFIMEQIYHLQLDTTSLLLVPGYIVAVIVGALSGKIGEYLNSKQAIITAIILIALSLILPAFAVGNHISIFVISMIFFAGSFALMYAPLLNEAIKTIDLNMTGVAIGFYNLIINVAVSVGIAIAAALIDFKALNFPGNDALSSHFGIILIILGLMSIVGLVLFVSLNRWTQSEK',\n",
       " 'MPFALYMLALAVFVMGTSEFMLAGLLPAIATELDVSVGTAGLLTSAFAVGMVVGAPVMAAFARRWPPRLTLIVCLLVFAGSHVIGAMTPVFSLLLITRVLSALANAGFLAVALSTATTLVPANQKGRALSILLSGTTIATVVGVPAGALLGTALGWRTTFWAIAILCIPAAVGVIRGVTNNVGRSETSATSPRLRVELSQLATPRLILAMALGALINGGTFAAFTFLAPIVTETAGLAEAWVSVALVMFGIGSFLGVTIAGRLSDQRPGLVLAVGGPLLLTGWIVLAVVASHPVALIVLVLVQGFLSFGVGSTLITRVLYAASGAPTMGGSYATAALNIGAAAGPVLGALGLATGLGLLAPVWVASVLTAIALVIMLLTRRALTKTAAEAN',\n",
       " 'MQFKKISCLLLSPLFIFSTSIYADNTPKDQEIKKLVDQNFKPLLEKYDVPGMAVGVIQNNKKYEMYYGLQSVQDKKAVNSSTIFELGSVSKLFTATAGGYAKNKGKISFDDTPGKYWKELKNTPIDQVNLLQLATYTSGNLALQFPDEVQTDQQVLTFFKDWKPKNPIGEYRQYSNPSIGLFGKVVALSMNKPFDQVLEKTIFPALGLKHSYVNVPKTQMQNYAFGYNQENQPIRVNPGPLDAPAYGVPAYGVKSTLPDMLSFIHANLNPQKYPADIQRAINETHQGFYQVNTMYQALGWEEFSYPATLQTLLDSNSEQIVMKPNKVTAISKEPSVKMYHKTGSTNGFGTYVVFIPKENIGLVMLTNKRIPNEERIKAAYAVLNAIKK',\n",
       " 'MQNLAINITNLQVRFGNQLELSIDSLRVYQQDRIGIIGENGVGKSTLLKLIAGELIPDQGKIQTEITFNYLPQLTYLTETKDLNLELASHFQLKLEETSERKWSGGEERKIELIRLLSSYEQGMLLDEPTTHLDRKSMDRLIEELRYYYGTLIFVSHDRYFLDELATKIWEVKDGEIREFSGNYSAYLTQKELEKKTQLREAESIMQEKKRLEKSLQEKKKQAEKLEKVSSKKKKQQIRPDRLSSSKQKDSVQKAIQKNTKTLERRLQKIGETTKPQQMKQIRFPVPKSLALHNRYPIMGQNIQLERNGRTLLINSDFQFSLGKKIAIVGENGSGKTTLLEHIRKQGEGILLSPKVSFQVYQQKDYQMTSEESVIRFVMRQTEFSESLVRSLLNHLGFAQETLAKPLCTLSGGEATRLTIALLFTKPSNVLLLDEPTNFIDMATIEALEKLMQVYPGTILFTSHDSYFVERTADEVYEIKGQKIKKVLTRNF',\n",
       " 'MQQRRAFALLTLGSLLLAPCTYASGEAPLTAAVDGIIQPMLKAYRIPGMAVAVLKDGKAHYFNYGVANRESGQRVSEQTLFEIGSVSKTLTATLGAYAAVKGGFELDDKVSQHAPWLKGSALDGVTMAELATYSAGGLPLQFPDEVDSNDKMRTYYRSWSPVYPAGTHRQYSNPSIGLFGHLAANSLGQPFEQLMSQTLLPKLGLHHTYIQVPESAMVNYAYGYSKEDKPVRVTPGVLAAEAYGIKTGSADLLKFAEANMGYQGDAAVKSAIALTHTGFYSVGDMTQGLGWESYAYPVTEQTLLAGNAPAVSFQANPVTRFAVPKAMGEQRLYNKTGSTGGFGAYVAFVPARGIAIVMLANRNYPIEARVKAAHAILSQLAE',\n",
       " 'MRDTRFPCLCGIAASTLLFATTPAIADEAPADRLKALVDAAVQPVMKANDIPGLTVAISLKGEPHYFSYGLASKEDGRRVTPETLFEIGSVSKTFTATLAGYALAQDKMRLDDRASQHWPALQGSRFDGISLLDLATYTAGGLPLQFPDSVQKDRAQIRDYYRQWQPTYAPGSQRLYSNPSIGLFGYLAARSLGQPFERLMEQQLFPALGLEQTHLDVPEAALAQYAQGYGKDDRPLRVGPGPLDAEGYGVKTSAADLLRFVDANLHPERLDRPWAQALDATHRGYYKVGDMTQGLGWEAYDWPISLKRLQAGNSTPMALQPHRIARLPAPQALEGQRLLNKTGSTNGFGAYVAFVPGRDLGLVILANRNYPNAERVKIAYAILSGLEQQAKVPLKR',\n",
       " 'MRDTRFPCLCGIAASTLLFATTPAIAGEAPADRLKALVDAAVQPVMKANDIPGLAVAISLKGEPHYFSYGLASKEDGRQVTPETLFEIGSVSKTFTATLAGYALAQDKMRLDDRASQHWPALQGSRFDGISLLDLATYTAGGLPLQFPDSVQKDQAQIRDYYRQWQPTYAPGSQRLYSNPSIGLFGYLAARSLGQPFERLMEQQVFPALGLEQTHLDVPEAALAQYAQGYGKDDRPLRVGPGPLDAEGYGVKTSAADLLRFVDANLHPERLDRPWAQALDATHRGYYKVGDMTQGLGWEAYDWPISLKRLQAGNSTPMALQPHRIARLPAPQALEGQRLLNKTGSTNGFGAYVAFVPGRDLGLVILANRNYPNAERVKIAYAILSGLEQQGKVPLKR',\n",
       " 'MRDTRFPCLCGIAASTLLFATTPAIAGEAPADRLKALVDAAVQPVMKANDIPGLAVAISLKGEPHYFSYGLASKEDGRRVTPETLFEIGSVSKTFTATLAGYALAQDKMRLDDRASQHWPALQGSRFDGISLLDLATYTAGGLPLQFPDSVQKDQAQIRDYYRQWQPTYAPGSQRLYSNPSIGLFGYLAARSLGQPFERLMEQQVFPALGLEQTHLDVPEAALAQYAQGYGKDDRPLRVGPGPLDAEGYGVKTSAADLLRFVDANLHPERLDRPWAQALDATHRGYYKVGDMTQGLGWEAYDWPISLKRLQAGNSTPMALQPHRIARLPAPQALEGQRLLNKTGSTNGFGAYVAFVPGRDLGLVILANRNYPNAERVKIAYAILSGLEQQGKVPLKR',\n",
       " 'MREAVIAEVSTQLSEVVGVIERHLEPTLLAVHLYGSAVDGGLKPHSDIDLLVTVTVRLDETTRRALINDLLETSASPGESEILRAVEVTIVVHDDIIPWRYPAKRELQFGEWQRNDILAGIFEPATIDIDLAILLTKAREHSVALVGPAAEELFDPVPEQDLFEALNETLTLWNSPPDWAGDERNVVLTLSRIWYSAVTGKIAPKDVAADWAMERLPAQYQPVILEARQAYLGQEDRLASRADQLEEFVHYVKGEITKVVGK',\n",
       " 'MRFKKISCLLLSPLFIFSTSIYAGNTPKDQEIKKLVDQNFKPLLEKYDVPGMAVGVIQNNKKYEMYYGLQSVQDKKAVNRSTIFELGSVSKLFTATAGGYAKNKGKISFDDTPGKYWKELKNTPIDQVNLLQLATYTSGNLALQFPDEVQTDQQVLTFFKDWQPKNPIGEYRQYSNPSIGLFGKVVALSMNKPFDQVLEKTIFPALGLKHSYVNVPKTQMQNYAFGYNQENQPIRVNPGPLDAPAYGVKSTLPDMLSFIHANLNPQKYPADIQRAINETHQGFYQVNTMYQALGWEEFSYPATLQTLLDSNSEQIVMKPNKVTAISKEPSVKMYHKTGSTNGFGTYVVFIPKENIGLVMLTNKRIPNEERIKAAYAVLDAIKK',\n",
       " 'MRHATILNLCGLAASTLFFATTSAFATEAPAERLKALVDAAVQPVMKANDIPGLAVAITLKGEPHYFSYGVASKEDARKVTPETLFEIGSVSKTFTATLAGYALAQDKMRLDDRASRHWPALQGSRFDGISLLDLGTYSAGGLPLQFPDAVQKDPAQIRDYYRQWQPTYAPGSHRQYSNPSIGLFGYLAARSLGQPFERSMERQLFPALGLEHTFIRVPAAQQGLYAQGYGKDDRPLRVGPGPLDAEAYGLKSSAADLLRFVEANLHPERLEKPWAQALDATHRGYYKVGDMTQGLGWEAYDWPIDLKRLQAGNSAPMALQAHKVARLPAPQALDGQRLLNKTGSTNGFGAYLAFIPGRDVGLVILANRNYPNAERVKIAYAILSGLEQQAKVPLVR',\n",
       " 'MRIVNGPIIMTREERMKIVHEIKERILDKYGDDVKAIGVYGSLGRQTDGPYSDIEMMCVMSTEEAEFSHEWTTGEWKVEVNFDSEEILLDYASQVESDWPLTHGQFFSILPIYDSGGYLEKVYQTAKSVEAQTFHDAICALIVEELFEYAGKWRNIRVQGPTTFLPSLTVQVAMAGAMLIGLHHRICYTTSASVLTEAVKQSDLPSGYDHLCQFVMSGQLSDSEKLLESLENFWNGIQEWTERHGYIVDVSKRIPF',\n",
       " 'MRLLFFSALLLLFGQAQASEWNDSQAVGKLFEAAGMKGTFVLYDVQRQGYDGHDRARAQTRFVPASTYKVAHSLIGLSTGAVKSADEVLPYGGKPQRFKAWEHDMSLREAIRASNVPVYQELARRIGLQRMRANLARMDYGNGEVGQVVDNFWLVGPLEISALEQTRFLARLAQGELPFPEQVQATVRDMTLMEHGPDWELHGKTGWCFDCTPELGWWVGWVKRDERLYTFALNIDMPGGEADIGKRLELGKAALKALGVLP',\n",
       " 'MRMLQSIPALPVGDIKKSIGFYCDKLGFTLVHHEDGFAVLMCNEVRIHLWEASDEGWRSRSNDSPVCTGAESFIAGTASCRIEVEGIDELYQHIKPLGILHPNTSLKDQWWDERDFAVIDPDNNLISFFQQIKS',\n",
       " 'MRPLLFSALLLLSAHAQASEWNDSQAVDKLFGAAGVKGTFVLYDVQRQLYVGHDRERAETRFVPASTYKVANSLIGLSTGAVRSADEVLPYGGKPQRFKAWEHDMSLRDAIKASNVPVYQELARRIGLERMRANVSRLGYGNAEIGQVVDNFWLVGPLKISAMEQTRFLLRLAQGELPFPAPVQSTVRAMTLLESGPGWELHGKTGWCFDCTPELGWWVGWVKRNERLYGFALNIDMPGGEADIGKRVELGKASLKALGILP',\n",
       " 'MRPLLFSALLLLSGHAQASEWNDSRAVDKLFGAAGVKGTFVLYDVQRQRYVGHDRERAETRFVPASTYKVANSLIGLSTGAVRSADEVLPYGGKPQRFKAWEHDMSLRDAIKASNVPVYQELARRIGLERMRANVSRLGYGNAEIGQVVDNFWLVGPLKISAMEQTRFLLRLAQGELPFPAPVQSTVRAMTLLESGPGWELHGKTGWCFDCTPELGWWVGWVKRNERLYGFALNIDMPGGEADIGKRVELGKASLKALGILP',\n",
       " 'MRPLLFSALLLLSGHTQASEWNDSQAVDKLFGAAGVKGTFVLYDVQRQRYVGHDRERAETRFVPASTYKVANSLIGLSTGAVRSADEVLPYGGKPQRFKAWEHDMSLRDAIKASNVPVYQELARRIGLERMRANVSRLGYGNAEIGQVVDNFWLVGPLKISAMEQTRFLLRLAQGELPFPAPVQSTVRAMTLLESGPGWELHGKTGWCFDCTPELGWWVGWVKRNERLYGFALNIDMPGGEADIGKRVELGKASLKALGILP',\n",
       " 'MRSEKEMMDLVLSLAEQDERIRIVTLEGSRANINIPKDEFQDYDITYFVSDIEPFISNDDWLNQFGNIIMMQKPEDMELFPPEEKGFSYLMLFDDYNKIDLTLLPLEELDNYLKGDKLIKVLIDKDCRIKRDIVPTDIDYHVRKPSAREYDDCCNEFWNVTPYVIKGLCRKEILFAIDHFNQIVRHELLRMISWKVGIETGFKLSVGKNYKFIERYISEDLWEKLLSTYRMDSYENIWEALFLCHQLFRAVSGEVAERLHYAYPEYDRNITKYTRDMYKKYTGKTGCLDSTYAADIEERREQ',\n",
       " 'MRSEKEMMDLVLSLAEQDERIRIVTLEGSRANINIPKDEFQDYDITYFVSDIEPFISNDDWLNQFGNIIMMQKPEDMELFPPEEKGFSYLMLFDDYNKIDLTLLPLEELDNYLKGDKLIKVLIDKDCRIKRDIVPTDIDYHVRKPSAREYDDCCNEFWNVTPYVIKGLCRKEILFAIDHLNQILRFELLRMMSWKVGIKTEFSLSVGKNYKYINKYIDEDLWNRLLSTYRMDSYENIWKSLFICHQLFREVSKEVAELLGFDYPEYGKNITRYTEDMYKKYVENDYF',\n",
       " 'MRSKNFSWRYSLAATVLLLSPFDLLASLGMDMYLPAVPFMPNALGTTASTIQLTLTTYLVMIGAGQLLFGPLSDRLGRRPVLLGGGLAYVVASMGLALTSSAEVFLGLRILQACGASACLVSTFATVRDIYAGREESNVIYGILGSMLAMVPAVGPLLGALVDMWLGWRAIFAFLGLGMIAASAAAWRFWPETRVQRVAGLQWSQLLLPVKCLNFWLYTLCYAAGMGSFFVFFSIAPGLMMGRQGVSQLGFSLLFATVAIAMVFTARFMGRVIPKWGSPSVLRMGMGCLIAGAVLLAITEIWASQSVLGFIAPMWLVGIGVATAVSVSPNGALRGFDHVAGTVTAVYFCLGGVLLGSIGTLIISLLPRNTAWPVVVYCLTLATVVLGLSCVSRVKGSRGQGEHDVVALQSAESTSNPNR',\n",
       " 'MRSSAIIALLIVGLDAMGLGLIMPVLPTLLRELVPAEQVAGHYGALLSLYALMQVVFAPMLGQLSDSYGRRPVLLASLAGAAVDYTIMASAPVLWVLYIGRLVSGVTGATGAVAASTIADSTGEGSRARWFGYMGACYGAGMIAGPALGGMLGGISAHAPFIAAALLNGFAFLLACIFLKETHHSHGGTGKPVRIKPFVLLRLDDALRGLGALFAVFFIIQLIGQVPAALWVIYGEDRFQWNTATVGLSLAAFGATHAIFQAFVTGPLSSRLGERRTLLFGMAADATGFVLLAFATQGWMVFPILLLLAAGGVGMPALQAMLSNNVSSNKQGALQGTLTSLTNLSSIAGPLGFTALYSATAGAWNGWVWIVGAILYLICLPILRRPFATSL',\n",
       " 'MRYIRLCIISLLATLPLAVHASPQPLEQIKLSESQLSGRVGMIEMDLASGRTLTAWRADERFPMMSTFKVVLCGAVLARVDAGDEQLERKIHYRQQDLVDYSPVSEKHLADGMTVGELCAAAITMSDNSAANLLLATVGGPAGLTAFLRQIGDNVTRLDRWETELNEALPGDARDTTTPASMAATLRKLLTSQRLSARSQRQLLQWMVDDRVAGPLIRSVLPAGWFIADKTGASKRGARGIVALLGPNNKAERIVVIYLRDTPASMAERNQQIAGIGAALIEHWQR',\n",
       " 'MRYIRLCIISLLATLPLAVHASPQPLEQIKQSESQLSGRVGMIEMDLASGRTLTAWRADERFPMMSTFKVVLCGAVLARVDAGDEQLERKIHYRQQDLVDYSPVSEKHLADGMTVGELCAAAITMSDNSAANLLLATVGGPAGLTAFLRQIGDNVTRLDRWETELNEALPGDARDTTTPASMAATLRKLLTSQRLSARSQRQLLQWMVDDRVAGPLIRSVLPAGWFIADKTGAGERGARGIVALLGPNNKAERIVVIYLRDTPASMAERNQQIAGIGAALIEHWQR',\n",
       " 'MRYIRLCIISLLATLPLAVHASPQPLEQIKQSESQLSGRVGMIEMDLASGRTLTAWRADERFPMMSTFKVVLCGAVLARVDAGDEQLERKIHYRQQDLVDYSPVSEKHLADGMTVGELCAAAITMSDNSAANLLLATVGGPAGLTAFLRQIGDNVTRLDRWETELNEALPGDARDTTTPASMAATLRKLLTSQRLSARSQRQLLQWMVDDRVAGPLIRSVLPAGWFIADKTGASKRGARGIVALLGPNNKAERIVVIYLRDTPASMAERNQQIAGIGAALIEHWQR',\n",
       " 'MSDFIQLEYLQEKLQQLLAESLFAIYLYGSAVDGGLGPESDLDVLVVVTQPLTSALREQLAQELLKISQPVGELQRPLEVTILLKDEIQSGNYPLSYEMQFGEWLREELKEGGTLSSQKDPDISILLRKARFHHAVLFGPALDQWAPEISDQELWQAMSDTYPEIVAHWDEDADERNQILALCRIYFSLVMKDIASKGNAARWVMPQLPPEQKFVLQRLIQEYRGEIGKQNWQEEHYALQPIVNFLSSKIEEQFEQKRNLIT',\n",
       " 'MSDFIQLEYLQEKLQQLLAKSLFAIYLYGSAVDGGLGPESDLDVMVVVTQPLTSALREQLAQELLKISQPVGELQRPLEVTILLKDEIQSGNYPLSYEMQFGEWLREELKEGGTLSSQKDPDISILLRKARFHHTVLFGPALNQWAPKISDQELWQAMSDTYPEIVAHWDEDADERNQILALCRIYFSLVMKDIASKDNAARWVMPQLPPEQKFVLQRLIQEYRGEIGKQNWQEEHYALQPIVNFLSSKIEEQFEQKRNLIT',\n",
       " 'MSHIQRETSCSRPRLNSNLDADLYGYRWARDNVGQSGATIYRLYGKPNAPELFLKHGKGSVANDVTDEMVRLNWLTAFMPLPTIKHFIRTPDDAWLLTTAIPGKTAFQVLEEYPDSGENIVDALAVFLRRLHSIPVCNCPFNSDRVFRLAQAQSRMNNGLVDASDFDDERNGWPVEQVWKEMHKLLPFSPDSVVTHGDFSLDNLIFDEGKLIGCIDVGRVGIADRYQDLAILWNCLGEFSPSLQKRLFQKYGIDNPDMNKLQFHLMLDEFF',\n",
       " 'MSHIQRETSCSRPRLNSNLDVDLYGYRWARDNVGQSGATIYRLYGKPDAPELFLKHGKGSVANDVTDEMVRLNWLTAFMPLPTIKHFIRTPDDAWLLTTAIPGKTAFQVLEEYPDSGENIVDALAAFLRRLHSIPVCNCPFNSDRVFRLAQAQSRMNNGLVDASDFDDERNGWPVEQVWKEMHKLLPFSPDSVVTHGDFSLDNLIFDEGKLIGCIDVGRVGIADRYQDLAILWNCLGEFSPSLQKRLFQKYGIDNPDMNKLQFHLMLDEFF',\n",
       " 'MSIDLSNKKIPKEAIQALKTIAELLDNMLIGVYLYGSAVMGGLRMNSDVDILVITNQSLSEKTRRNLTNRLMLISGKIGNIKDMRPLEVTVINQKDIVPWHFPPKYEFMYGEWLREQFEKGEIPESTYDPDLAILLAQLRKNSINLLGPKATEVIEPVPMTDIRKAIKESLPGLIASINGDERNVILTLARMWLTASTGEIRSKDLAAEWAIPQLPDEHATLLNKAREAYLGECVDKWEGMESEVAELVNHMKKSIESSLNIQLPFRIV',\n",
       " 'MSIQHFRVALIPFFAAFCLPVFAHPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMSTFKVLLCGAVLSRVDAGQEQLGRRIHYSQNDLVEYSPVTEKHLTDGMTVRELCSAAITMSDNTAANLLLTTIGGPKELTAFLHNMGDHVTRLDRWEPELNEAIPNDERDTTMPAAMATTLRKLLTGELLTLASRQQLIDWMEADKVAGPLLRSALPAGWFIADKSGAGERGSRGIIAALGPDGKPSRIVVIYTTGSQATMDERNRQIAEIGASLIKHW',\n",
       " 'MSLIIKARNIRLDYAGRDVLDIDELEIHSYDRIGLVGDNGAGKSSLLKVLNGEIVLAEATLQRFGDFAHISQLGGIEIETVEDRAMLSRLGVSNVQNDTMSGGEETRAKIAAAFSQQVHGILADEPTSHLDLNGIDLLIGQLKAFDGALLVISHDRYFLDMVVDKIWELKDGKITEYWGGYSDYLRQKEEERQHQAVEYELMMKERERLESAVQEKRQQANRLDNKKKGEKSKNSTESAGRLGHAKMTGTKQRKLYQAAKSMEKRLAALEDIQAPEHLRSIRFRQSSALELHNKFPITADGLSLKFGSRTIFDDANFIIPLGAKVAITGSNGTGKTSLLKMISERADGLTISPKAEIGYFTQTGYKFNTHKSVLSFMQEECEYTVAEIRAVLASMGIGANDIQKNLSDLSGGEIIKLLLSKMLLGKYNILLMDEPGNYLDLKSIAALETMMKSYAGTIIFVSHDKQLVDNIADIIYEIKDHKIIKTFERDC',\n",
       " 'MSLINVSNLTFSYEGSYDNIFENVSFQIDTDWKLGFIGRNGRGKTTFLNLLLGKYAYSGNISSTVKFEYFPYDVEDKSLYTIEVMKSICTECMDWEIFREISLLDVQEDALYRPFNTLSNGEQTKVLLAALFLTASCFLLIDEPTNHLDIDARNVVQNYLKRKKGFILVSHDRSLLDQCVDHILSINKTNIEIQKGNFTSWWENKTLQDNFELAENKKLLKEIGRLSYAAKRSSNWSNKVEKSKYGTTNSGSKLDKGYVGHKAAKAMKRAKNIESRHQEAVLQKSELLHNIEQYDDLKISPLEFHKECLIEANDLSLSYGDKEVCSNLNFRVNIGDRVAIIGKNGSGKSSILKLINGDDIKFTGNFMLASGLKISYISQDTSYLKGNLSEFAYNNKIDETLFKTILRKLDFNREQFDKNMVDFSAGQKKKVLIAKSLCESAHLYIWDEPLNYIDIFSRIQIEKMILEYCPTLLFVEHDDAFCNNICTKNINLGL',\n",
       " 'MSLQKTWGNIHLTALGAMMLSFLLVGCDDSVAQNAAPPAPTVSAAKVLVKSISQWDSFNGRIEAVESVQLRPRVSGYIDKVNYTDGQEVKKGQVLFTIDDRTYRAALEQAQAALARAKTQASLAQSEANRTDKLVHTNLVSREEWEQRRSAAVQAQADIRAAQAAVDAAQLNLDFTKVTAPIDGRASRALITSGNLVTAGDTASVLTTLVSQKTVYVYFDVDESTYLHYQNLARRGQGASSDNQALPVEIGLVGEEGYPHQGKVDFLDNQLTPSTGTIRMRALLDNSQRLFTPGLFARVRLPGSAEFKATLIDDKAVLTDQDRKYVYIVDKDGKAQRRDITPGRLADGLRIVQKGLNPGDSVIVDGLQKVFMPGMPVNAKTVAMTSSATLN',\n",
       " 'MSLYRRLVLLSCLSWPLAGFSATALTNLVAEPFAKLEQDFGGSIGVYAMDTGSGATVSYRAEERFPLCSSFKGFLAAAVLARSQQQAGLLDTPIRYGKNALVPWSPISEKYLTTGMTVAELSAAAVQYSDNAAANLLLKELGGPAGLTAFMRSIGDTTFRLDRWELELNSAIPGDARDTSSPRAVTESLQKLTLGSALAAPQRQQFVDWLKGNTTGNHRIRAAVPADWAVGDKTGTCGVYGTANDYAVVWPTGRAPIVLAVYTRAPNKDDKHSEAVIAAAARLALEGLGVNGQ',\n",
       " 'MSLYRRLVLLSCLSWPLAGFSATALTNLVAEPFAKLEQDFGGSIGVYAMDTGSGATVSYRAEERFPLCSSFKGFLAAAVLARSQQQAGLLDTPIRYGKNALVPWSPISEKYLTTGMTVAELSAAAVQYSDNAAANLLLKELGGPAGLTAFMRSIGDTTFRLDRWELELNSAIPGDARDTSSPRAVTESLQKLTLGSALAAPQRQQFVDWLKGNTTGNHRIRAAVPADWAVGDKTGTCGVYGTANDYAVVWPTGRAPIVLAVYTRAPNKDDKYSEAVIAAAARLALEGLGVNGQ',\n",
       " 'MSNLINGKIPNQAIQTLKIVKDLFGSSIVGVYLFGSAVNGGLRINSDVDVLVVVNHSLPQLTRKKLTERLMTISGKIGNTDSVRPLEVTVINRSEVVPWQYPPKREFIYGEWLRGEFENGQIQEPSYDPDLAIVLAQARKNSISLFGPDSSSILVSVPLTDIRRAIKDSLPELIEGIKGDERNVILTLARMWQTVTTGEITSKDVAAEWAIPLLPKEHVTLLDIARKGYRGECDDKWEGLYSKVKALVKYMKNSIETSLN',\n",
       " 'MSYTRVDISSWNRREHFEIFRGDGQCTFSQTVQLDITRLLDFTRSRGYRFYPVFIHSIAKVVNRFPEYRMAMKGDELIVWDCVHPNYTTFHPDTETFSSFWSHYHDDLARFLAEYSMDREKYRNDHSYFPKGFIENVFYVSANPWVSFTSFDFNFASANNFFAPLFTVGKYYSQAGKTLVPLAVQVHHAVCDGFHAARLVTELQKLCDEAGG',\n",
       " 'MTDQATPNLPSRDFDSTAAFYERLGFGIVFRDAGWMILQRGDLMLEFFAHPGLDPLASWFSCCLRLDDLAEFYRQCKSVGIQETSSGYPRIHAPELQEWGGTMAALVDPDGTLLRLIQNELLAGIS',\n",
       " 'MTFNIINLETWDRKEYFNHYFNQQTTYSVTKELDITLLKSMIKDKGYELYPALIHAIVSVINRNKVFRTGINSEGNLGYWDKLEPLYTVFNKETEKFSNIWTESNASFNSFYNSYKNDLFKYKDKNEMFPKKPIPENTVPISMIPWIDFSSFNLNIGNNSRFLLPIITIGKFYSKDDKIYLPFSLQVHHAVCDGYHVSLFMNEFQNIIDNVNEWI',\n",
       " 'MTIQDIQSLAEAHGLLLTDKMNFNEMGIDFKVVFALDTKGQQWLLRIPRRDGMREQIKKEKRILELVKKHLSVEVPDWRISSTELVAYPILKDNPVLNLDAETYEIIWNMDKDSPKYITSLAKTLFEIHSIPEKEVRENDLKIMKPSDLRPEIANNLQLVKSEIGISEQLETRYRKWLDNDVLWADFTQFIHGDLYAGHVLASKDGAVSGVIDWSTAHIDDPAIDFAGHVTLFGEESLKTLIIEYEKLGGKVWNKLYEQTLERAAASPLMYGLFALETQNESLIVGAKAQLGVI',\n",
       " 'MTKIMPEYTVVDLSRWARKEHFEVFQGFAQSTFNQTVQLDITVLLKHIKEVGWKFYPAIISLISDVVNRHPEFRMAMKDDELVIWNEVHPSYTLFHKETETFSSLWSHYDGNIHHFQRVYAEDVARYGNILAYWPKEESRENIFFISDIPWVSFSSFNVNVANMRNFFAPMFTFGKYYNQDGKVLLPFAVQVHHSVCDGFHVARLINELQELCDNLPHHSEALNV',\n",
       " 'MTNKQVEISMAEWDVMNIIWDKKSVSANEIVVEIQKYKEVSDKTIRTLITRLYKKEIIKRYKSENIYFYSSNIKEDDIKMKTAKTFLNKLYGGDMKSLVLNFAKNEELNNKEIEELRDILNDISKK',\n",
       " 'MTNSNDSVTLRLMTEHDLAMLYEWLNRSHIVEWWGGEEARPTLADVQEQYLPSVLAQESVTPYIAMLNGEPIGYAQSYVALGSGDGWWEEETDPGVRGIDQLLANASQLGKGLGTKLVRALVELLFNDPEVTKIQTDPSPSNLRAIRCYEKAGFERQGTVTTPDGPAVYMVQTRQAFERTRSDA',\n",
       " 'MTNSNDSVTLRLMTEHDLAMLYEWLNRSHIVEWWGGEEARPTLADVQEQYLPSVLAQESVTPYIAMLNGEPIGYAQSYVALGSGDGWWEEETDPGVRGIDQLLANASQLGKGLGTKLVRALVELLFNDPEVTKIQTDPSPSNLRAIRCYEKAGFERQGTVTTPDGPAVYMVQTRQAFERTRSVA',\n",
       " 'MTNSTDSVTLRLMTEHDLAMLYEWLNRSHIVEWWGGEEARPTLADVQEQYLPSVLAQESVTPYIAMLNGEPIGYAQSYVALGSGDGWWEEETDPGVRGIDQSLANASQLGKGLGTKLVRALVELLFNDPEVTKIQTDPSPSNLRAIRCYEKAGFERQGTVTTPDGPAVYMVQTRQAFERTRSDA',\n",
       " 'MTNYFDSPFKGKLLSEQVKNPNIKVGRYSYYSGYYHGHSFDDCARYLFPDRDDVDKLIIGSFCSIGSGASFIMAGNQGHRYDWASSFPFFYMQEEPAFSSALDAFQKAGNTVIGNDVWIGSEAMVMPGIKIGHGAVIGSRSLVTKDVEPYAIVGGNPAKKIKKRFTDEEISLLLEMEWWNWSLEKIKAAMPMLCSSNIVGLHKYWLEFAV',\n",
       " 'MTTTRPAWAYTLPAALLLMAPFDILASLAMDIYLPVVPAMPGVLNTTPSIIQLTLSLYMVMLGVGQVIFGPLSDRVGRRPILLVGATAFVAASLGAACSSTALAFVAFRLVQAVGASAMLVATFATVRDVYANRPEGAVIYGLFSSMLAFVPALGPIAGALIGEFWGWQAIFITLAALASLALLNASFRWHETRPLDQARTQRSVLPIFASPAFWVYTVGFSAGMGTFFVFFSTAPRVLIGQAGYSEIGFSLAFATVALVMVTTTRFAKSFVAKWGIAGCVARGMALLVSGAILLGIGQLFGSPSFFSFILPMWVVAVGIVFTVSVTANGALAQFDDIAGSAVAFYFCIQSLIVSIVGTLAVTLLNGDTAWPVICYATAMAVLVSLGLALLRSRDAATEKSPVV',\n",
       " 'MVKDWIPISHDNYKQVQGPFYHGTKANLAIGDLLTTGFISHFEDGRILKHIYFSALMEPAVWGAELAMSLSGLEGRGYIYIVEPTGPFEDDPNLTNKKFPGNPTQSYRTCEPLRIVGVVEDWEGHPVELIRGMLDSLEDLKRRGLHVIED',\n",
       " 'MVTVFGILNLTEDSFFDESRRLDPAGAVTAAIEMLRVGADVVDVGPAASHPDARPVSPADEIRRIAPLLDALSDQMHRVSIDSFQPETQRYALKRGVGYLNDIQGFPDPALYPDIAEADCRLVVMHSAQRDGIATRTGHLRPEDALDEIVRFFEARVSALRRSGVAADRLILDPGMGFFLSPAPETSLHVLSNLQKLKSALGLPLLVSVSRKSFLGATVGLPVKDLGPASLAAELHAIGNGADYVRTHAPGDLRSAITFSETLAKFRSRDARDRGLDHA',\n",
       " 'MVTVFGILNLTEDSFFDESRRLDPAGAVTAAIEMLRVGSDVVDVGPAASHPDARPVSPADEIRRIAPLLDALSDQMHRVSIDSFQPETQRYALKRGVGYLNDIQGFPDPALYPDIAEADCRLVVMHSAQRDGIATRTGHLRPEDALDEIVRFFEARVSALRRSGVAADRLILDPGMDQRREHHIEQRDATRNRDGRIFYRRNLLATLREREVARAGAEMAEGKALPFRAAKDGESVSGKFTGTVQLTSGKFAIVEKSHEFTLVPWRPIIDRQLGREVAGIMQGGSVSWQLGRQRGLGL',\n",
       " 'MVTVFGILNLTEDSFFDESRRLDPAGAVTAAIEMLRVGSDVVDVGPAASHPDARPVSPADEIRRIAPLLDALSDQMHRVSIDSFQPETQRYALKRGVGYLNDIQGFPDPALYPDIAEADCRLVVMHSAQRDGIATRTGHLRPEDALDEIVRFFEARVSALRRSGVAADRLILDPGMGFFLSPAPETSLHVLSNLQKLKSALGLPLLVSVSRKSFLGATVGLPVKDLGPASLAAELHAIGNGADYVRTHAPGDLRSAITFSETLAKFRSRDARDRGLDHA',\n",
       " 'MVTVFGILNLTEDSFFDESRRLDPAGAVTAAIEMLRVGSDVVDVGPAASHPDARPVSPADEIRRIAPLLDALSDQMHRVSIDSFQPETQRYALKRGVGYLNDIQGFPDPALYPDIAEADCRLVVMHSAQRDGIATRTGHLRPEDALDEIVRFFEARVSALRRSGVAADRLILDPGMGFFLSPAPETSLHVLSNLQKLKSALGLPLLVSVSRKSFLGATVGLPVKDLGPASLAAELHAIGNGADYVRTHAPGDLRSAITISETLAKFRSRDARDRGLDHA',\n",
       " 'MVYGECMSVQIPTPWMEKLSAYDWAQQTTGCSEASVFRGAAPGKSTLFVKTEPGGPLSELQDEAARLRWLATTGLPCAQVLDAAHEAGRDWLLLSAVPGEDLLSSPIDPVEKVTIIADVLRQLHGLDPVTCPFDHRAEHRIERARARMQAGLVDQDDLDDEHQGLEPAELFARLQARGPAYEDLVVTHGDACLPNIMVENGRFSGFIDCGRLGVADRYQDLALATRDVAEELGDEWVQPFLDRYGIGRLDPDRVAFYRLLDEFY'}"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "####tring a faster way instead!\n",
    "# Convert the list of protein sequences to a set for faster lookup\n",
    "sequence_set = set(amr_protein_df['sequence'])\n",
    "sequence_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "80dda7b0-6a02-4f9b-b8a8-747d975f93c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter rows on an exact match for any protein sequence\n",
    "# Here, we use a list comprehension to find headers that exactly match any protein sequence from the set\n",
    "sub_sample_all_df['AMR_status'] = sub_sample_all_df['sequence'].apply(lambda x: \"AMR\" if x in sequence_set else \"non-AMR\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "id": "49a135ff-5f4e-437a-9c91-72c88181524e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>header</th>\n",
       "      <th>sequence</th>\n",
       "      <th>AMR_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3089</th>\n",
       "      <td>ABR81255.1</td>\n",
       "      <td>M       R       L       L       F       F     ...</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3095</th>\n",
       "      <td>ABR81261.1</td>\n",
       "      <td>M       T       D       Q       A       T     ...</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3367</th>\n",
       "      <td>ABR81550.1</td>\n",
       "      <td>M       S       Y       T       R       V     ...</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3706</th>\n",
       "      <td>ABR81912.1</td>\n",
       "      <td>M       I       E       Q       D       G     ...</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6733</th>\n",
       "      <td>ABR85124.1</td>\n",
       "      <td>M       R       H       A       T       I     ...</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118356</th>\n",
       "      <td>UDP42657.1</td>\n",
       "      <td>M       R       S       E       K       E     ...</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118357</th>\n",
       "      <td>UDP42658.1</td>\n",
       "      <td>M       N       K       N       I       K     ...</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118361</th>\n",
       "      <td>UDP42662.1</td>\n",
       "      <td>M       N       T       S       Y       S     ...</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118363</th>\n",
       "      <td>UDP42664.1</td>\n",
       "      <td>M       R       S       E       K       E     ...</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118367</th>\n",
       "      <td>UDP42668.1</td>\n",
       "      <td>M       L       K       Q       K       E     ...</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>180 rows  3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            header                                           sequence  \\\n",
       "3089    ABR81255.1  M       R       L       L       F       F     ...   \n",
       "3095    ABR81261.1  M       T       D       Q       A       T     ...   \n",
       "3367    ABR81550.1  M       S       Y       T       R       V     ...   \n",
       "3706    ABR81912.1  M       I       E       Q       D       G     ...   \n",
       "6733    ABR85124.1  M       R       H       A       T       I     ...   \n",
       "...            ...                                                ...   \n",
       "118356  UDP42657.1  M       R       S       E       K       E     ...   \n",
       "118357  UDP42658.1  M       N       K       N       I       K     ...   \n",
       "118361  UDP42662.1  M       N       T       S       Y       S     ...   \n",
       "118363  UDP42664.1  M       R       S       E       K       E     ...   \n",
       "118367  UDP42668.1  M       L       K       Q       K       E     ...   \n",
       "\n",
       "       AMR_status  \n",
       "3089          AMR  \n",
       "3095          AMR  \n",
       "3367          AMR  \n",
       "3706          AMR  \n",
       "6733          AMR  \n",
       "...           ...  \n",
       "118356        AMR  \n",
       "118357        AMR  \n",
       "118361        AMR  \n",
       "118363        AMR  \n",
       "118367        AMR  \n",
       "\n",
       "[180 rows x 3 columns]"
      ]
     },
     "execution_count": 224,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sub_sample_all_df[sub_sample_all_df['AMR_status'] == 'AMR']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "id": "203939b8-fe4d-4269-8f25-5f97d8843e9d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>header</th>\n",
       "      <th>sequence</th>\n",
       "      <th>AMR_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAI79689.1</td>\n",
       "      <td>M       S       E       K       E       I     ...</td>\n",
       "      <td>non-AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CAI79690.1</td>\n",
       "      <td>M       M       E       F       T       I     ...</td>\n",
       "      <td>non-AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CAI79691.1</td>\n",
       "      <td>M       I       I       L       V       Q     ...</td>\n",
       "      <td>non-AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CAI79692.1</td>\n",
       "      <td>M       K       L       N       T       L     ...</td>\n",
       "      <td>non-AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CAI79693.1</td>\n",
       "      <td>M       V       T       A       L       S     ...</td>\n",
       "      <td>non-AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118374</th>\n",
       "      <td>UDP42675.1</td>\n",
       "      <td>M       N       L       F       K       E     ...</td>\n",
       "      <td>non-AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118375</th>\n",
       "      <td>UDP42676.1</td>\n",
       "      <td>M       D       Y       G       F       I     ...</td>\n",
       "      <td>non-AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118376</th>\n",
       "      <td>UDP42677.1</td>\n",
       "      <td>M       P       I       A       I       G     ...</td>\n",
       "      <td>non-AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118377</th>\n",
       "      <td>UDP42678.1</td>\n",
       "      <td>D       P       V       S       F       K     ...</td>\n",
       "      <td>non-AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118378</th>\n",
       "      <td>UDP42679.1</td>\n",
       "      <td>M       A       N       E       N       Q     ...</td>\n",
       "      <td>non-AMR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>118379 rows  3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            header                                           sequence  \\\n",
       "0       CAI79689.1  M       S       E       K       E       I     ...   \n",
       "1       CAI79690.1  M       M       E       F       T       I     ...   \n",
       "2       CAI79691.1  M       I       I       L       V       Q     ...   \n",
       "3       CAI79692.1  M       K       L       N       T       L     ...   \n",
       "4       CAI79693.1  M       V       T       A       L       S     ...   \n",
       "...            ...                                                ...   \n",
       "118374  UDP42675.1  M       N       L       F       K       E     ...   \n",
       "118375  UDP42676.1  M       D       Y       G       F       I     ...   \n",
       "118376  UDP42677.1  M       P       I       A       I       G     ...   \n",
       "118377  UDP42678.1  D       P       V       S       F       K     ...   \n",
       "118378  UDP42679.1  M       A       N       E       N       Q     ...   \n",
       "\n",
       "       AMR_status  \n",
       "0         non-AMR  \n",
       "1         non-AMR  \n",
       "2         non-AMR  \n",
       "3         non-AMR  \n",
       "4         non-AMR  \n",
       "...           ...  \n",
       "118374    non-AMR  \n",
       "118375    non-AMR  \n",
       "118376    non-AMR  \n",
       "118377    non-AMR  \n",
       "118378    non-AMR  \n",
       "\n",
       "[118379 rows x 3 columns]"
      ]
     },
     "execution_count": 225,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sub_sample_all_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f260fa4-ad3d-401e-964d-8d8bb0795ad2",
   "metadata": {},
   "source": [
    "#### Total number of AMR protein sequences found = 180\n",
    "#### Total number of non-AMR protein sequences found = 118199"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "dfb060db-d3ec-434f-9d01-e6a65b967b86",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Save this as a training dataset to a TSV file\n",
    "sub_sample_all_df.to_csv(\"training_set.tsv\", sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40ded068-3bc2-405f-851b-cd22298c7a49",
   "metadata": {},
   "source": [
    "#### ##Now trying transformer"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d00e733c-2f2d-42f2-8f12-5b11beb9cc5a",
   "metadata": {},
   "source": [
    "### Formatting the training set for as input into the transformer model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "id": "849d4005-06b7-4b52-a223-c5a778b9d317",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>header</th>\n",
       "      <th>sequence</th>\n",
       "      <th>AMR_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAI79689.1</td>\n",
       "      <td>M S E K E I W E K V L K I A Q E K L S A V S Y ...</td>\n",
       "      <td>non-AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CAI79690.1</td>\n",
       "      <td>M M E F T I K R D Y F I T Q L N D T L K A I S ...</td>\n",
       "      <td>non-AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CAI79691.1</td>\n",
       "      <td>M I I L V Q E V V V E G D I N L G Q F L K T E ...</td>\n",
       "      <td>non-AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CAI79692.1</td>\n",
       "      <td>M K L N T L Q L E N Y R N Y D E V T L K C H P ...</td>\n",
       "      <td>non-AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CAI79693.1</td>\n",
       "      <td>M V T A L S D V N N T D N Y G A G Q I Q V L E ...</td>\n",
       "      <td>non-AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118374</th>\n",
       "      <td>UDP42675.1</td>\n",
       "      <td>M N L F K E E I G N W Q D W S N V S Q S L F A ...</td>\n",
       "      <td>non-AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118375</th>\n",
       "      <td>UDP42676.1</td>\n",
       "      <td>M D Y G F I S T I V R S E L F M M Q L D S V L ...</td>\n",
       "      <td>non-AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118376</th>\n",
       "      <td>UDP42677.1</td>\n",
       "      <td>M P I A I G N K R L P V T L D E K R Q K E L Q ...</td>\n",
       "      <td>non-AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118377</th>\n",
       "      <td>UDP42678.1</td>\n",
       "      <td>D P V S F K P I Y G R S G K L E T A Y W G S R ...</td>\n",
       "      <td>non-AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118378</th>\n",
       "      <td>UDP42679.1</td>\n",
       "      <td>M A N E N Q L S E I Q S K I E G R G A E K R T ...</td>\n",
       "      <td>non-AMR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>118379 rows  3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            header                                           sequence  \\\n",
       "0       CAI79689.1  M S E K E I W E K V L K I A Q E K L S A V S Y ...   \n",
       "1       CAI79690.1  M M E F T I K R D Y F I T Q L N D T L K A I S ...   \n",
       "2       CAI79691.1  M I I L V Q E V V V E G D I N L G Q F L K T E ...   \n",
       "3       CAI79692.1  M K L N T L Q L E N Y R N Y D E V T L K C H P ...   \n",
       "4       CAI79693.1  M V T A L S D V N N T D N Y G A G Q I Q V L E ...   \n",
       "...            ...                                                ...   \n",
       "118374  UDP42675.1  M N L F K E E I G N W Q D W S N V S Q S L F A ...   \n",
       "118375  UDP42676.1  M D Y G F I S T I V R S E L F M M Q L D S V L ...   \n",
       "118376  UDP42677.1  M P I A I G N K R L P V T L D E K R Q K E L Q ...   \n",
       "118377  UDP42678.1  D P V S F K P I Y G R S G K L E T A Y W G S R ...   \n",
       "118378  UDP42679.1  M A N E N Q L S E I Q S K I E G R G A E K R T ...   \n",
       "\n",
       "       AMR_status  \n",
       "0         non-AMR  \n",
       "1         non-AMR  \n",
       "2         non-AMR  \n",
       "3         non-AMR  \n",
       "4         non-AMR  \n",
       "...           ...  \n",
       "118374    non-AMR  \n",
       "118375    non-AMR  \n",
       "118376    non-AMR  \n",
       "118377    non-AMR  \n",
       "118378    non-AMR  \n",
       "\n",
       "[118379 rows x 3 columns]"
      ]
     },
     "execution_count": 240,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import re\n",
    "###first introduce spaces between each protein sequence character\n",
    "sub_sample_all_df['sequence'] = sub_sample_all_df['sequence'].apply(lambda x: re.sub(r'\\s+','',x))\n",
    "sub_sample_all_df['sequence'] = sub_sample_all_df['sequence'].apply(lambda x: ' '.join(list(x)))\n",
    "sub_sample_all_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "9d5249d6-663d-478c-a6f5-befcf2a02c18",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: datasets in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (3.0.0)\n",
      "Requirement already satisfied: filelock in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from datasets) (3.16.1)\n",
      "Requirement already satisfied: numpy>=1.17 in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from datasets) (2.1.1)\n",
      "Requirement already satisfied: pyarrow>=15.0.0 in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from datasets) (17.0.0)\n",
      "Requirement already satisfied: dill<0.3.9,>=0.3.0 in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from datasets) (0.3.8)\n",
      "Requirement already satisfied: pandas in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from datasets) (2.2.3)\n",
      "Requirement already satisfied: requests>=2.32.2 in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from datasets) (2.32.3)\n",
      "Requirement already satisfied: tqdm>=4.66.3 in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from datasets) (4.66.5)\n",
      "Requirement already satisfied: xxhash in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from datasets) (3.5.0)\n",
      "Requirement already satisfied: multiprocess in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from datasets) (0.70.16)\n",
      "Requirement already satisfied: fsspec<=2024.6.1,>=2023.1.0 in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from fsspec[http]<=2024.6.1,>=2023.1.0->datasets) (2024.6.1)\n",
      "Requirement already satisfied: aiohttp in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from datasets) (3.10.5)\n",
      "Requirement already satisfied: huggingface-hub>=0.22.0 in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from datasets) (0.25.1)\n",
      "Requirement already satisfied: packaging in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from datasets) (24.1)\n",
      "Requirement already satisfied: pyyaml>=5.1 in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from datasets) (6.0.2)\n",
      "Requirement already satisfied: aiohappyeyeballs>=2.3.0 in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from aiohttp->datasets) (2.4.0)\n",
      "Requirement already satisfied: aiosignal>=1.1.2 in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from aiohttp->datasets) (1.3.1)\n",
      "Requirement already satisfied: attrs>=17.3.0 in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from aiohttp->datasets) (24.2.0)\n",
      "Requirement already satisfied: frozenlist>=1.1.1 in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from aiohttp->datasets) (1.4.1)\n",
      "Requirement already satisfied: multidict<7.0,>=4.5 in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from aiohttp->datasets) (6.1.0)\n",
      "Requirement already satisfied: yarl<2.0,>=1.0 in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from aiohttp->datasets) (1.12.1)\n",
      "Requirement already satisfied: typing-extensions>=3.7.4.3 in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from huggingface-hub>=0.22.0->datasets) (4.12.2)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from requests>=2.32.2->datasets) (3.3.2)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from requests>=2.32.2->datasets) (3.7)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from requests>=2.32.2->datasets) (2.2.3)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from requests>=2.32.2->datasets) (2024.8.30)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from pandas->datasets) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from pandas->datasets) (2024.2)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from pandas->datasets) (2024.2)\n",
      "Requirement already satisfied: six>=1.5 in /home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages (from python-dateutil>=2.8.2->pandas->datasets) (1.16.0)\n"
     ]
    }
   ],
   "source": [
    "## split between training and test dataset\n",
    "\n",
    "! pip install datasets\n",
    "from datasets import Dataset, DatasetDict\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "id": "6f9a2474-9d39-4741-aebf-90c3410e1e66",
   "metadata": {},
   "outputs": [],
   "source": [
    "amr_df = sub_sample_all_df[sub_sample_all_df['AMR_status']=='AMR']\n",
    "non_amr_df = sub_sample_all_df[sub_sample_all_df['AMR_status']=='non-AMR'].sample(720,random_state=7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "id": "0e6783e5-0082-4410-b869-55ccdd2bde77",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Dataset({\n",
       "    features: ['header', 'sequence', 'AMR_status'],\n",
       "    num_rows: 900\n",
       "})"
      ]
     },
     "execution_count": 242,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hf_dataset = pd.concat([non_amr_df,amr_df]).reset_index(drop=True)\n",
    "hf_dataset = Dataset.from_pandas(hf_dataset)\n",
    "hf_dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "id": "42f6286c-8778-46a0-a439-5b7d2c6fa5c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Convert to Hugging Face Dataset\n",
    "# hf_dataset = Dataset.from_pandas(sub_sample_all_df)\n",
    "# hf_dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "id": "ed4d9d55-2a8a-4127-a3c4-ea38352722ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split into train and test datasets (80% train, 20% test)\n",
    "train_test = hf_dataset.train_test_split(test_size=0.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "id": "b97545ba-0bb9-4bb3-8b3d-6af1f463459a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split the train dataset further into train and validation sets (80% train, 20% validation)\n",
    "train_valid = train_test['train'].train_test_split(test_size=0.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "id": "81c42559-cc79-4ed6-984d-2149e000c70a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DatasetDict({\n",
      "    train: Dataset({\n",
      "        features: ['header', 'sequence', 'AMR_status'],\n",
      "        num_rows: 576\n",
      "    })\n",
      "    validation: Dataset({\n",
      "        features: ['header', 'sequence', 'AMR_status'],\n",
      "        num_rows: 144\n",
      "    })\n",
      "    test: Dataset({\n",
      "        features: ['header', 'sequence', 'AMR_status'],\n",
      "        num_rows: 180\n",
      "    })\n",
      "})\n"
     ]
    }
   ],
   "source": [
    "# Combine train, validation, and test sets into a single DatasetDict\n",
    "final_splits = DatasetDict({\n",
    "    'train': train_valid['train'],\n",
    "    'validation': train_valid['test'],\n",
    "    'test': train_test['test']\n",
    "})\n",
    "\n",
    "# Print the final splits\n",
    "print(final_splits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "id": "20dedd48-fe8b-48e0-80a2-49190bfb2cfe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'header': 'BAB42435.1',\n",
       " 'sequence': 'M F N E K D Q L A V D T L R A L S I D T I E K A N S G H P G L P M G A A P M A Y T L W T R H L N F N P Q S K D Y F N R D R F V L S A G H G S A L L Y S L L H V S G S L E L E E L K Q F R Q W G S K T P G H P E Y R H T D G V E V T T G P L G Q G F A M S V G L A L A E D H L A G K F N K E G Y N V V D H Y T Y V L A S D G D L M E G I S H E A A S F A G H N K L S K L V V L Y D S N D I S L D G E L N K A F S E N T K A R F E A Y G W N Y L L V K D G N D L E E I D K A I T T A K S Q E G P T I I E V K T T I G F G S P N K A G T N G V H G A P L G E V E R K L T F E N Y G L D P E K R F N V S E E V Y E I F Q N T M L K R A N E D E S Q W N S L L E K Y A E T Y P E L A E E F K L A I S G K L P K N Y K D E L P R F E L G H N G A S R A D S G T V I Q A I S K T V P S F F G G S A D L A G S N K S N V N D A T D Y S S E T P E G K N V W F G V R E F A M G A A V N G M A A H G G L H P Y G A T F F V F S D Y L K P A L R L S S I M G L N A T F I F T H D S I A V G E D G P T H E P I E Q L A G L R A I P N M N V I R P A D G N E T R V A W E V A L E S E S T P T S L V L T R Q N L P V L D V P E D V V E E G V R K G A Y T V Y G S E E T P E F L L L A S G S E V S L A V E A A K D L E K Q G K S V R V V S M P N W N A F E Q Q S E E Y K E S V I P S S V T K R V A I E M A S P L G W H K Y V G T A G K V I A I D G F G A S A P G D L V V E K Y G F T K E N I L N Q V M S L',\n",
       " 'AMR_status': 'non-AMR'}"
      ]
     },
     "execution_count": 247,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_splits['test'][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "394f15d4-baa6-489a-9cf5-c3dc96ff10c6",
   "metadata": {},
   "source": [
    "#### Setting up transformer model to run with the above dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "id": "fba21a58-5267-4530-85c3-edc1737108fb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# ! pip install transformers datasets evaluate accelerate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "id": "273e776d-4700-4c51-9658-4ab9d1820164",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Some weights of BertForSequenceClassification were not initialized from the model checkpoint at Rostlab/prot_bert and are newly initialized: ['classifier.bias', 'classifier.weight']\n",
      "You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.\n"
     ]
    }
   ],
   "source": [
    "from transformers import BertTokenizer, BertForSequenceClassification\n",
    "tokenizer = BertTokenizer.from_pretrained('Rostlab/prot_bert')\n",
    "model = BertForSequenceClassification.from_pretrained('Rostlab/prot_bert', num_labels=2)  # Assuming binary classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "id": "7ebac9ad-a528-4708-bef2-90a43d4e1f56",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "da4bc7e2a33b418fa19adc8254740dd0",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Map:   0%|          | 0/576 [00:00<?, ? examples/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "345311245d404d619e39fca3f7350d16",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Map:   0%|          | 0/144 [00:00<?, ? examples/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f0dc6c3928c945c1a65a6f9b9b9ee085",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Map:   0%|          | 0/180 [00:00<?, ? examples/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from datasets import DatasetDict\n",
    "from transformers import BertTokenizer\n",
    "\n",
    "# Dataset should be in DatasetDict\n",
    "def preprocess_function(examples):\n",
    "    tokenized = tokenizer(examples[\"sequence\"], padding=\"max_length\", truncation=True, max_length=340)\n",
    "    tokenized[\"labels\"] = [1 if status == \"AMR\" else 0 for status in examples[\"AMR_status\"]]    \n",
    "    return tokenized\n",
    "\n",
    "\n",
    "# Tokenize your dataset\n",
    "tokenized_datasets = final_splits.map(preprocess_function, batched=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "id": "87df40f6-18b6-4041-8f89-2155ca7763df",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import precision_score, recall_score, f1_score, accuracy_score\n",
    "\n",
    "def compute_metrics(p):\n",
    "    preds = p.predictions.argmax(-1)  # Get the predicted class (for binary classification, will give 0 or 1)\n",
    "    labels = p.label_ids  # labels (0 or 1)\n",
    "    precision = precision_score(labels, preds, average='binary')\n",
    "    recall = recall_score(labels, preds, average='binary')\n",
    "    f1 = f1_score(labels, preds, average='binary')\n",
    "    accuracy = accuracy_score(labels, preds)\n",
    "\n",
    "    return {\n",
    "        'accuracy': accuracy,\n",
    "        'precision': precision,\n",
    "        'recall': recall,\n",
    "        'f1': f1\n",
    "    }\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "id": "10c3997e-3f82-4e28-a130-eb3e4338075f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages/transformers/training_args.py:1545: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of  Transformers. Use `eval_strategy` instead\n",
      "  warnings.warn(\n",
      "/home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages/torch/nn/parallel/_functions.py:68: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.\n",
      "  warnings.warn('Was asked to gather along dimension 0, but all '\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "    <div>\n",
       "      \n",
       "      <progress value='50' max='50' style='width:300px; height:20px; vertical-align: middle;'></progress>\n",
       "      [50/50 01:37, Epoch 10/10]\n",
       "    </div>\n",
       "    <table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       " <tr style=\"text-align: left;\">\n",
       "      <th>Epoch</th>\n",
       "      <th>Training Loss</th>\n",
       "      <th>Validation Loss</th>\n",
       "      <th>Accuracy</th>\n",
       "      <th>Precision</th>\n",
       "      <th>Recall</th>\n",
       "      <th>F1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>No log</td>\n",
       "      <td>0.560694</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>No log</td>\n",
       "      <td>0.588408</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>No log</td>\n",
       "      <td>0.440004</td>\n",
       "      <td>0.756944</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.027778</td>\n",
       "      <td>0.054054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>No log</td>\n",
       "      <td>0.303561</td>\n",
       "      <td>0.944444</td>\n",
       "      <td>0.888889</td>\n",
       "      <td>0.888889</td>\n",
       "      <td>0.888889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>No log</td>\n",
       "      <td>0.219373</td>\n",
       "      <td>0.951389</td>\n",
       "      <td>0.967742</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.895522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>No log</td>\n",
       "      <td>0.178150</td>\n",
       "      <td>0.958333</td>\n",
       "      <td>0.894737</td>\n",
       "      <td>0.944444</td>\n",
       "      <td>0.918919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>No log</td>\n",
       "      <td>0.211285</td>\n",
       "      <td>0.916667</td>\n",
       "      <td>0.772727</td>\n",
       "      <td>0.944444</td>\n",
       "      <td>0.850000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>No log</td>\n",
       "      <td>0.170183</td>\n",
       "      <td>0.944444</td>\n",
       "      <td>0.850000</td>\n",
       "      <td>0.944444</td>\n",
       "      <td>0.894737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>No log</td>\n",
       "      <td>0.146385</td>\n",
       "      <td>0.965278</td>\n",
       "      <td>0.918919</td>\n",
       "      <td>0.944444</td>\n",
       "      <td>0.931507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>No log</td>\n",
       "      <td>0.141696</td>\n",
       "      <td>0.965278</td>\n",
       "      <td>0.918919</td>\n",
       "      <td>0.944444</td>\n",
       "      <td>0.931507</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><p>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
      "/home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages/torch/nn/parallel/_functions.py:68: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.\n",
      "  warnings.warn('Was asked to gather along dimension 0, but all '\n",
      "/home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
      "/home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages/torch/nn/parallel/_functions.py:68: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.\n",
      "  warnings.warn('Was asked to gather along dimension 0, but all '\n",
      "/home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages/torch/nn/parallel/_functions.py:68: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.\n",
      "  warnings.warn('Was asked to gather along dimension 0, but all '\n",
      "/home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages/torch/nn/parallel/_functions.py:68: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.\n",
      "  warnings.warn('Was asked to gather along dimension 0, but all '\n",
      "/home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages/torch/nn/parallel/_functions.py:68: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.\n",
      "  warnings.warn('Was asked to gather along dimension 0, but all '\n",
      "/home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages/torch/nn/parallel/_functions.py:68: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.\n",
      "  warnings.warn('Was asked to gather along dimension 0, but all '\n",
      "/home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages/torch/nn/parallel/_functions.py:68: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.\n",
      "  warnings.warn('Was asked to gather along dimension 0, but all '\n",
      "/home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages/torch/nn/parallel/_functions.py:68: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.\n",
      "  warnings.warn('Was asked to gather along dimension 0, but all '\n",
      "/home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages/torch/nn/parallel/_functions.py:68: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.\n",
      "  warnings.warn('Was asked to gather along dimension 0, but all '\n",
      "/home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages/torch/nn/parallel/_functions.py:68: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.\n",
      "  warnings.warn('Was asked to gather along dimension 0, but all '\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "TrainOutput(global_step=50, training_loss=0.22764545440673828, metrics={'train_runtime': 100.3408, 'train_samples_per_second': 57.404, 'train_steps_per_second': 0.498, 'total_flos': 4452701487667200.0, 'train_loss': 0.22764545440673828, 'epoch': 10.0})"
      ]
     },
     "execution_count": 265,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from transformers import Trainer, TrainingArguments\n",
    "\n",
    "training_args = TrainingArguments(\n",
    "    output_dir=\"./fine-tuned-prot-bert\",\n",
    "    evaluation_strategy=\"epoch\",\n",
    "    learning_rate=5e-5,\n",
    "    per_device_train_batch_size=32,\n",
    "    per_device_eval_batch_size=32,\n",
    "    num_train_epochs=10,\n",
    "    weight_decay=0.01,\n",
    ")\n",
    "\n",
    "trainer = Trainer(\n",
    "    model=model,\n",
    "    args=training_args,\n",
    "    train_dataset=tokenized_datasets['train'],\n",
    "    eval_dataset=tokenized_datasets['validation'],\n",
    "    compute_metrics=compute_metrics\n",
    ")\n",
    "\n",
    "trainer.train()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "id": "684424f0-902a-4dc3-899c-6702af13b018",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/maaly7/anaconda3/envs/general/lib/python3.11/site-packages/torch/nn/parallel/_functions.py:68: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.\n",
      "  warnings.warn('Was asked to gather along dimension 0, but all '\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "    <div>\n",
       "      \n",
       "      <progress value='2' max='2' style='width:300px; height:20px; vertical-align: middle;'></progress>\n",
       "      [2/2 00:00]\n",
       "    </div>\n",
       "    "
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'eval_loss': 0.053683504462242126, 'eval_accuracy': 0.9888888888888889, 'eval_precision': 1.0, 'eval_recall': 0.9285714285714286, 'eval_f1': 0.9629629629629629, 'eval_runtime': 0.897, 'eval_samples_per_second': 200.662, 'eval_steps_per_second': 2.23, 'epoch': 10.0}\n"
     ]
    }
   ],
   "source": [
    "results = trainer.evaluate(tokenized_datasets['test'])\n",
    "print(results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "id": "3365bb3e-6ec5-4e96-b0c9-04834c5785a9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The predicted label for the sequence is: AMR\n"
     ]
    }
   ],
   "source": [
    "import torch\n",
    "\n",
    "# Check if a GPU is available\n",
    "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
    "\n",
    "\n",
    "model.to(device)\n",
    "\n",
    "sequence = \"MTADERDRARSALPFLVITQLMIVLDASIVNIALPSMGRELGMDQTGLQWVVNAYTLTFGGFLMLGGRMADLIGRRLVFVSGICLFGAASLAAALAPVAGVLVAARAVQGLSAAVASAAALSIIVATFPEGKGRNQALAMWGAVSGVGGAVGVLLGGVLTSGPGWPWIFYINVPIVVVVVLGVFRSVSGARGDTRGRLDVAGAVTLTGGLTLLVYAIVSGQSGDPVTILLALGLAVVLLVSFFLVQRKVREPLVPLSSFRNRNLSVASVVGLFAGAAPYAMFFLLSLHLQNVVGLTPLQTGLGFLPVSLISMVGAAALAPLAMARIGMRFTLLLSLGVLAVGLVLLSRLTEEDGFGATVAGQLVAGLGLGTTFVAVTTAAVAGLAENESGLASGLINTAQQLGGALGLGALAALSGAYSAAELAKEPPVSEVAALSSGYQVAFLGAAVFAVAGALIALALPRRESVPATTPHE\" \n",
    "# sequence = \"MSVHIEPIGRFLLAVGVIVAVCHLGGLLCHRIRQPPVIGEIAAGLLLGPTLLGAVAPSLQRALFPEEVLQAVGMAAQLGLVTFMFLLGSELRVDHVRGNGKVVWALVAGSILLPFLAGTGFALLTRPAFGTPQVSTTAYALFVGLAMSITALPVLARILADFRADQSFLGTLALMAAAVGDALAWAALTVILAVTGSGSTGELVLRSALALTLVLLTVFVVKPALRTLLHRLPVNSRVTVPALVVGTTAFAATTEVIGLHPVIGAFLFGCAMPRGSAVLQRASAQLRGFTVSVLLPLFFAGVAMKTAFDAFGTAGNWLLFAAALAVATVTKFVGASSGALLAGLDRARAFQLGALMNCRGVTELVVATVGLQNGFVNEFGYTVLVLIALVTTALTGPLARLRAEEAPQENHRIPMKHGGTFHVRQD\"\n",
    "sequence = re.sub(r'\\s+','',sequence)\n",
    "sequence = ' '.join(list(sequence))\n",
    "\n",
    "inputs = tokenizer(sequence, padding=\"max_length\", truncation=True, max_length=340, return_tensors=\"pt\")\n",
    "inputs = {key: value.to(device) for key, value in inputs.items()}  # Move inputs to the same device\n",
    "\n",
    "\n",
    "model.eval()\n",
    "\n",
    "# Run inference with no_grad() to avoid gradient computations\n",
    "with torch.no_grad():\n",
    "    outputs = model(**inputs)\n",
    "\n",
    "# Convert logits to probabilities using sigmoid (for binary classification)\n",
    "probabilities = torch.sigmoid(outputs.logits)\n",
    "\n",
    "# predictions is likely a 2-element tensor, so you need to get the index of the max value or threshold\n",
    "predictions = (probabilities > 0.5).int()\n",
    "\n",
    "# If it's a single sequence, you can convert the predictions tensor into a Python list\n",
    "# Since this is binary classification, you likely want the first prediction\n",
    "predicted_label = label_map[predictions.argmax().item()]  # Get the predicted class based on the index\n",
    "\n",
    "print(f\"The predicted label for the sequence is: {predicted_label}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a51e6e78-1f53-4eb5-9a9d-f2702bb81270",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
