{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2c395b5-019b-4919-923c-652812b3b4b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ! conda install bioconda::prodigal -y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "463c4f03-cd23-4a9c-b2f6-e203013a9f30",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ! pip install -r requirements.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2122873f-c6fd-4bc3-ad6f-db009e375f39",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ! pip install -r transformers/examples/pytorch/text-classification/requirements.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54519a36-9253-479e-be7e-153a3b069508",
   "metadata": {},
   "outputs": [],
   "source": [
    "from classes.genes import genes\n",
    "genes = genes()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "836384db-a0e8-46b1-bbe4-002dc189dd29",
   "metadata": {},
   "source": [
    "## GENERATE ML DATASET FROM ASSEMBLY ACCESSIONS"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8caa4f05-9cf6-426f-9775-0995db914ea5",
   "metadata": {},
   "source": [
    "#### EXAMPLES\n",
    "- KQ970957.1 - Bacteroides fragilis strain KLE1257 and diabetes - https://www.ncbi.nlm.nih.gov/nuccore/KQ970957.1\n",
    "- AANQ01000001.1 - C. jejuni HB93-13 and neuropathy - https://europepmc.org/article/MED/35442154\n",
    "- NZ_NQLD01000016.1 - Methanobrevibacter smithii and multiple sclerosis - https://www.ncbi.nlm.nih.gov/nuccore/NZ_NQLD01000016.1\n",
    "- JAMGSH000000000.1 - Akkermansia muciniphila and multiple sclerosis - https://www.ncbi.nlm.nih.gov/nuccore/JAMGSH000000000.1\n",
    "- LAPF01000001.1 - S. Typhimurium - https://www.ncbi.nlm.nih.gov/nuccore/LAPF01000001.1\n",
    "- AF047478.1 - Brucella melitensis - https://www.ncbi.nlm.nih.gov/nuccore/AF047478.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "06d08e00-103b-4aed-94b2-2b159b360b5e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No genes were annotated in genbank for this accession GCA_018086425.1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>id</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>gene</th>\n",
       "      <th>product</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>translation</th>\n",
       "      <th>text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [accession, id, locus_tag, gene, product, protein_id, translation, text]\n",
       "Index: []"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### USING ASSEBMLY ACCESSIONS\n",
    "wg_accs = ['GCA_018086425.1']\n",
    "genes_df = genes.genbank(wg_accs)\n",
    "genes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f3aa0563-9945-4a0f-b01c-ab1078f2c50f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "--2024-09-25 15:03:24--  https://www.ebi.ac.uk/ena/browser/api/fasta/GCA_018086425.1?download=true\n",
      "Resolving www.ebi.ac.uk (www.ebi.ac.uk)... 193.62.193.80\n",
      "Connecting to www.ebi.ac.uk (www.ebi.ac.uk)|193.62.193.80|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 \n",
      "Length: unspecified [text/plain]\n",
      "Saving to: ‘fasta/GCA_018086425.1.fna’\n",
      "\n",
      "     0K .......... .......... .......... .......... ..........  174K\n",
      "    50K .......... .......... .......... .......... ..........  262K\n",
      "   100K .......... .......... .......... .......... ..........  521K\n",
      "   150K .......... .......... .......... .......... ..........  523K\n",
      "   200K .......... .......... .......... .......... ..........  523K\n",
      "   250K .......... .......... .......... .......... ..........  521K\n",
      "   300K .......... .......... .......... .......... ..........  522K\n",
      "   350K .......... .......... .......... .......... ..........  531K\n",
      "   400K .......... .......... .......... .......... .......... 26.7M\n",
      "   450K .......... .......... .......... .......... ..........  524K\n",
      "   500K .......... .......... .......... .......... ..........  114M\n",
      "   550K .......... .......... .......... .......... ..........  521K\n",
      "   600K .......... .......... .......... .......... ..........  135M\n",
      "   650K .......... .......... .......... .......... ..........  524K\n",
      "   700K .......... .......... .......... .......... ..........  536K\n",
      "   750K .......... .......... .......... .......... .......... 21.0M\n",
      "   800K .......... .......... .......... .......... ..........  133M\n",
      "   850K .......... .......... .......... .......... ..........  531K\n",
      "   900K .......... .......... .......... .......... .......... 37.1M\n",
      "   950K .......... .......... .......... .......... .......... 98.8M\n",
      "  1000K .......... .......... .......... .......... ..........  532K\n",
      "  1050K .......... .......... .......... .......... .......... 37.8M\n",
      "  1100K .......... .......... .......... .......... ..........  535K\n",
      "  1150K .......... .......... .......... .......... .......... 40.6M\n",
      "  1200K .......... .......... .......... .......... .......... 40.1M\n",
      "  1250K .......... .......... .......... .......... .......... 84.5M\n",
      "  1300K .......... .......... .......... .......... ..........  537K\n",
      "  1350K .......... .......... .......... .......... .......... 52.2M\n",
      "  1400K .......... .......... .......... .......... .......... 37.7M\n",
      "  1450K .......... .......... .......... .......... .......... 96.9M\n",
      "  1500K .......... .......... .......... .......... ..........  534K\n",
      "  1550K .......... .......... .......... .......... .......... 69.1M\n",
      "  1600K .......... .......... .......... .......... .......... 63.9M\n",
      "  1650K .......... .......... .......... .......... ..........  113M\n",
      "  1700K .......... .......... .......... .......... ..........  537K\n",
      "  1750K .......... .......... .......... .......... .......... 42.9M\n",
      "  1800K .......... .......... .......... .......... .......... 48.5M\n",
      "  1850K .......... .......... .......... .......... ..........  109M\n",
      "  1900K .......... .......... .......... .......... ..........  537K\n",
      "  1950K .......... .......... .......... .......... .......... 43.0M\n",
      "  2000K .......... .......... .......... .......... .......... 96.3M\n",
      "  2050K .......... .......... .......... .......... .......... 87.9M\n",
      "  2100K .......... .......... .......... .......... .......... 35.4M\n",
      "  2150K .......... .......... .......... .......... ..........  543K\n",
      "  2200K .......... .......... .......... .......... .......... 43.9M\n",
      "  2250K .......... .......... .......... .......... .......... 70.3M\n",
      "  2300K .......... .......... .......... .......... ..........  111M\n",
      "  2350K .......... .......... .......... .......... .......... 39.3M\n",
      "  2400K .......... .......... .......... .......... ..........  169M\n",
      "  2450K .......... .......... .......... .......... ..........  543K\n",
      "  2500K .......... .......... .......... .......... .......... 45.6M\n",
      "  2550K .......... .......... .......... .......... .......... 99.2M\n",
      "  2600K .......... .......... .......... .......... .......... 91.3M\n",
      "  2650K .......... .......... .......... .......... .......... 41.9M\n",
      "  2700K .......... .......... .......... .......... ..........  135M\n",
      "  2750K .......... .......... .......... .......... ..........  544K\n",
      "  2800K .......... .......... .......... .......... .......... 41.6M\n",
      "  2850K .......... .......... .......... .......... ..........  183M\n",
      "  2900K .......... .......... .......... .......... ..........  132M\n",
      "  2950K .......... .......... .......... .......... .......... 68.8M\n",
      "  3000K .......... .......... .......... .......... .....      49.5M=2.4s\n",
      "\n",
      "2024-09-25 15:03:27 (1.24 MB/s) - ‘fasta/GCA_018086425.1.fna’ saved [3118821]\n",
      "\n",
      "-------------------------------------\n",
      "PRODIGAL v2.6.3 [February, 2016]         \n",
      "Univ of Tenn / Oak Ridge National Lab\n",
      "Doug Hyatt, Loren Hauser, et al.     \n",
      "-------------------------------------\n",
      "Request:  Single Genome, Phase:  Training\n",
      "Reading in the sequence(s) to train...3036002 bp seq created, 37.69 pct GC\n",
      "Locating all potential starts and stops...121957 nodes\n",
      "Looking for GC bias in different frames...frame bias scores: 2.60 0.17 0.23\n",
      "Building initial set of genes to train from...done!\n",
      "Creating coding model and scoring nodes...done!\n",
      "Examining upstream regions and training starts...done!\n",
      "-------------------------------------\n",
      "Request:  Single Genome, Phase:  Gene Finding\n",
      "Finding genes in sequence #1 (157108 bp)...done!\n",
      "Finding genes in sequence #2 (145294 bp)...done!\n",
      "Finding genes in sequence #3 (110486 bp)...done!\n",
      "Finding genes in sequence #4 (101870 bp)...done!\n",
      "Finding genes in sequence #5 (86098 bp)...done!\n",
      "Finding genes in sequence #6 (83208 bp)...done!\n",
      "Finding genes in sequence #7 (72634 bp)...done!\n",
      "Finding genes in sequence #8 (72433 bp)...done!\n",
      "Finding genes in sequence #9 (72141 bp)...done!\n",
      "Finding genes in sequence #10 (68212 bp)...done!\n",
      "Finding genes in sequence #11 (53759 bp)...done!\n",
      "Finding genes in sequence #12 (53394 bp)...done!\n",
      "Finding genes in sequence #13 (52623 bp)...done!\n",
      "Finding genes in sequence #14 (51226 bp)...done!\n",
      "Finding genes in sequence #15 (48707 bp)...done!\n",
      "Finding genes in sequence #16 (47063 bp)...done!\n",
      "Finding genes in sequence #17 (43679 bp)...done!\n",
      "Finding genes in sequence #18 (42112 bp)...done!\n",
      "Finding genes in sequence #19 (37943 bp)...done!\n",
      "Finding genes in sequence #20 (36804 bp)...done!\n",
      "Finding genes in sequence #21 (36568 bp)...done!\n",
      "Finding genes in sequence #22 (36230 bp)...done!\n",
      "Finding genes in sequence #23 (35719 bp)...done!\n",
      "Finding genes in sequence #24 (34636 bp)...done!\n",
      "Finding genes in sequence #25 (34196 bp)...done!\n",
      "Finding genes in sequence #26 (32310 bp)...done!\n",
      "Finding genes in sequence #27 (32037 bp)...done!\n",
      "Finding genes in sequence #28 (29498 bp)...done!\n",
      "Finding genes in sequence #29 (29142 bp)...done!\n",
      "Finding genes in sequence #30 (28107 bp)...done!\n",
      "Finding genes in sequence #31 (27214 bp)...done!\n",
      "Finding genes in sequence #32 (26493 bp)...done!\n",
      "Finding genes in sequence #33 (25807 bp)...done!\n",
      "Finding genes in sequence #34 (24395 bp)...done!\n",
      "Finding genes in sequence #35 (24081 bp)...done!\n",
      "Finding genes in sequence #36 (23639 bp)...done!\n",
      "Finding genes in sequence #37 (23579 bp)...done!\n",
      "Finding genes in sequence #38 (23571 bp)...done!\n",
      "Finding genes in sequence #39 (21935 bp)...done!\n",
      "Finding genes in sequence #40 (21358 bp)...done!\n",
      "Finding genes in sequence #41 (21309 bp)...done!\n",
      "Finding genes in sequence #42 (20799 bp)...done!\n",
      "Finding genes in sequence #43 (20784 bp)...done!\n",
      "Finding genes in sequence #44 (19578 bp)...done!\n",
      "Finding genes in sequence #45 (19566 bp)...done!\n",
      "Finding genes in sequence #46 (19452 bp)...done!\n",
      "Finding genes in sequence #47 (19059 bp)...done!\n",
      "Finding genes in sequence #48 (18895 bp)...done!\n",
      "Finding genes in sequence #49 (18379 bp)...done!\n",
      "Finding genes in sequence #50 (18094 bp)...done!\n",
      "Finding genes in sequence #51 (17751 bp)...done!\n",
      "Finding genes in sequence #52 (17689 bp)...done!\n",
      "Finding genes in sequence #53 (17589 bp)...done!\n",
      "Finding genes in sequence #54 (17308 bp)...done!\n",
      "Finding genes in sequence #55 (17253 bp)...done!\n",
      "Finding genes in sequence #56 (17005 bp)...done!\n",
      "Finding genes in sequence #57 (16679 bp)...done!\n",
      "Finding genes in sequence #58 (16387 bp)...done!\n",
      "Finding genes in sequence #59 (16231 bp)...done!\n",
      "Finding genes in sequence #60 (16075 bp)...done!\n",
      "Finding genes in sequence #61 (15963 bp)...done!\n",
      "Finding genes in sequence #62 (14817 bp)...done!\n",
      "Finding genes in sequence #63 (14114 bp)...done!\n",
      "Finding genes in sequence #64 (13917 bp)...done!\n",
      "Finding genes in sequence #65 (13725 bp)...done!\n",
      "Finding genes in sequence #66 (13451 bp)...done!\n",
      "Finding genes in sequence #67 (13291 bp)...done!\n",
      "Finding genes in sequence #68 (13146 bp)...done!\n",
      "Finding genes in sequence #69 (12946 bp)...done!\n",
      "Finding genes in sequence #70 (12670 bp)...done!\n",
      "Finding genes in sequence #71 (12583 bp)...done!\n",
      "Finding genes in sequence #72 (11749 bp)...done!\n",
      "Finding genes in sequence #73 (11255 bp)...done!\n",
      "Finding genes in sequence #74 (10661 bp)...done!\n",
      "Finding genes in sequence #75 (10496 bp)...done!\n",
      "Finding genes in sequence #76 (10034 bp)...done!\n",
      "Finding genes in sequence #77 (9967 bp)...done!\n",
      "Finding genes in sequence #78 (9665 bp)...done!\n",
      "Finding genes in sequence #79 (9210 bp)...done!\n",
      "Finding genes in sequence #80 (8555 bp)...done!\n",
      "Finding genes in sequence #81 (8555 bp)...done!\n",
      "Finding genes in sequence #82 (8528 bp)...done!\n",
      "Finding genes in sequence #83 (8292 bp)...done!\n",
      "Finding genes in sequence #84 (7763 bp)...done!\n",
      "Finding genes in sequence #85 (7760 bp)...done!\n",
      "Finding genes in sequence #86 (7600 bp)...done!\n",
      "Finding genes in sequence #87 (7353 bp)...done!\n",
      "Finding genes in sequence #88 (6643 bp)...done!\n",
      "Finding genes in sequence #89 (6569 bp)...done!\n",
      "Finding genes in sequence #90 (6281 bp)...done!\n",
      "Finding genes in sequence #91 (6135 bp)...done!\n",
      "Finding genes in sequence #92 (5892 bp)...done!\n",
      "Finding genes in sequence #93 (5873 bp)...done!\n",
      "Finding genes in sequence #94 (5796 bp)...done!\n",
      "Finding genes in sequence #95 (5758 bp)...done!\n",
      "Finding genes in sequence #96 (5719 bp)...done!\n",
      "Finding genes in sequence #97 (5577 bp)...done!\n",
      "Finding genes in sequence #98 (5556 bp)...done!\n",
      "Finding genes in sequence #99 (5341 bp)...done!\n",
      "Finding genes in sequence #100 (5257 bp)...done!\n",
      "Finding genes in sequence #101 (4937 bp)...done!\n",
      "Finding genes in sequence #102 (4807 bp)...done!\n",
      "Finding genes in sequence #103 (4798 bp)...done!\n",
      "Finding genes in sequence #104 (4416 bp)...done!\n",
      "Finding genes in sequence #105 (4397 bp)...done!\n",
      "Finding genes in sequence #106 (4170 bp)...done!\n",
      "Finding genes in sequence #107 (4056 bp)...done!\n",
      "Finding genes in sequence #108 (3923 bp)...done!\n",
      "Finding genes in sequence #109 (3872 bp)...done!\n",
      "Finding genes in sequence #110 (3771 bp)...done!\n",
      "Finding genes in sequence #111 (3728 bp)...done!\n",
      "Finding genes in sequence #112 (3719 bp)...done!\n",
      "Finding genes in sequence #113 (3404 bp)...done!\n",
      "Finding genes in sequence #114 (3311 bp)...done!\n",
      "Finding genes in sequence #115 (3274 bp)...done!\n",
      "Finding genes in sequence #116 (3146 bp)...done!\n",
      "Finding genes in sequence #117 (3142 bp)...done!\n",
      "Finding genes in sequence #118 (3043 bp)...done!\n",
      "Finding genes in sequence #119 (3022 bp)...done!\n",
      "Finding genes in sequence #120 (3021 bp)...done!\n",
      "Finding genes in sequence #121 (2776 bp)...done!\n",
      "Finding genes in sequence #122 (2745 bp)...done!\n",
      "Finding genes in sequence #123 (2717 bp)...done!\n",
      "Finding genes in sequence #124 (2645 bp)...done!\n",
      "Finding genes in sequence #125 (2618 bp)...done!\n",
      "Finding genes in sequence #126 (2484 bp)...done!\n",
      "Finding genes in sequence #127 (2398 bp)...done!\n",
      "Finding genes in sequence #128 (2316 bp)...done!\n",
      "Finding genes in sequence #129 (2313 bp)...done!\n",
      "Finding genes in sequence #130 (2300 bp)...done!\n",
      "Finding genes in sequence #131 (2259 bp)...done!\n",
      "Finding genes in sequence #132 (2001 bp)...done!\n",
      "Finding genes in sequence #133 (1912 bp)...done!\n",
      "Finding genes in sequence #134 (1897 bp)...done!\n",
      "Finding genes in sequence #135 (1829 bp)...done!\n",
      "Finding genes in sequence #136 (1772 bp)...done!\n",
      "Finding genes in sequence #137 (1748 bp)...done!\n",
      "Finding genes in sequence #138 (1704 bp)...done!\n",
      "Finding genes in sequence #139 (1703 bp)...done!\n",
      "Finding genes in sequence #140 (1688 bp)...done!\n",
      "Finding genes in sequence #141 (1583 bp)...done!\n",
      "Finding genes in sequence #142 (1466 bp)...done!\n",
      "Finding genes in sequence #143 (1453 bp)...done!\n",
      "Finding genes in sequence #144 (1432 bp)...done!\n",
      "Finding genes in sequence #145 (1424 bp)...done!\n",
      "Finding genes in sequence #146 (1414 bp)...done!\n",
      "Finding genes in sequence #147 (1324 bp)...done!\n",
      "Finding genes in sequence #148 (1215 bp)...done!\n",
      "Finding genes in sequence #149 (1194 bp)...done!\n",
      "Finding genes in sequence #150 (1165 bp)...done!\n",
      "Finding genes in sequence #151 (1161 bp)...done!\n",
      "Finding genes in sequence #152 (1127 bp)...done!\n",
      "Finding genes in sequence #153 (1058 bp)...done!\n",
      "Finding genes in sequence #154 (1024 bp)...done!\n",
      "Finding genes in sequence #155 (986 bp)...done!\n",
      "Finding genes in sequence #156 (985 bp)...done!\n",
      "Finding genes in sequence #157 (975 bp)...done!\n",
      "Finding genes in sequence #158 (970 bp)...done!\n",
      "Finding genes in sequence #159 (900 bp)...done!\n",
      "Finding genes in sequence #160 (865 bp)...done!\n",
      "Finding genes in sequence #161 (863 bp)...done!\n",
      "Finding genes in sequence #162 (861 bp)...done!\n",
      "Finding genes in sequence #163 (859 bp)...done!\n",
      "Finding genes in sequence #164 (843 bp)...done!\n",
      "Finding genes in sequence #165 (825 bp)...done!\n",
      "Finding genes in sequence #166 (823 bp)...done!\n",
      "Finding genes in sequence #167 (821 bp)...done!\n",
      "Finding genes in sequence #168 (813 bp)...done!\n",
      "Finding genes in sequence #169 (793 bp)...done!\n",
      "Finding genes in sequence #170 (773 bp)...done!\n",
      "Finding genes in sequence #171 (639 bp)...done!\n",
      "Finding genes in sequence #172 (593 bp)...done!\n",
      "Finding genes in sequence #173 (583 bp)...done!\n",
      "Finding genes in sequence #174 (581 bp)...done!\n",
      "Finding genes in sequence #175 (548 bp)...done!\n",
      "Finding genes in sequence #176 (531 bp)...done!\n",
      "Finding genes in sequence #177 (529 bp)...done!\n",
      "Finding genes in sequence #178 (507 bp)...done!\n",
      "Finding genes in sequence #179 (503 bp)...done!\n",
      "Finding genes in sequence #180 (498 bp)...done!\n",
      "Finding genes in sequence #181 (498 bp)...done!\n",
      "Finding genes in sequence #182 (487 bp)...done!\n",
      "Finding genes in sequence #183 (476 bp)...done!\n",
      "Finding genes in sequence #184 (468 bp)...done!\n",
      "Finding genes in sequence #185 (455 bp)...done!\n",
      "Finding genes in sequence #186 (450 bp)...done!\n",
      "Finding genes in sequence #187 (412 bp)...done!\n",
      "Finding genes in sequence #188 (375 bp)...done!\n",
      "Finding genes in sequence #189 (368 bp)...done!\n",
      "Finding genes in sequence #190 (358 bp)...done!\n",
      "Finding genes in sequence #191 (341 bp)...done!\n",
      "Finding genes in sequence #192 (333 bp)...done!\n",
      "Finding genes in sequence #193 (321 bp)...done!\n",
      "Finding genes in sequence #194 (317 bp)...done!\n",
      "Finding genes in sequence #195 (309 bp)...done!\n",
      "Finding genes in sequence #196 (302 bp)...done!\n",
      "Finding genes in sequence #197 (286 bp)...done!\n",
      "Finding genes in sequence #198 (268 bp)...done!\n",
      "Finding genes in sequence #199 (241 bp)...done!\n",
      "Finding genes in sequence #200 (235 bp)...done!\n",
      "Finding genes in sequence #201 (20985 bp)...done!\n",
      "Finding genes in sequence #202 (12882 bp)...done!\n",
      "Finding genes in sequence #203 (10113 bp)...done!\n",
      "Finding genes in sequence #204 (9772 bp)...done!\n",
      "Finding genes in sequence #205 (7959 bp)...done!\n",
      "Finding genes in sequence #206 (4742 bp)...done!\n",
      "Finding genes in sequence #207 (4479 bp)...done!\n",
      "Finding genes in sequence #208 (3356 bp)...done!\n",
      "Finding genes in sequence #209 (3107 bp)...done!\n",
      "Finding genes in sequence #210 (2973 bp)...done!\n",
      "Finding genes in sequence #211 (2973 bp)...done!\n",
      "Finding genes in sequence #212 (2973 bp)...done!\n",
      "Finding genes in sequence #213 (2438 bp)...done!\n",
      "Finding genes in sequence #214 (2360 bp)...done!\n",
      "Finding genes in sequence #215 (1781 bp)...done!\n",
      "Finding genes in sequence #216 (1727 bp)...done!\n",
      "Finding genes in sequence #217 (1689 bp)...done!\n",
      "Finding genes in sequence #218 (1689 bp)...done!\n",
      "Finding genes in sequence #219 (1689 bp)...done!\n",
      "Finding genes in sequence #220 (1689 bp)...done!\n",
      "Finding genes in sequence #221 (1689 bp)...done!\n",
      "Finding genes in sequence #222 (1689 bp)...done!\n",
      "Finding genes in sequence #223 (1689 bp)...done!\n",
      "Finding genes in sequence #224 (1670 bp)...done!\n",
      "Finding genes in sequence #225 (1670 bp)...done!\n",
      "Finding genes in sequence #226 (1670 bp)...done!\n",
      "Finding genes in sequence #227 (1664 bp)...done!\n",
      "Finding genes in sequence #228 (1664 bp)...done!\n",
      "Finding genes in sequence #229 (1660 bp)...done!\n",
      "Finding genes in sequence #230 (1638 bp)...done!\n",
      "Finding genes in sequence #231 (1633 bp)...done!\n",
      "Finding genes in sequence #232 (626 bp)...done!\n",
      "Finding genes in sequence #233 (625 bp)...done!\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>id</th>\n",
       "      <th>translation</th>\n",
       "      <th>text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_018086425.1</td>\n",
       "      <td>00001</td>\n",
       "      <td>MEVKVYGSAHCVTSLKAKQWLEHHQLPYQFVDLEERDLSTEEAEEL...</td>\n",
       "      <td>M E V K V Y G S A H C V T S L K A K Q W L E H ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_018086425.1</td>\n",
       "      <td>00002</td>\n",
       "      <td>MIKVYTSPSCTSCRKARAWLQANQLEFEEKNIFAEPLTESEIKKIL...</td>\n",
       "      <td>M I K V Y T S P S C T S C R K A R A W L Q A N ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_018086425.1</td>\n",
       "      <td>00003</td>\n",
       "      <td>MNEKEILRLQKELKRWENKYKKLQQVFRLQEEIIATYEAAYGEAPG...</td>\n",
       "      <td>M N E K E I L R L Q K E L K R W E N K Y K K L ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_018086425.1</td>\n",
       "      <td>00004</td>\n",
       "      <td>MQTSFHEGRTIGLLLTFIGGAMDSYTYIQYGAFASAQTGNIILSII...</td>\n",
       "      <td>M Q T S F H E G R T I G L L L T F I G G A M D ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCA_018086425.1</td>\n",
       "      <td>00005</td>\n",
       "      <td>MIIVAGTIGAGKSTLTEMLAQDLETKPFYENVEDNEVLPLFYSNPE...</td>\n",
       "      <td>M I I V A G T I G A G K S T L T E M L A Q D L ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3004</th>\n",
       "      <td>GCA_018086425.1</td>\n",
       "      <td>00002</td>\n",
       "      <td>MEIGFTFLDEIVHGVRWDAKYATWDNFTGKPVDGYEVNRIVGTYEL...</td>\n",
       "      <td>M E I G F T F L D E I V H G V R W D A K Y A T ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3005</th>\n",
       "      <td>GCA_018086425.1</td>\n",
       "      <td>00001</td>\n",
       "      <td>MHGLLVKKNHEYEINHVDVAFSALHGKSGEDGSIQGLFELSGIPFV...</td>\n",
       "      <td>M H G L L V K K N H E Y E I N H V D V A F S A ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3006</th>\n",
       "      <td>GCA_018086425.1</td>\n",
       "      <td>00002</td>\n",
       "      <td>MEIGFTFLDEIVHGVRWDAKYATWDNFTGKPVDGYEVNRIVGTYEL...</td>\n",
       "      <td>M E I G F T F L D E I V H G V R W D A K Y A T ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3007</th>\n",
       "      <td>GCA_018086425.1</td>\n",
       "      <td>00001</td>\n",
       "      <td>MIGLIVARSKNNVIGKNGNIPWKIKGEQKQFRELTTGNVVIMGRKS...</td>\n",
       "      <td>M I G L I V A R S K N N V I G K N G N I P W K ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3008</th>\n",
       "      <td>GCA_018086425.1</td>\n",
       "      <td>00001</td>\n",
       "      <td>MIGLIVARSKNNVIGKNGNIPWKIKGEQKQFRELTTGNVVIMGRKS...</td>\n",
       "      <td>M I G L I V A R S K N N V I G K N G N I P W K ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3009 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            accession     id  \\\n",
       "0     GCA_018086425.1  00001   \n",
       "1     GCA_018086425.1  00002   \n",
       "2     GCA_018086425.1  00003   \n",
       "3     GCA_018086425.1  00004   \n",
       "4     GCA_018086425.1  00005   \n",
       "...               ...    ...   \n",
       "3004  GCA_018086425.1  00002   \n",
       "3005  GCA_018086425.1  00001   \n",
       "3006  GCA_018086425.1  00002   \n",
       "3007  GCA_018086425.1  00001   \n",
       "3008  GCA_018086425.1  00001   \n",
       "\n",
       "                                            translation  \\\n",
       "0     MEVKVYGSAHCVTSLKAKQWLEHHQLPYQFVDLEERDLSTEEAEEL...   \n",
       "1     MIKVYTSPSCTSCRKARAWLQANQLEFEEKNIFAEPLTESEIKKIL...   \n",
       "2     MNEKEILRLQKELKRWENKYKKLQQVFRLQEEIIATYEAAYGEAPG...   \n",
       "3     MQTSFHEGRTIGLLLTFIGGAMDSYTYIQYGAFASAQTGNIILSII...   \n",
       "4     MIIVAGTIGAGKSTLTEMLAQDLETKPFYENVEDNEVLPLFYSNPE...   \n",
       "...                                                 ...   \n",
       "3004  MEIGFTFLDEIVHGVRWDAKYATWDNFTGKPVDGYEVNRIVGTYEL...   \n",
       "3005  MHGLLVKKNHEYEINHVDVAFSALHGKSGEDGSIQGLFELSGIPFV...   \n",
       "3006  MEIGFTFLDEIVHGVRWDAKYATWDNFTGKPVDGYEVNRIVGTYEL...   \n",
       "3007  MIGLIVARSKNNVIGKNGNIPWKIKGEQKQFRELTTGNVVIMGRKS...   \n",
       "3008  MIGLIVARSKNNVIGKNGNIPWKIKGEQKQFRELTTGNVVIMGRKS...   \n",
       "\n",
       "                                                   text  \n",
       "0     M E V K V Y G S A H C V T S L K A K Q W L E H ...  \n",
       "1     M I K V Y T S P S C T S C R K A R A W L Q A N ...  \n",
       "2     M N E K E I L R L Q K E L K R W E N K Y K K L ...  \n",
       "3     M Q T S F H E G R T I G L L L T F I G G A M D ...  \n",
       "4     M I I V A G T I G A G K S T L T E M L A Q D L ...  \n",
       "...                                                 ...  \n",
       "3004  M E I G F T F L D E I V H G V R W D A K Y A T ...  \n",
       "3005  M H G L L V K K N H E Y E I N H V D V A F S A ...  \n",
       "3006  M E I G F T F L D E I V H G V R W D A K Y A T ...  \n",
       "3007  M I G L I V A R S K N N V I G K N G N I P W K ...  \n",
       "3008  M I G L I V A R S K N N V I G K N G N I P W K ...  \n",
       "\n",
       "[3009 rows x 4 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### USING PRODIGAL\n",
    "wg_accs = ['GCA_018086425.1']\n",
    "genes_df = genes.prodigal(wg_accs,'fasta')\n",
    "genes_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43e933e7-9e48-4eb9-978b-583514e38e23",
   "metadata": {},
   "source": [
    "#### AMR ASSEMBLY GENES WITH THEIR TRANSLATIONS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a567e12-db30-41cb-9af4-d07c6d2e67b1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "031495fa-3224-4b6d-83e0-50f719fbc2b0",
   "metadata": {},
   "source": [
    "#### AMR ML DATASET"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0fe96413-87d6-4256-97a7-a44df8573627",
   "metadata": {},
   "outputs": [],
   "source": [
    "from datasets import load_dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7a5e117-2e25-4938-92b1-05a9177e87e6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "7f93d47e-f4de-433b-877f-3b854b7eafd1",
   "metadata": {},
   "source": [
    "#### TRANSFORMER CLASSIFIER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e5f11f8-663b-4b30-99ab-26540f3a7559",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ! git clone https://github.com/huggingface/transformers.git"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d65491f7-f95a-4a56-a611-ce3ba2a25b02",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ! pip install -r transformers/examples/pytorch/text-classification/requirements.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d687e59-eda0-4c6c-b317-3a17e0d3ab12",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
